// @ncigdc/routes/legacy/SmartSearchRoute.js
/* @flow */
/* global _ angular */
/* eslint
  fp/no-class:0,
  fp/no-this:0,
  fp/no-mutation:0,
  class-methods-use-this:0,
  react/prefer-stateless-function:0,
  max-len:0,
  flowtype/require-parameter-type:0,
  flowtype/require-return-type:0,
  no-param-reassign:0,
  no-useless-escape:0,
  no-unused-vars:1,
  no-else-return:0,
*/

import React from 'react';

_.pluck = _.map;

angular
  .module('legacyAngularWrapper', ['ngApp'])
  .config(($locationProvider: ng.ILocationProvider) => {
    $locationProvider.html5Mode({
      enabled: true,
      requireBase: false,
    });
  })
  .directive('smartSearchWrapper', () => ({
    controller: 'QueryController as qc',
    templateUrl: 'query/templates/query.html',
  }))
  ;

const angularBootstrapHtml = `
  <smart-search-wrapper></smart-search-wrapper>
`;

export default class extends React.Component {
  componentDidMount() {
    angular.bootstrap(this.container, ['legacyAngularWrapper']);
  }

  render(): void {
    return (
      <div
        ref={c => { this.container = c; }}
        dangerouslySetInnerHTML={{ __html: angularBootstrapHtml }}
      />
    );
  }
}
