// @ncigdc/components/Modals/NoAccessModal.js
// @flow
import React from 'react';
import { connect } from 'react-redux';

import { setModal } from '@ncigdc/dux/modal';
import Button from '@ncigdc/uikit/Button';
import { Column, Row } from '@ncigdc/uikit/Flex';
import LoginButton from '../LoginButton';

const LoginModal = ({
  dispatch,
  message = "You don't have access",
  primaryButton,
}: {
  dispatch: Function,
  message: string,
}) => (
  <Column>
    <h2 style={{ paddingLeft: 15 }}>Access Error</h2>
    <Column
      style={{
        borderBottom: '1px solid #e5e5e5',
        borderTop: '1px solid #e5e5e5',
        padding: 15,
        marginBottom: 15,
      }}
    >
      {message}
      <p>Please <LoginButton /></p>
    </Column>
    <Row
      style={{
        marginBottom: 15,
      }}
    >
      <Button
        style={{ margin: '0 10px 0 auto' }}
        onClick={() => dispatch(setModal(null))}
      >
        Cancel
      </Button>
      {primaryButton}
    </Row>
  </Column>
);

export default connect()(LoginModal);
