// @ncigdc/components/Modals/SessionExpiredModal.js
// @flow
import React from 'react';
import { connect } from 'react-redux';

import { setModal } from '@ncigdc/dux/modal';
import Button from '@ncigdc/uikit/Button';
import { Column, Row } from '@ncigdc/uikit/Flex';
import { resetError } from '@ncigdc/dux/error';
import LoginButton from '../LoginButton';

const SessionExpired = ({
  dispatch,
}: {
  dispatch: Function,
  message: string,
}) => (
  <Column>
    <h2 style={{ paddingLeft: 15 }}>Session Expired</h2>
    <Column
      style={{
        borderBottom: '1px solid #e5e5e5',
        borderTop: '1px solid #e5e5e5',
        padding: 15,
        marginBottom: 15,
      }}
    >
      Your session has expired.
      <p>Please <LoginButton /></p>
    </Column>
    <Row
      style={{
        marginBottom: 15,
      }}
    >
      <Button
        style={{ margin: '0 10px 0 auto' }}
        onClick={() => {
          dispatch(resetError());
          dispatch(setModal(null));
        }}
      >
        Cancel
      </Button>
    </Row>
  </Column>
);

export default connect()(SessionExpired);
