// @ncigdc/components/AddToCartButtonAll.js
// @flow

import React from 'react';
import { connect } from 'react-redux';
import { style, hover } from 'glamor';

import _ from 'lodash';

import ShoppingCartIcon from 'react-icons/lib/fa/shopping-cart';
import DownCaretIcon from 'react-icons/lib/fa/caret-down';

import { addAllFilesInCart, toggleFilesInCart } from '@ncigdc/dux/cart';

import Row from '@ncigdc/uikit/Flex/Row';
import Dropdown from '@ncigdc/uikit/Dropdown';
import Button from '@ncigdc/uikit/Button';

import styled from '@ncigdc/theme/styled';

const styles= {
  cart: {
    padding: '0 5px',
    color: '#FFF',
    borderRadius: '4px',
    borderTopRightRadius: '0px',
    borderBottomRightRadius: '0px',
    cursor: 'pointer',
  },
  addToCart: {
    backgroundColor: '#0d95a1',
  },
  removeFromCart: {
    backgroundColor: '#255425',
  }
};

const DropDownCaret = styled.span({
  backgroundColor: '#FFF',
  borderTopRightRadius: '4px',
  borderBottomRightRadius: '4px',
});

const filesInCart = (edges, files) => {
  if(files.length < edges.length) return false;
  else return _.every(edges, edge => _.some(files, {'file_id':edge.file_id}))
}

const AddToCartButtonAll = ({ edges, files, dispatch }) => (
  <Row>
    <Button
      onClick={() => filesInCart(edges, files) ? dispatch(toggleFilesInCart(edges)) : dispatch(addAllFilesInCart(edges))}
      {...style({ ...styles.cart, ...(filesInCart(edges, files) ? styles.removeFromCart : styles.addToCart) })}
      {...hover({ ...(filesInCart(edges, files) ? styles.removeFromCart : styles.addToCart) })}
    >
      <ShoppingCartIcon />
    </Button>
    <Dropdown
      button={
        <Row>
          <DropDownCaret>
            <DownCaretIcon style={{ marginLeft: 'auto' }} />
          </DropDownCaret>
        </Row>
      }
    >
      <span>All</span>
      <span>Some</span>
    </Dropdown>
  </Row>
);

export default connect(state => state.cart)(AddToCartButtonAll);
