// @ncigdc/components/AddToCartButtonAll.js
// @flow

import React from 'react';
import { connect } from 'react-redux';

import ShoppingCartIcon from 'react-icons/lib/fa/shopping-cart';
import DownCaretIcon from 'react-icons/lib/fa/caret-down';

import { toggleInCart } from '@ncigdc/dux/cart';

import Row from '@ncigdc/uikit/Flex/Row';
import Dropdown from '@ncigdc/uikit/Dropdown';

import styled from '@ncigdc/theme/styled';

const ShoppingCart = styled.span({
  padding: '0 5px',
  color: '#FFF',
  backgroundColor: '#0d95a1',
  borderTopLeftRadius: '4px',
  borderBottomLeftRadius: '4px',
  cursor: 'pointer',
});

const DropDownCaret = styled.span({
  backgroundColor: '#FFF',
  borderTopRightRadius: '4px',
  borderBottomRightRadius: '4px',
});

const AddToCartButtonAll = ({ edges, files, dispatch }) => (
  <Row>
    <ShoppingCart
      onClick={() => dispatch(toggleInCart(edges))}
    >
      <ShoppingCartIcon />
    </ShoppingCart>
    <Dropdown
      button={
        <Row>
          <DropDownCaret>
            <DownCaretIcon style={{ marginLeft: 'auto' }} />
          </DropDownCaret>
        </Row>
      }
    >
      <span>All</span>
      <span>Some</span>
    </Dropdown>
  </Row>
);

export default connect(state => state.cart)(AddToCartButtonAll);
