// @ncigdc/components/BiospecimenCard.js
// @flow

import React from 'react';
import _ from 'lodash';
import SearchIcon from 'react-icons/lib/fa/search';
import { compose, withState } from 'recompose';

import { search, idFields, formatValue } from '@ncigdc/utils/biotree';

import Card from '@ncigdc/uikit/Card';
import { Row, Column } from '@ncigdc/uikit/Flex';
import Input from '@ncigdc/uikit/Form/Input';

import EntityPageVerticalTable from '@ncigdc/components/EntityPageVerticalTable';
import BioTreeView from '@ncigdc/components/BioTreeView';

import { withTheme } from '@ncigdc/theme';

const styles = {
  cardHeader: theme => ({
    padding: '1rem',
    color: theme.greyScale7,
  }),
  button: {
    color: '#333',
    backgroundColor: '#fff',
    borderColor: '#ccc',
    minWidth: '115px',
    minHeight: '34px',
    display: 'inline-flex',
    outline: 'none',
  },
  searchIcon: theme => ({
    backgroundColor: theme.greyScale5,
    color: theme.greyScale2,
    padding: '0.8rem',
    width: '3.4rem',
    height: '3.4rem',
    borderRadius: '4px 0 0 4px',
    border: `1px solid ${theme.greyScale4}`,
    borderRight: 'none',
  }),
};

const BiospecimenCard = ({ p, state: { selectedEntity: se, type, query }, setState, theme }) => {
  const selectedEntity =
    (query && (_.flatten(p.samples.map(e => search(query, e))) || [])[0]) || se;

  return (
    <Card title="Biospecimen" headerStyle={styles.cardHeader(theme)} style={{ flex: 1 }}>
      <Row>
        <Column flex="3" style={{ padding: '0 15px' }}>
          <Row spacing="30px">
            <Row flex="1">
              <SearchIcon style={styles.searchIcon(theme)} />
              <Input
                onChange={({ target }) => setState(s => ({ ...s, query: target.value }))}
                placeholder="Search"
              />
            </Row>
          </Row>

          <Column style={{ padding: '10px' }}>
            <BioTreeView
              entities={p.samples}
              type={{ s: 'sample', p: 'samples' }}
              query={query}
              selectedEntity={selectedEntity}
              selectEntity={selectedEntity => setState(s => ({ ...s, selectedEntity }))}
              defaultExpanded
            />
          </Column>

        </Column>
        <Column flex="4">
          <EntityPageVerticalTable
            thToTd={[
              { th: 'Submitter ID', td: selectedEntity.submitter_id },
              { th: type, td: selectedEntity[idFields.find(id => selectedEntity[id])] },
              ...Object.entries(selectedEntity)
                .filter(([key]) =>
                  key !== 'submitter_id' && key !== 'expanded' && key !== `${type}_id`
                )
                .map(([key, val]) => ({
                  th: key, td: formatValue(val),
                })),
            ]}
            style={{ flex: 1 }}
          />
        </Column>
      </Row>
    </Card>
  );
};

export default compose(
  withState(
    'state',
    'setState',
    ({ p }) => ({
      selectedEntity: p.samples[0],
      type: 'sample',
      query: '',
    })
  ),
  withTheme
)(BiospecimenCard);
