// @ncigdc/components/CancerDistribution.js
// @flow

import React from 'react';
import withSize from 'react-sizeme';
import { compose } from 'recompose';

import { Row, Column } from '@ncigdc/uikit/Flex';
import { withTheme } from '@ncigdc/theme';

import BarChart from './Charts/BarChart';
import DownloadVisualizationButton from './DownloadVisualizationButton';
import EntityPageHorizontalTable from './EntityPageHorizontalTable';

const styles = {
  column: {
    width: '100%',
    minWidth: 450,
  },
};

const enhance = compose(
  withSize(),
  withTheme
);

const CancerDistribution = ({
  chartData,
  tableData,
  tableHeadings,
  tagline,
  theme,
  size: {
    width,
  },
}) => {
  const chartMargin = { top: 30, right: 50, bottom: 105, left: 40 };
  const bandWidth = (width - chartMargin.right - chartMargin.left) / (chartData.length + 1) * 0.7;

  return (
    <Column>
      <h5 style={{ textTransform: 'uppercase', padding: '0 2rem' }}>
        {tagline}
      </h5>
      {chartData.length >= 5 && <Row style={{ padding: '0 2rem', justifyContent: 'center' }}>
        <span>
          <div style={{ textAlign: 'right' }}>
            <DownloadVisualizationButton
              svg="#cancer-distribution svg"
              data={chartData.map(d => ({ label: d.label, value: d.value }))}
              slug="bar-chart"
              noText
              tooltipHTML="Download image or data"
            />
          </div>
          <BarChart
            margin={chartMargin}
            bandwidth={bandWidth}
            data={chartData}
            yAxis={{ title: '% of Cases Affected' }}
            styles={{
              xAxis: { stroke: theme.greyScale4, textFill: theme.greyScale3 },
              yAxis: { stroke: theme.greyScale4, textFill: theme.greyScale3 },
              bars: { fill: theme.secondary },
              tooltips: {
                fill: '#fff',
                stroke: theme.greyScale4,
                textFill: theme.greyScale3,
              },
            }}
          />
        </span>
        </Row>}
      <Column style={{ ...styles.column }}>
        <EntityPageHorizontalTable
          headings={tableHeadings}
          data={tableData}
        />
      </Column>
    </Column>
  );
};

export default enhance(CancerDistribution);
