// @ncigdc/components/CartDownloadButton.js
// @flow

// Vendor
import React, { PropTypes } from 'react'
import DownloadIcon from 'react-icons/lib/fa/download'
import DownCaretIcon from 'react-icons/lib/fa/caret-down'

// Custom
import withDropdown from '@ncigdc/uikit/withDropdown'
import Button from '@ncigdc/uikit/Button'
import { Column, Row } from '@ncigdc/uikit/Flex'
import { dropdown } from '@ncigdc/theme/mixins'
import { withTheme } from '@ncigdc/theme'

/*----------------------------------------------------------------------------*/

const styles = {
  row: theme => ({
    color: theme.greyScale2,
    padding: '0.6rem 1rem',
    ':hover': {
      backgroundColor: theme.greyScale6,
    },
  }),
  iconSpacing: {
    marginRight: '0.6rem',
  },
}

const CartDownloadButton = ({
  active,
  setActive,
  mouseDownHandler,
  mouseUpHandler,
  style,
  theme,
}) => (
  <Button
    style={style}
    leftIcon={<DownloadIcon />}
    rightIcon={<DownCaretIcon />}
    onClick={() => setActive(true)}
  >
    Download

    {active &&
      <Column
        style={{ ...dropdown, width: '22rem' }}
        onMouseDown={mouseDownHandler}
        onMouseUp={mouseUpHandler}
      >
        <Row style={styles.row(theme)}>
          <DownloadIcon style={styles.iconSpacing} /> Manifest
        </Row>
        <Row style={styles.row(theme)}>
          <DownloadIcon style={styles.iconSpacing} /> Cart
        </Row>
      </Column>

    }
  </Button>
)

CartDownloadButton.propTypes = {
  active: PropTypes.bool,
  setActive: PropTypes.func,
  mouseDownHandler: PropTypes.func,
  mouseUpHandler: PropTypes.func,
  style: PropTypes.object,
}

/*----------------------------------------------------------------------------*/

export default withTheme(withDropdown(CartDownloadButton))
