// @ncigdc/components/CartDownloadButton.js
// @flow

// Vendor
import React, { PropTypes } from 'react';
import _ from 'lodash';

import { connect } from 'react-redux';
import { compose } from 'recompose';

import DownloadIcon from 'react-icons/lib/fa/download';
import DownCaretIcon from 'react-icons/lib/fa/caret-down';

import { setModal } from '@ncigdc/dux/modal';

// Custom
import withDropdown from '@ncigdc/uikit/withDropdown';
import Button from '@ncigdc/uikit/Button';
import { Column, Row } from '@ncigdc/uikit/Flex';
import { dropdown } from '@ncigdc/theme/mixins';
import { withTheme } from '@ncigdc/theme';

import download from '@ncigdc/utils/download';
/*----------------------------------------------------------------------------*/

const styles = {
  row: theme => ({
    color: theme.greyScale2,
    padding: '0.6rem 1rem',
    ':hover': {
      backgroundColor: theme.greyScale6,
    },
  }),
  iconSpacing: {
    marginRight: '0.6rem',
  },
};

const downloadManifests = (files) => {
  download({
    url: 'http://gdc-portal.nci.nih.gov/auth/api/manifest',
    params: {
      ids: files.map((file) => file.file_id),
    },
    method: 'POST',
  })(
    () => {},
    () => {}
  );
};

const downloadCart = (files, dispatch) => {
  const controlledFiles = files.filter((file) => file.access === 'controlled');

  if (controlledFiles && controlledFiles.length) {
    dispatch(setModal(
      <Column
        style={{ margin: '-15px' }}
      >
        <div className="modal-header">
          <h4> Access Error </h4>
        </div>
        <div className="modal-body">
          <p>You are attempting to download files that you are not authorized to access.</p>
          <p>
            <span className='label label-success'>
              {files.length - controlledFiles.length}
            </span> files that you are authorized to download.
          </p>
          <p>
            <span className='label label-danger'>
              {controlledFiles.length}
            </span>
            files that you are not authorized to download.
          </p>
          <p>
            Please <a href=''>login <i className='fa fa-sign-in' /></a>
          </p>
        </div>
        <div
          className="modal-footer"
          style={{
            display: 'flex',
            flexDirection: 'row',
            justifyContent: 'flex-end',
          }}
        >
          <Button
            onClick={() => dispatch(setModal(null))}
            style={{ margin: '0 10px' }}
          >
            <span>Cancel</span>
          </Button>
          <Button
            onClick={() => downloadCart(files.filter((file) => file.access === 'open'), dispatch)}
            style={{ margin: '0 10px' }}
          >
            <span>Download {files.length - controlledFiles.length} authorized files</span>
          </Button>
        </div>
      </Column>
    ));
  } else {
    dispatch(setModal(null));

    download({
      url: 'http://gdc-portal.nci.nih.gov/auth/api/data',
      params: {
        ids: files.map((file) => file.file_id),
      },
      method: 'POST',
    })(
      () => {},
      () => {}
    );
  }
};

const CartDownloadButton = ({
  active,
  setActive,
  mouseDownHandler,
  mouseUpHandler,
  style,
  theme,
  files,
  dispatch,
}) => (
  <Button
    style={style}
    leftIcon={<DownloadIcon />}
    rightIcon={<DownCaretIcon />}
    onClick={() => setActive(true)}
  >
    Download

    {active &&
      <Column
        style={{ ...dropdown, width: '22rem' }}
        onMouseDown={mouseDownHandler}
        onMouseUp={mouseUpHandler}
      >
        <Row
          style={styles.row(theme)}
          onClick={() => downloadManifests(files.files)}
        >
          <DownloadIcon style={styles.iconSpacing} /> Manifest
        </Row>
        <Row
          style={styles.row(theme)}
          onClick={() => downloadCart(files.files, dispatch)}
        >
          <DownloadIcon style={styles.iconSpacing} /> Cart
        </Row>
      </Column>

    }
  </Button>
);

CartDownloadButton.propTypes = {
  active: PropTypes.bool,
  setActive: PropTypes.func,
  mouseDownHandler: PropTypes.func,
  mouseUpHandler: PropTypes.func,
  style: PropTypes.object,
};

/*----------------------------------------------------------------------------*/

export default compose(
  connect(),
  withTheme,
  withDropdown
)(CartDownloadButton);
