// @ncigdc/components/DownloadButton.js
// @flow

import React from 'react';
import { connect } from 'react-redux';
import { compose, withState } from 'recompose';

import download from '@ncigdc/utils/download';
import Button from '@ncigdc/uikit/Button';

const styles = {
  buttons: {
    flex: 'none',
    marginLeft: '0.2rem',
  },
};

const DownloadButton = ({
  url,
  disabled = false,
  filename,
  dataExportExpands,
  activeText,
  inactiveText,
  returnType,
  size = 10000,
  format = 'JSON',
  fields = [],
  filters = {},
  active,
  setActive,
  altMessage = false,
  style = {},
  extraParams = {},
  SetParentActive,
  dispatch,
}) => {
  const icon = <i className={`fa fa-download ${active ? 'fa-spinner fa-pulse' : ''}`} />;
  const text = active ? activeText : inactiveText;

  return (
    <Button
      style={{ ...style, ...styles.buttons, ...((disabled || active) && styles.noHover) }}
      leftIcon={text && icon}
      disabled={disabled || active}
      onClick={() => {
        const params = {
          size,
          attachment: true,
          format,
          fields: fields.join(),
          filters,
          expand: dataExportExpands ? dataExportExpands.join() : undefined,
          pretty: true,
          ...(returnType ? { return_type: returnType } : {}),
          ...(filename ? { filename } : {}),
          ...extraParams,
        };
        SetParentActive(true);
        setActive(true);

        download({
          params,
          url,
          method: 'POST',
          altMessage,
          dispatch,
        })(
          () => {},
          () => { SetParentActive(false); setActive(false); }
        );
      }}
    >
      {text || icon}
    </Button>
  );
};

DownloadButton.propTypes = {
  url: React.PropTypes.string,
  active: React.PropTypes.bool,
  disabled: React.PropTypes.bool,
  filename: React.PropTypes.string,
  dataExportExpands: React.PropTypes.array,
  setActive: React.PropTypes.func,
  activeText: React.PropTypes.string,
  inactiveText: React.PropTypes.string,
  returnType: React.PropTypes.string,
  size: React.PropTypes.number,
  format: React.PropTypes.string,
  fields: React.PropTypes.array,
  filters: React.PropTypes.object,
  altMessage: React.PropTypes.bool,
  style: React.PropTypes.object,
  extraParams: React.PropTypes.object,
};

const enhance = compose(
  connect(),
  withState('active', 'setActive', false)
);

export default enhance(DownloadButton);
