// @ncigdc/components/GeneTable.js
// @flow
import React from 'react';
import { scaleOrdinal, schemeCategory10 } from 'd3';

import EntityPageHorizontalTable from '@ncigdc/components/EntityPageHorizontalTable';
import CohortLink from '@ncigdc/components/Links/CohortLink';
import GeneLink from '@ncigdc/components/Links/GeneLink';
import SurvivalIcon from '@ncigdc/theme/icons/SurvivalIcon';
import { tableToolTipHint } from '@ncigdc/theme/mixins';

import TogglableUl from '@ncigdc/uikit/TogglableUl';
import { Tooltip } from '@ncigdc/uikit/Tooltip';
import { getSurvivalCurves } from '@ncigdc/utils/survivalplot';
import { makeFilter } from '@ncigdc/utils/filters';
import Hidden from '@ncigdc/components/Hidden';

const colors = scaleOrdinal(schemeCategory10);

type TProps = {
  numCasesAggByProject: Object,
  projectId: string,
  totalNumCases: number,
  selectedSurvivalData: Object,
  setSelectedSurvivalData: Function,
  mutatedGenesChartData: Array<Object>,
};

const GeneTable = ({
  numCasesAggByProject = {},
  projectId = '',
  totalNumCases = 0,
  selectedSurvivalData = {},
  setSelectedSurvivalData = () => {},
  mutatedGenesChartData = [],
}: TProps) => (
  <EntityPageHorizontalTable
    headings={[
      { key: 'symbol', title: 'Symbol' },
      { key: 'name', title: 'Name' },
      { key: 'cytoband', title: 'Cytoband' },
      ...(projectId ?
      [{
        key: 'num_affected_cases_project',
        title: (
          <Tooltip
            Component={
              <span>Breakdown of Affected Cases in {projectId} <br /> # of Cases where Gene is <br />mutated /# SSM tested Cases</span>
            }
            style={tableToolTipHint()}
          >
            # Affected Cases<br />in {projectId}
          </Tooltip>),
      }] : []),
      {
        key: 'num_affected_cases_all',
        title: (
          <Tooltip
            Component={
              <span>
                # of Cases where gene contains <br />
                SSM filtered by current criteria <br />
                / # SSM tested donors portal wide
              </span>
            }
            style={tableToolTipHint()}
          >
            # Affected Cases<br /> Across all Projects
          </Tooltip>
        ),
      },
      {
        key: 'num_mutations',
        title: (
          <Tooltip
            Component='# of SSM in the gene'
            style={tableToolTipHint()}
          >
            # Mutations
          </Tooltip>
        ),
      },
      {
        title: 'Survival Analysis',
        key: 'survival_plot',
        style: { textAlign: 'center', width: '100px' },
      },
    ]}
    data={mutatedGenesChartData.map(g => ({
      ...g,
      symbol: <GeneLink id={g.gene_id}>{g.symbol}</GeneLink>,
      cytoband: (g.cytoband || []).join(', '),
      num_affected_cases_project:
        `${g.num_affected_cases_project} / ${numCasesAggByProject[projectId]}
        (${((g.num_affected_cases_project / numCasesAggByProject[projectId]) * 100).toFixed(2)}%)`,
      num_affected_cases_all: (
        <TogglableUl
          items={[
            /*  TODO: Commented out as part of PRTL-683
                Should be revert back once Exploration page is ready
            <CohortLink
              query={{
                searchTableTab: 'cases',
                filters: makeFilter([{ field: 'genes.gene_id', value: [g.gene_id] }], false),
              }}
            >
              {g.num_affected_cases_all.toLocaleString()} / {totalNumCases.toLocaleString()}
              &nbsp;({((g.num_affected_cases_all / totalNumCases) * 100).toFixed(2)}%)
            </CohortLink>, */
            <span>
              {g.num_affected_cases_all.toLocaleString()} / {totalNumCases.toLocaleString()}
              &nbsp;({((g.num_affected_cases_all / totalNumCases) * 100).toFixed(2)}%)
            </span>,
            ...Object.keys(g.num_affected_cases_by_project)
              .map(k => `
                ${k}: ${g.num_affected_cases_by_project[k]} / ${numCasesAggByProject[k]}
                (${((g.num_affected_cases_by_project[k] / numCasesAggByProject[k]) * 100).toFixed(2)}%)`),
          ]}
        />
      ),
      num_mutations: (
        /*  TODO: Commented out as part of PRTL-683
            Should be revert back once Exploration page is ready
        <CohortLink
          query={{
            searchTableTab: 'mutations',
            filters: makeFilter([{ field: 'genes.gene_id', value: [g.gene_id] }], false),
          }}
        >
          {g.case.reduce((acc, c) => acc + (c.ssm || []).length, 0).toLocaleString()}
        </CohortLink>*/
        <span>
          {g.case.reduce((acc, c) => acc + (c.ssm || []).length, 0).toLocaleString()}
        </span>
      ),
      survival_plot: (
        <Tooltip Component={`Click icon to plot ${g.symbol}`}>
          <button
            onClick={() => {
              if (g.symbol !== selectedSurvivalData.id) {
                getSurvivalCurves({
                  field: 'gene.symbol',
                  value: g.symbol,
                  currentFilters: projectId ? { op: 'and', content: [{ op: '=', content: { field: 'cases.project.project_id', value: projectId } }] } : null,
                })
                  .then(setSelectedSurvivalData);
              } else {
                setSelectedSurvivalData({});
              }
            }}
          >
            <span
              style={{
                color: colors(selectedSurvivalData.id === g.symbol ? 1 : 0),
                cursor: 'pointer',
              }}
            >
              <SurvivalIcon />
              <Hidden>add to survival plot</Hidden>
            </span>
          </button>
        </Tooltip>
      ),
    }))}
  />
);

export default GeneTable;
