// @ncigdc/components/MostAffectedCasesContainer.js
// @flow

import React from 'react';
import { withState, withProps, lifecycle, compose } from 'recompose';

import { replaceFilters } from '@ncigdc/utils/filters';
import { PaginationContainer } from '@ncigdc/uikit/Pagination';
import { fetchApi } from '@ncigdc/utils/ajax';

import MostAffectedCases from './MostAffectedCases';

export default
compose(
  withState('state', 'setState', { pagination: { total: 0 }, mostAffectedCases: [], loading: true }),
  withProps({
    fetchData: async ({ currentFilters, offset = 0, setState, projectId, first }) => {
      setState(s => ({ ...s, loading: true }));

      const filters = replaceFilters({
        op: '=',
        content: {
          field: 'project.project_id',
          value: projectId,
        },
      }, currentFilters);

      const data = await fetchApi(
        'analysis/top_mutated_cases_by_ssm',
        {
          headers: { 'Content-Type': 'application/json' },
          body: {
            from: offset,
            size: first,
            filters,
            fields: [
              'case_id',
              'gene.ssm.ssm_id',
              'summary.data_categories.data_category',
              'summary.data_categories.file_count',
              'project.primary_site',
              'demographic.gender',
              'diagnoses.days_to_last_follow_up',
            ].join(),
          },
        }
      );

      setState(s => ({ ...s, mostAffectedCases: data.data.hits, loading: false, pagination: data.data.pagination }));
    },
  }),
  lifecycle({
    componentDidMount() {
      this.props.fetchData(this.props);
    },
    componentWillReceiveProps(next) {
      if (this.props.currentFilters !== next.currentFilters) {
        this.props.fetchData(next);
      }
    },
  })
)(({ state, ...props }) => {
  return (
    <PaginationContainer
      total={state.pagination.total}
      onChange={pageInfo => props.fetchData({ ...props, ...pageInfo })}
      entityType="Cases"
      loading={state.loading}
    >
      <MostAffectedCases mostAffectedCases={state.mostAffectedCases} />
    </PaginationContainer>
  );
});
