// @ncigdc/components/PortalSummary.js
// @flow

import React from 'react';

import { zDepth1 } from '@ncigdc/theme/mixins';
import DoubleHelix from '@ncigdc/theme/icons/DoubleHelix';
import MutationIcon from '@ncigdc/theme/icons/Mutation';

import ProjectsCount from '@ncigdc/components/ProjectsCount';
import PrimarySitesCount from '@ncigdc/components/PrimarySitesCount';
import CasesCount from '@ncigdc/components/CasesCount';
import FilesCount from '@ncigdc/components/FilesCount';

import { Row, Column } from '@ncigdc/uikit/Flex';

const styles = {
  summaryBoxContainer: {
    ...zDepth1,
    marginTop: '2rem',
    backgroundColor: 'white',
    borderTop: '3px solid rgb(37, 208, 182)',
  },
  dataCountBox: {
    padding: '1.5rem',
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  dataCountBoxTitle: {
    justifyContent: 'center',
    textTransform: 'uppercase',
    marginBottom: '0.5rem',
  },
};

type TProps = {
  dataRelease: string,
  projectsCountData: Object,
  primarySitesCountData: Object,
  casesCountData: Object,
  filesCountData: Object,
  genesCountData: Object,
  ssmsCountData: Object,
};
const PortalSummary = (props: TProps) => (
  <Column style={styles.summaryBoxContainer}>
    <Row style={{ padding: '2rem', alignItems: 'baseline' }}>
      <div style={{ fontSize: '2.3rem', color: 'rgb(70, 70, 70)' }}>Data Portal Summary</div>
      <div style={{ fontSize: '1.3rem', color: 'rgb(37, 97, 122)', marginLeft: '2rem' }}>
        <a
          target="_blank"
          rel="noopener noreferrer"
          href="https://gdc-docs.nci.nih.gov/Data_Portal/Release_Notes/Data_Portal_Release_Notes/"
        >
          {props.dataRelease}
        </a>
      </div>
    </Row>
    <Column>
      <Row>
        <Column style={styles.dataCountBox}>
          <Row style={styles.dataCountBoxTitle}>Projects</Row>
          <Row>
            <i style={{ color: '#01b987', fontSize: '3rem' }} className="icon-gdc-projects project-icon" />
            <span style={{ fontSize: '2.5rem', marginLeft: '0.5rem' }}>
              <ProjectsCount hits={props.projectsCountData} />
            </span>
          </Row>
        </Column>
        <Column style={styles.dataCountBox}>
          <Row style={{ textTransform: 'uppercase' }}>Primary Sites</Row>
          <Row>
            <i style={{ color: '#01b987', fontSize: '3rem' }} className="icon-gdc-cases data-icon" />
            <span style={{ fontSize: '2.5rem', marginLeft: '0.5rem' }}>
              <PrimarySitesCount aggregations={props.primarySitesCountData} />
            </span>
          </Row>
        </Column>
        <Column style={styles.dataCountBox}>
          <Row style={{ textTransform: 'uppercase' }}>Cases</Row>
          <Row>
            <i style={{ color: '#01b987', fontSize: '3rem' }} className="icon-gdc-cases data-icon" />
            <span style={{ fontSize: '2.5rem', marginLeft: '0.5rem' }}>
              <CasesCount hits={props.casesCountData} />
            </span>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column style={styles.dataCountBox}>
          <Row style={{ textTransform: 'uppercase' }}>Files</Row>
          <Row>
            <i style={{ color: '#01b987', fontSize: '3rem' }} className="fa fa-file-o data-icon" />
            <span style={{ fontSize: '2.5rem', marginLeft: '0.5rem' }}>
              <FilesCount hits={props.filesCountData} />
            </span>
          </Row>
        </Column>
        <Column style={styles.dataCountBox}>
          <Row style={{ textTransform: 'uppercase' }}>Genes</Row>
          <Row>
            <DoubleHelix
              color="#01b987"
              width="20px"
            />
            <span style={{ fontSize: '2.5rem', marginLeft: '0.5rem' }}>
              <CasesCount hits={props.genesCountData} />
            </span>
          </Row>
        </Column>
        <Column style={styles.dataCountBox}>
          <Row style={{ textTransform: 'uppercase' }}>Mutations</Row>
          <Row>
            <MutationIcon
              color="#01b987"
              width="32px"
            />
            <span style={{ fontSize: '2.5rem', marginLeft: '0.5rem' }}>
              <FilesCount hits={props.ssmsCountData} />
            </span>
          </Row>
        </Column>
      </Row>
    </Column>
  </Column>
);

export default PortalSummary;
