// @ncigdc/components/Project.js
// @flow
import React from 'react';
import { scaleOrdinal, schemeCategory20 } from 'd3';

import { makeFilter } from '@ncigdc/utils/filters';

import { Row, Column } from '@ncigdc/uikit/Flex';
import { Tooltip } from '@ncigdc/uikit/Tooltip';

import EntityPageVerticalTable from '@ncigdc/components/EntityPageVerticalTable';
import CountCard from '@ncigdc/components/CountCard';
import DownloadButton from '@ncigdc/components/DownloadButton';
import SummaryCard from '@ncigdc/components/SummaryCard';
import ProjectVisualizations from '@ncigdc/components/ProjectVisualizations';

import FileIcon from '@ncigdc/theme/icons/File';
import CaseIcon from '@ncigdc/theme/icons/Case';
import EditIcon from '@ncigdc/theme/icons/Edit';

const colors20 = scaleOrdinal(schemeCategory20);

const CLINICAL_EXPORT_EXPANDS = ['demographic', 'diagnoses', 'family_histories', 'exposures'];
const BIOSPECIMEN_EXPORT_EXPANDS = [
  'samples',
  'samples.portions',
  'samples.portions.analytes',
  'samples.portions.analytes.aliquots',
  'samples.portions.analytes.aliquots.annotations',
  'samples.portions.analytes.annotations',
  'samples.portions.submitter_id',
  'samples.portions.slides',
  'samples.portions.annotations',
  'samples.portions.center',
];

const SPACING = '2rem';

const styles = {
  countCard: {
    width: 'auto',
    marginBottom: SPACING,
  },
  column: {
    flexGrow: 1,
  },
  margin: {
    marginBottom: SPACING,
  },
  icon: {
    width: '4rem',
    height: '4rem',
    color: '#888',
  },
  coloredSquare: {
    display: 'inline-block',
    width: 10,
    height: 10,
    marginRight: 5,
  },
};

const Project = ({
  projectId,
  fileCount,
  projectName,
  programName,
  caseCount,
  diseaseType,
  primarySite,
  experimentalStrategies,
  dataCategories,
  totalAnnotations,
  annotations,
  clinicalCount,
  biospecimenCount,
  api = API,
}: {
  projectId: string,
  fileCount: number,
  projectName: string,
  programName: string,
  caseCount: number,
  diseaseType: Array<string>,
  primarySite: Array<string>,
  experimentalStrategies: Array<Object>,
  dataCategories: Array<Object>,
  totalAnnotations: number,
  annotations: Array<{annotation_id: string}>,
  clinicalCount: number,
  biospecimenCount: number,
  api?: string,
}) => {
  const projectFilter = {
    field: 'project.project_id',
    value: projectId,
  };

  const dataExportFilters = makeFilter([projectFilter], false);

  const experimentalStrategiesFilters = (value) => makeFilter([
    projectFilter,
    { field: 'files.experimental_strategy', value: [value] },
  ], true);

  const dataCategoriesFilters = (value) => makeFilter([
    projectFilter,
    { field: 'files.data_category', value: [value] },
  ], true);

  return (
    <span>
      <Row style={{ ...styles.margin, flexDirection: 'row-reverse' }}>
        <DownloadButton
          disabled={!biospecimenCount}
          filename={`biospecimen.project-${projectId}`}
          dataExportExpands={BIOSPECIMEN_EXPORT_EXPANDS}
          url={`${api}/cases`}
          activeText="Processing"
          inactiveText={biospecimenCount ? 'Download Biospecimen' : 'No Biospecimen Data'}
          fields={['case_id']}
          filters={dataExportFilters}
        />

        <DownloadButton
          disabled={!clinicalCount}
          filename={`clinical.project-${projectId}`}
          dataExportExpands={CLINICAL_EXPORT_EXPANDS}
          url={`${api}/cases`}
          activeText="Processing"
          inactiveText={clinicalCount ? 'Download Clinical' : 'No Clinical Data'}
          fields={['case_id']}
          filters={dataExportFilters}
        />

        <Tooltip
          dir="down"
          Component={`Download a manifest for use with the GDC Data Transfer Tool.
            The GDC Data Transfer Tool is recommended for transferring large volumes of data.`}
          maxWidth="250px"
        >
          <DownloadButton
            disabled={!fileCount}
            url={`${api}/files`}
            activeText="Downloading"
            inactiveText="Download manifest"
            fields={['file_id']}
            size={fileCount}
            returnType="manifest"
            format="TSV"
            filters={
              makeFilter([{ field: 'cases.project.project_id', value: projectId }], false)
            }
          />
        </Tooltip>
      </Row>

      <Row style={{ flexWrap: 'wrap' }} spacing={SPACING}>
        <span style={{ ...styles.column, ...styles.margin }}>
          <EntityPageVerticalTable
            id="summary"
            title={<span><i className="fa fa-table" /> Summary</span>}
            thToTd={[
              { th: 'Project ID', td: projectId },
              { th: 'Project Name', td: projectName },
              { th: 'Disease Type', td: diseaseType.map(p => <div key={p}>{p}</div>) },
              { th: 'Primary Site', td: primarySite.map(p => <div key={p}>{p}</div>) },
              { th: 'Program', td: programName },
            ]}
          />
        </span>

        <Column style={{ ...styles.margin, width: '200px' }}>
          <CountCard
            title="CASES"
            count={caseCount.toLocaleString()}
            icon={<CaseIcon style={styles.icon} className="fa-3x" />}
            style={styles.countCard}
            onCountClick={() => {
              window.location = `/search/c?filters=${
                makeFilter([{ field: 'cases.project.project_id', value: projectId }])
              }`;
            }}
          />

          <CountCard
            title="FILES"
            count={fileCount.toLocaleString()}
            icon={<FileIcon style={styles.icon} className="fa-3x" />}
            style={styles.countCard}
            onCountClick={() => {
              window.location = `/search/f?filters=${
                makeFilter([{ field: 'cases.project.project_id', value: projectId }])
              }`;
            }}
          />

          <CountCard
            title="ANNOTATIONS"
            count={(totalAnnotations).toLocaleString()}
            icon={<EditIcon style={styles.icon} className="fa-3x" />}
            style={{ ...styles.countCard, marginBottom: 0 }}
            {
              ...(totalAnnotations > 0 ? { onCountClick: () => {
                if (totalAnnotations > 1 || !annotations) {
                  window.location = `/annotations?filters=${
                      makeFilter([{ field: 'project.project_id', value: projectId }])
                  }`;
                } else {
                  window.location = `/annotations/${annotations[0].annotation_id}`;
                }
              } } : {})
            }
          />
        </Column>
      </Row>

      <Row style={{ flexWrap: 'wrap' }} spacing={SPACING}>
        <span style={{ ...styles.column, ...styles.margin, flex: 1 }}>
          <SummaryCard
            tableTitle="Cases and File Counts by Experimental Strategy"
            pieChartTitle="File Counts by Experimental Strategy"
            data={
              experimentalStrategies.map((item, i) => {
                const builtFilters = experimentalStrategiesFilters(item.experimental_strategy);

                return {
                  id: item.experimental_strategy,
                  experimental_strategy: (
                    <span>
                      <div style={{ ...styles.coloredSquare, backgroundColor: colors20(i) }} />
                      {item.experimental_strategy}
                    </span>
                  ),
                  case_count: <a href={`/search/c?filters=${builtFilters}`}>{item.case_count.toLocaleString()}</a>,
                  file_count: <a href={`/search/f?filters=${builtFilters}`}>{item.file_count.toLocaleString()}</a>,
                  file_count_value: item.file_count,
                  tooltip: (
                    <span>
                      <b>{item.experimental_strategy}</b><br />
                      {item.file_count} file{item.file_count > 1 ? 's' : 0}
                    </span>
                  ),
                };
              })
            }
            footer={`${experimentalStrategies.length} Experimental Strategies`}
            path="file_count_value"
            headings={[
              { key: 'experimental_strategy', title: 'Experimental Strategy', color: true },
              {
                key: 'case_count',
                title: 'Cases',
                style: { textAlign: 'right' },
              },
              {
                key: 'file_count',
                title: 'Files',
                style: { textAlign: 'right' },
              },
            ]}
          />
        </span>
        <span style={{ ...styles.column, ...styles.margin, flex: 1 }}>
          <SummaryCard
            tableTitle="Cases and File Counts by Data Category"
            pieChartTitle="File Counts by Data Category"
            data={
              dataCategories.map((item, i) => {
                const builtFilters = dataCategoriesFilters(item.data_category);

                return {
                  id: item.data_category,
                  data_category: (
                    <span>
                      <div style={{ ...styles.coloredSquare, backgroundColor: colors20(i) }} />
                      {item.data_category}
                    </span>
                  ),
                  case_count: <a href={`/search/c?filters=${builtFilters}`}>{item.case_count.toLocaleString()}</a>,
                  file_count: <a href={`/search/f?filters=${builtFilters}`}>{item.file_count.toLocaleString()}</a>,
                  file_count_value: item.file_count,
                  tooltip: (
                    <span>
                      <b>{item.data_category}</b><br />
                      {item.file_count} file{item.file_count > 1 ? 's' : 0}
                    </span>
                  ),
                };
              })
            }
            footer={`${dataCategories.length} Data Categories`}
            path="file_count_value"
            headings={[
              { key: 'data_category', title: 'Data Category', color: true },
              {
                key: 'case_count',
                title: 'Cases',
                style: { textAlign: 'right' },
              },
              {
                key: 'file_count',
                title: 'Files',
                style: { textAlign: 'right' },
              },
            ]}
          />
        </span>
      </Row>
      <ProjectVisualizations
        projectId={projectId}
        api={api}
      />
    </span>
  );
};

export default Project;
