// @ncigdc/components/ProteinLolliplot.js
// @flow

import React from 'react';
import moment from 'moment';
import { insertRule } from 'glamor';

import { Row, Column } from '@ncigdc/uikit/Flex';
import Dropdown from '@ncigdc/uikit/Dropdown';
import DropdownItem from '@ncigdc/uikit/DropdownItem';
import Button from '@ncigdc/uikit/Button';

import { visualizingButton } from '@ncigdc/theme/mixins';
import DoubleHelix from '@ncigdc/theme/icons/DoubleHelix';
import SpinnerParticle from '@ncigdc/uikit/Loaders/Particle';
import DownloadVisualizationButton from './DownloadVisualizationButton';

const rootId = 'protein-viewer-root';

insertRule(`
  #${rootId} text { user-select: none; }
`);

const styles = {
  heading: {
    flexGrow: 1,
    fontSize: '2.2rem',
    marginBottom: 7,
    marginTop: 7,
    display: 'flex',
    alignItems: 'center',
  },
};

const ProteinLolliplot = ({
  gene = {},
  activeTranscript = {},
  transcripts = [],
  proteinLolliplotData = {},
  reset = () => {},
  selectTranscript = () => {},
  loading = true,
}) => (
  <Column>
    <Row>
      <h1 style={{ ...styles.heading, padding: '1rem' }} id="protein">
        <DoubleHelix width="12px" />
        <span style={{ marginLeft: '1rem' }}>Protein</span>
      </h1>
    </Row>
    <Row style={{ marginBottom: '2rem', padding: '0 2rem' }} spacing="1rem">
      <span style={{ alignSelf: 'center' }}>
        Transcript:
      </span>
      <Dropdown
        selected={
          <span
            style={{
              fontWeight: activeTranscript.transcript_id === gene.canonical_transcript_id
                ? 'bold' : 'initial',
            }}
          >
            {activeTranscript.transcript_id} ({activeTranscript.length_amino_acid} aa)
          </span>
        }
      >
        {transcripts
        .filter(t => t.transcript_id === gene.canonical_transcript_id)
        .map(t =>
          <DropdownItem
            key={t.transcript_id}
            style={{
              fontWeight: 'bold',
              ...(activeTranscript.transcript_id === t.transcript_id && {
                backgroundColor: 'rgb(44, 136, 170)',
                color: 'white',
              }),
            }}
            onClick={() => selectTranscript(t)}
          >
            {t.transcript_id} ({t.length_amino_acid} aa)
          </DropdownItem>
        )}
        {transcripts
        .filter(t => t.length_amino_acid && t.transcript_id !== gene.canonical_transcript_id)
        .map(t =>
          <DropdownItem
            key={t.transcript_id}
            style={{
              ...(activeTranscript.transcript_id === t.transcript_id && {
                backgroundColor: 'rgb(44, 136, 170)',
                color: 'white',
              }),
            }}
            onClick={() => selectTranscript(t)}
          >
            {t.transcript_id} ({t.length_amino_acid} aa)
          </DropdownItem>
        )}
      </Dropdown>
      <Button
        style={visualizingButton}
        onClick={reset}
        leftIcon={<i className="fa fa-refresh" />}
      >
        Reset
      </Button>
      <DownloadVisualizationButton
        svg="#protein-viewer-root svg.chart"
        data={proteinLolliplotData}
        stylePrefix="#protein-viewer-root"
        slug={`protein_viewer-${gene.symbol}-${moment().format('YYYY-MM-DD')}`}
      />
    </Row>
    {loading &&
      <Column style={{ alignItems: 'center', padding: '20px' }}><SpinnerParticle /></Column>
    }
    <div style={{ padding: '0 3rem' }} id={rootId} />
  </Column>
);

export default ProteinLolliplot;
