// @ncigdc/components/SearchPage.js
// @flow

// Vendor
import React, { PropTypes } from 'react';
import LeftArrow from 'react-icons/lib/fa/long-arrow-left';
import ShoppingCartIcon from 'react-icons/lib/fa/shopping-cart';

// Custom
import { Row, Column } from '@ncigdc/uikit/Flex';
import Button from '@ncigdc/uikit/Button';
import Info from '@ncigdc/uikit/Info';
import theme from '@ncigdc/theme';
import CurrentFilters from './CurrentFilters';

/*----------------------------------------------------------------------------*/

const styles = {
  container: {
    padding: '2rem 2.5rem 13rem',
  },
  facetsPanel: {
    width: theme.facetsPanelWidth,
  },
  facet: {
    backgroundColor: 'white',
  },
  content: {
    marginLeft: '18px',
    width: `calc(100% - ${theme.facetsPanelWidth})`,
  },
};

const SearchPage = ({
  Results,
  Facets,
}) => (
  <Row style={styles.container}>
    <Column style={styles.facetsPanel}>
      {Facets}
    </Column>

    <Column style={styles.content}>
      {/* {!location.query.filters &&
        <Info>
          <LeftArrow />
          <span style={{ marginLeft: '0.6rem' }}>
            Start searching by selecting a facet
          </span>
        </Info>
      } */}

      {/* <CurrentFilters /> */}

      <Info>
        <Button leftIcon={<ShoppingCartIcon />}>Add all files to the cart</Button>
      </Info>

      {Results}
    </Column>
  </Row>
);

SearchPage.propTypes = {
  Results: PropTypes.node,
  Facets: PropTypes.node,
};

/*----------------------------------------------------------------------------*/

export default SearchPage;
