// @ncigdc/components/SearchPage.js
// @flow

// Vendor
import React, { PropTypes } from 'react';

import { Row, Column } from '@ncigdc/uikit/Flex';
import Button from '@ncigdc/uikit/Button';
import Info from '@ncigdc/uikit/Info';

import { withTheme } from '@ncigdc/theme';

import CurrentFilters from './CurrentFilters';

/*----------------------------------------------------------------------------*/

const styles = theme => ({
  container: {
    padding: '2rem 2.5rem 13rem',
  },
  facetsPanel: {
    width: theme.facetsPanelWidth,
  },
  facet: {
    backgroundColor: 'white',
  },
  content: {
    marginLeft: '18px',
    width: `calc(100% - ${theme.facetsPanelWidth})`,
  },
});

const SearchPage = ({
  Results,
  Facets,
  theme,
  charts = <span />
}) => (
  <Row style={styles(theme).container}>
    <Column style={styles(theme).facetsPanel}>
      {Facets}
    </Column>

    <Column style={styles(theme).content}>
      {charts}
      <CurrentFilters />
      {Results}
    </Column>
  </Row>
);

SearchPage.propTypes = {
  Results: PropTypes.node,
  Facets: PropTypes.node,
};

/*----------------------------------------------------------------------------*/

export default withTheme(SearchPage);
