// @ncigdc/components/SearchPage.js
// @flow

import React from 'react';
import { Row, Column } from '@ncigdc/uikit/Flex';
import styled from '@ncigdc/theme/styled';
import { DoubleArrowRightIcon } from '@ncigdc/theme/icons';
import CurrentFilters from '@ncigdc/components/CurrentFilters';

const sidePadding = '2.5rem';

const Container = styled(Row, {
  padding: `2rem ${sidePadding} 13rem`,
});

const FacetsPanel = styled(Column, {
  width: ({ theme }) => theme.facetsPanelWidth,
  flex: 'none',
  marginRight: '18px',
});

const Content = styled(Column, {
  flex: 1,
  width: 0,
});

const ShowFacetsButton = styled.button({
  flex: 'none',
  padding: 10,
  backgroundColor: ({ theme }) => theme.white,
  border: ({ theme }) => `1px solid ${theme.greyScale4}`,
  borderLeft: 'none',
  margin: `2.1rem 4rem auto -${sidePadding}`,
  outline: 'none',
  borderRadius: '0 0.4rem 0.4rem 0',
});

type TProps = {|
  charts?: mixed,
  facets?: mixed,
  results?: mixed,
  showFacets?: boolean,
  toggleShowFacets?: Function,
|};

const SearchPage = ({
  charts = <span />,
  facets = <span />,
  results = <span />,
  showFacets = true,
  toggleShowFacets,
}: TProps = {}) => (
  <Container>
    {showFacets && <FacetsPanel>{facets}</FacetsPanel>}
    <Content>
      <Row style={{ marginBottom: '2rem' }}>
        {
          !showFacets && toggleShowFacets && (
            <ShowFacetsButton onClick={toggleShowFacets}>
              <span>Show Filters <DoubleArrowRightIcon /></span>
            </ShowFacetsButton>
          )
        }
        <CurrentFilters style={{ flex: 1 }} />
      </Row>
      {charts}
      {results}
    </Content>
  </Container>
);

export default SearchPage;
