// @ncigdc/components/SortTableButton.js
// @flow

// Vendor
import React from 'react';
import { getContext, compose } from 'recompose';
import ArrowDownIcon from 'react-icons/lib/fa/long-arrow-down';
import ArrowUpIcon from 'react-icons/lib/fa/long-arrow-up';

// Custom
import { SortIcon } from '@ncigdc/theme/icons';
import { Row } from '@ncigdc/uikit/Flex';
import Button from '@ncigdc/uikit/Button';
import Dropdown from '@ncigdc/uikit/Dropdown';
import { Tooltip } from '@ncigdc/uikit/Tooltip';
import { withTheme } from '@ncigdc/theme';
import Link from '@ncigdc/components/Links/Link';

/*----------------------------------------------------------------------------*/

const styles = {
  checkboxRow: (theme) => ({
    padding: '0.3rem 0.6rem',
    ':hover': {
      backgroundColor: theme.greyScale6,
      borderRight: `2px solid ${theme.secondary}`,
    },
  }),
  radioRow: {
    padding: '0.3rem 0.6rem',
    marginLeft: 'auto',
    alignItems: 'center',
  },
};

const enhance = compose(
  getContext({ router: React.PropTypes.object }),
  withTheme
);

const SortTableButton = ({
  theme,
  style,
  options,
  query: q,
  sortKey,
}: {
  theme: Object,
  style: Object,
  options: Array<Object>,
  query: Object,
  sortKey: string,
}) => {
  const { [sortKey]: sort = '', ...query } = q;
  const fields = sort.split(',').filter(Boolean);

  return (
    <Dropdown
      button={
        <Tooltip Component={<span>Sort Table</span>} >
          <Button style={style} ><SortIcon /></Button>
        </Tooltip>
      }
      dropdownStyle={{ top: '100%', marginTop: 5, whiteSpace: 'nowrap' }}
    >
      {options.map(x => {
        const sameField = fields.find(f => f.split(':')[0] === x.id);
        const otherFields = fields.filter(f => f.split(':')[0] !== x.id);

        const nextSort = sameField
          ? otherFields.join(',')
          : [...fields, `${x.id}:asc`].join(',');

        return (
          <Row key={x.id} style={{ lineHeight: '1.5' }}>
            <Row style={styles.checkboxRow(theme)} flex="1">
              <Link
                style={{ width: '100%' }}
                query={{
                  ...query,
                  ...(nextSort ? { [sortKey]: nextSort } : {}),
                }}
              >
                <input
                  readOnly
                  style={{ pointerEvents: 'none' }}
                  type="checkbox"
                  checked={!!sameField}
                />
                <span style={{ marginLeft: '0.3rem' }}>{x.name}</span>
              </Link>
            </Row>
            <Row style={styles.radioRow}>
              <Link
                style={{ width: '100%' }}
                query={{
                  ...query,
                  [sortKey]: [...otherFields, `${x.id}:asc`].join(','),
                }}
              >
                <ArrowDownIcon />
                <input
                  readOnly
                  type="radio"
                  checked={!!sameField && sameField.split(':')[1] === 'asc'}
                />
              </Link>
              <Link
                style={{ width: '100%' }}
                query={{
                  ...query,
                  [sortKey]: [...otherFields, `${x.id}:desc`].join(','),
                }}
              >
                <ArrowUpIcon />
                <input
                  readOnly
                  type="radio"
                  checked={!!sameField && sameField.split(':')[1] === 'desc'}
                />
              </Link>
            </Row>
          </Row>
        );
      })}
    </Dropdown>
  );
};

/*----------------------------------------------------------------------------*/

export default enhance(SortTableButton);
