// @ncigdc/components/SortTableButton.js
// @flow

import React from 'react';
import { compose } from 'recompose';
import ArrowDownIcon from 'react-icons/lib/fa/long-arrow-down';
import ArrowUpIcon from 'react-icons/lib/fa/long-arrow-up';

import withRouter from '@ncigdc/utils/withRouter';
import { SortIcon } from '@ncigdc/theme/icons';
import { Row } from '@ncigdc/uikit/Flex';
import Button from '@ncigdc/uikit/Button';
import Dropdown from '@ncigdc/uikit/Dropdown';
import { Tooltip } from '@ncigdc/uikit/Tooltip';
import styled from '@ncigdc/theme/styled';
import Link from '@ncigdc/components/Links/Link';
import Hidden from '@ncigdc/components/Hidden';

/*----------------------------------------------------------------------------*/

const CheckboxRow = styled(Row, {
  borderRight: '2px solid transparent',
  padding: '0.3rem 0.6rem',
  ':hover': {
    backgroundColor: ({ theme }) => theme.greyScale6,
    borderRight: ({ theme }) => `2px solid ${theme.secondary}`,
  },
});

const RadioRow = styled(Row, {
  padding: '0.3rem 0.6rem',
  marginLeft: 'auto',
  alignItems: 'center',
});

type TSortTableButtonProps = {
  style: Object,
  options: Array<Object>,
  query: Object,
  sortKey: string,
};

const SortTableButton = compose(
  withRouter
)(({
  style,
  options,
  query: q,
  sortKey,
}: TSortTableButtonProps = {}) => {
  const { [sortKey]: sort = '', ...query } = q;
  const fields = sort.split(',').filter(Boolean);

  return (
    <Dropdown
      button={
        <Tooltip Component={<span>Sort Table</span>} >
          <Button style={style} ><SortIcon /><Hidden>Sort Table</Hidden></Button>
        </Tooltip>
      }
      dropdownStyle={{ top: '100%', marginTop: 5, whiteSpace: 'nowrap' }}
    >
      {options.map(x => {
        const sameField = fields.find(f => f.split(':')[0] === x.id);
        const otherFields = fields.filter(f => f.split(':')[0] !== x.id);

        const nextSort = sameField
          ? otherFields.join(',')
          : [...fields, `${x.id}:asc`].join(',');

        return (
          <Row key={x.id} style={{ lineHeight: '1.5' }}>
            <CheckboxRow flex="1">
              <Link
                style={{ width: '100%' }}
                query={{
                  ...query,
                  ...(nextSort ? { [sortKey]: nextSort } : {}),
                }}
              >
                <input
                  readOnly
                  style={{ pointerEvents: 'none' }}
                  type="checkbox"
                  checked={!!sameField}
                />
                <span style={{ marginLeft: '0.3rem' }}>{x.name}</span>
              </Link>
            </CheckboxRow>
            <RadioRow>
              <Link
                style={{ width: '100%' }}
                query={{
                  ...query,
                  [sortKey]: [...otherFields, `${x.id}:asc`].join(','),
                }}
              >
                <ArrowDownIcon />
                <input
                  readOnly
                  type="radio"
                  checked={!!sameField && sameField.split(':')[1] === 'asc'}
                />

              </Link>
              <Link
                style={{ width: '100%' }}
                query={{
                  ...query,
                  [sortKey]: [...otherFields, `${x.id}:desc`].join(','),
                }}
              >
                <ArrowUpIcon />
                <input
                  readOnly
                  type="radio"
                  checked={!!sameField && sameField.split(':')[1] === 'desc'}
                />
              </Link>
            </RadioRow>
          </Row>
        );
      })}
    </Dropdown>
  );
});

/*----------------------------------------------------------------------------*/

export default SortTableButton;
