// @ncigdc/components/SummaryCard.js
// @flow

// Vendor
import React, { PropTypes } from 'react';
import { compose, withState } from 'recompose';

// Custom
import PieChart from '@ncigdc/components/Charts/PieChart';
import { Row, Column } from '@ncigdc/uikit/Flex';
import Card from '@ncigdc/uikit/Card';
import { withTheme } from '@ncigdc/theme';
import { center } from '@ncigdc/theme/mixins';
import EntityPageHorizontalTable from './EntityPageHorizontalTable';

/*----------------------------------------------------------------------------*/

const styles = {
  header: theme => ({
    padding: '1rem',
    color: theme.greyScale7,
  }),
  footer: {
    padding: '1rem',
    ...center,
  },
};

const SummaryCard = ({
  data,
  title,
  pieChartTitle,
  tableTitle,
  style = {},
  footer,
  path,
  showTable,
  setShowTable,
  headings,
  theme,
}) => (
  <Card style={style}>
    <Column>
      <Row style={styles.header(theme)}>
        <span style={{ flexGrow: 1, fontSize: '1.7rem' }}>
          {(showTable ? tableTitle : pieChartTitle) || title}
        </span>
        <span onClick={() => setShowTable(!showTable)}>
          <i
            style={{ color: '#000' }}
            className={`fa ${showTable ? 'fa-pie-chart' : 'fa-table'}`}
          />
        </span>
      </Row>

      { showTable &&
        <EntityPageHorizontalTable
          headings={headings}
          data={data}
          style={{ overflow: 'hidden', borderLeft: 0, borderTop: 0 }}
        />
      }
      {!showTable &&
        [
          <PieChart key="chart" data={data} path={path} height={250} width={250} />,
          footer && <Row key={footer} style={styles.footer}>{footer}</Row>,
        ]
      }
    </Column>
  </Card>
);

SummaryCard.propTypes = {
  style: PropTypes.object,
  title: PropTypes.string,
  data: PropTypes.array,
  footer: PropTypes.node,
  path: PropTypes.string,
  table: PropTypes.object,
};

/*----------------------------------------------------------------------------*/

const enhance = compose(
  withState('showTable', 'setShowTable', true),
  withTheme
);

export default enhance(SummaryCard);
