// @ncigdc/components/TableActions.js
/* @flow */

import React from 'react';

import { Row } from '@ncigdc/uikit/Flex';
import { Tooltip } from '@ncigdc/uikit/Tooltip';
import { parseFilterParam } from '@ncigdc/utils/uri';
import DownloadButton from '@ncigdc/components/DownloadButton';
import SortTableButton from '@ncigdc/components/SortTableButton';
import LocationSubscriber from '@ncigdc/components/LocationSubscriber';

import type { TRawQuery } from '@ncigdc/utils/uri/types';

const TableActions = ({
  prefix,
  total,
  sortOptions,
  sortKey,
  endpoint,
  downloadFields,
  downloadable = true,
}: {
  prefix: string,
  total: number,
  sortOptions: Array<Object>,
  sortKey: string,
  endpoint: string,
  downloadFields: Array<string>,
  downloadable?: boolean,
}) => (
  <LocationSubscriber>
    {
      ({ query }: {| query: TRawQuery |}) => (
        <Row>
          <Tooltip Component={<span>Sort Table</span>} >
            <SortTableButton
              options={sortOptions}
              query={query || {}}
              sortKey={sortKey}
            />
          </Tooltip>
          {downloadable &&
            (
              <Tooltip Component={<span>Export Table</span>} >
                <DownloadButton
                  filters={parseFilterParam((query || {}).filters, {})}
                  disabled={!total}
                  filename={prefix}
                  url={endpoint}
                  fields={downloadFields}
                />
              </Tooltip>
            )
          }
        </Row>
      )
    }
  </LocationSubscriber>
);
export default TableActions;
