// @ncigdc/routes/CohortRoute.js
/* @flow */
/* eslint fp/no-class:0 */

import React from 'react';
import Relay from 'react-relay';
import { connect } from 'react-redux';
import { parse } from 'query-string';

import { handleStateChange } from '@ncigdc/dux/relayProgress';
import CohortPage from '@ncigdc/containers/cohort/CohortPage';
import { parseIntParam, parseFilterParam } from '@ncigdc/utils/uri';

import { viewerQuery } from './queries';

class CohortRoute extends Relay.Route {
  static routeName = 'CohortRoute';

  static queries = {
    ...viewerQuery,
    //autocomplete: () => Relay.QL`query { cohort }`,
  };

  static prepareParams = ({ location: { search } }) => {
    const q = parse(search);

    return ({
      cases_offset: parseIntParam(q.cases_offset, 0),
      cases_size: parseIntParam(q.cases_size, 20),
      cases_sort: parseFilterParam(q.cases_sort, null),
      files_offset: parseIntParam(q.files_offset, 0),
      files_size: parseIntParam(q.files_size, 20),
      files_sort: parseFilterParam(q.files_sort, null),
      filters: parseFilterParam(q.filters, null),
    });
  }
}

export default connect()((props: mixed) => (
  <Relay.Renderer
    Container={CohortPage}
    queryConfig={new CohortRoute(props)}
    environment={Relay.Store}
    onReadyStateChange={handleStateChange(props)}
  />
));
