// @ncigdc/routes/RepositoryRoute.js
/* @flow */
/* eslint fp/no-class:0 */

import React from 'react';
import Relay from 'react-relay';
import { connect } from 'react-redux';
import { parse } from 'query-string';

import { handleStateChange } from '@ncigdc/dux/relayProgress';
import RepositoryPage from '@ncigdc/containers/RepositoryPage';
import { parseIntParam, parseFilterParam } from '@ncigdc/utils/uri';

import { viewerQuery } from './queries';

class RepositoryRoute extends Relay.Route {
  static routeName = 'RepositoryRoute';
  static queries = viewerQuery;
  static prepareParams = ({ location: { search } }) => {
    const q = parse(search);
    return ({
      cases_offset: parseIntParam(q.cases_offset, 0),
      cases_size: parseIntParam(q.cases_size, 20),
      cases_sort: q.cases_sort || '',
      files_offset: parseIntParam(q.files_offset, 0),
      files_size: parseIntParam(q.files_size, 20),
      files_sort: q.files_sort || '',
      filters: parseFilterParam(q.filters, null),
    });
  }
}

export default connect()((props: mixed) => (
  <Relay.Renderer
    Container={RepositoryPage}
    queryConfig={new RepositoryRoute(props)}
    environment={Relay.Store}
    onReadyStateChange={handleStateChange(props)}
  />
));
