// @ncigdc/routes/SearchRoute.js
/* @flow */
/* eslint fp/no-class:0 */

import React from 'react';
import Relay from 'react-relay';

import SearchPage from '@ncigdc/containers/SearchPage';
import { parseIntParam, parseFilterParam } from '@ncigdc/utils/uri';
import { viewerQuery } from './queries';

class SearchRoute extends Relay.Route {
  static routeName = 'SearchRoute';
  static queries = viewerQuery;
  static prepareParams = ({ location: { query } }) => {
    console.log(1, query);
    const q = query || {};
    return ({
      cases_offset: parseIntParam(q.cases_offset, 0),
      cases_size: parseIntParam(q.cases_size, 20),
      cases_sort: q.cases_sort || '',
      files_offset: parseIntParam(q.files_offset, 0),
      files_size: parseIntParam(q.files_size, 20),
      files_sort: q.files_sort || '',
      filters: parseFilterParam(q.filters, null),
    });
  }
}

export default (props: mixed) => (
  <Relay.Renderer
    Container={SearchPage}
    queryConfig={new SearchRoute(props)}
    environment={Relay.Store}
  />
);
