// @ncigdc/routes/SmartSearchRoute.js
// @flow
/* eslint fp/no-class:0 */

import React from 'react';
import Relay from 'react-relay';
import { connect } from 'react-redux';
import { parse } from 'query-string';
import JSURL from 'jsurl';

import { handleStateChange } from '@ncigdc/dux/relayProgress';
import SmartSearchPage from '@ncigdc/containers/SmartSearchPage';
import { parseIntParam, parseFilterParam } from '@ncigdc/utils/uri';

import { viewerQuery } from './queries';

const DEFAULT_CASES_SORT = null;
const DEFAULT_FILES_SORT = null;

class SmartSearchRoute extends Relay.Route {
  static routeName = 'RepositoryRoute';
  static queries = viewerQuery;
  static prepareParams = ({ location: { search } }) => {
    const q = parse(search);
    const casesSort = q.cases_sort ? JSURL.parse(q.cases_sort) : DEFAULT_CASES_SORT;
    const filesSort = q.files_sort ? JSURL.parse(q.files_sort) : DEFAULT_FILES_SORT;

    return ({
      cases_offset: parseIntParam(q.cases_offset, 0),
      cases_size: parseIntParam(q.cases_size, 20),
      cases_sort: casesSort,
      files_offset: parseIntParam(q.files_offset, 0),
      files_size: parseIntParam(q.files_size, 20),
      files_sort: filesSort,
      filters: parseFilterParam(q.filters, null),
    });
  }
}

export default connect()((props: mixed) => (
  <Relay.Renderer
    Container={SmartSearchPage}
    queryConfig={new SmartSearchRoute(props)}
    environment={Relay.Store}
    onReadyStateChange={handleStateChange(props)}
  />
));
