// containers/AnnotationsAggregations.js
// @flow
import React from 'react';
import Relay from 'react-relay';
import { compose } from 'recompose';
import { createContainer } from 'recompose-relay';

import TermFacet from 'components/TermFacet';

export type Bucket = {
  doc_count: number,
  key: string,
}

export type Props = {
  aggregations: {
    category: { buckets: [Bucket] },
    classification: { buckets: [Bucket] },
    entity_type: { buckets: [Bucket] },
    project__primary_site: { buckets: [Bucket] },
    project__program__name: { buckets: [Bucket] },
    project__project_id: { buckets: [Bucket] },
    status: { buckets: [Bucket] },
  },
  relay: {
    route: {
      params: Object
    }
  }
}

const AnnotationsAggregations = (props: Props) => {
  const docType = 'annotations';
  const facets = [
    'classification',
    'category',
    'entity_type',
    'project__primary_site',
    'project__program__name',
    'project__project_id',
    'status',
  ];

  return (
    <div>
      {facets.map(f => (
        <TermFacet
          key={`${docType}.${f}`}
          pathname={`/${docType}`}
          field={`${docType}.${f}`}
          params={props.relay.route.params}
          buckets={props.aggregations[f].buckets}
        />
      ))}
    </div>
  );
};


const AnnotationsAggregationsQuery = {
  fragments: {
    aggregations: () => Relay.QL`
      fragment on AnnotationsAgg {
        category {
          buckets {
            doc_count
            key
          }
        }
        classification {
          buckets {
            doc_count
            key
          }
        }
        entity_type {
          buckets {
            doc_count
            key
          }
        }
        project__primary_site {
          buckets {
            doc_count
            key
          }
        }
        project__program__name {
          buckets {
            doc_count
            key
          }
        }
        project__project_id {
          buckets {
            doc_count
            key
          }
        }
        status {
          buckets {
            doc_count
            key
          }
        }
      }
    `,
  },
};

export default compose(
  createContainer(AnnotationsAggregationsQuery)
)(AnnotationsAggregations);
