// containers/App.js
import React from 'react';
import Relay from 'react-relay';
import { compose } from 'recompose';
import { createContainer } from 'recompose-relay';
import Router from 'react-router/BrowserRouter';
import Match from 'react-router/Match';
import Link from 'react-router/Link';

import FilesRoute from 'routes/FilesRoute';
import AnnotationsRoute from 'routes/AnnotationsRoute';

const App = () => (
  <Router>
    <div>
      <ul>
        <li><Link to="/">Home</Link></li>
        <li><Link to="/files">Files</Link></li>
        <li><Link to="/annotations">Annotations</Link></li>
      </ul>

      <hr />

      <Match pattern="/files" component={FilesRoute} />
      <Match pattern="/annotations" component={AnnotationsRoute} />
    </div>
  </Router>
);

const AppQuery = {
  initialVariables: {
    first: 0,
    offset: 0,
    filters: null,
  },
  fragments: {
    viewer: () => Relay.QL`
      fragment on Root {
        summary {
          aggregations {
            access {
              buckets {
                doc_count
              }
            }
          }
        }
      }
    `,
  },
};

export default compose(
  createContainer(AppQuery)
)(App);
