// containers/CasesAggregations.js
import React from 'react';
import Relay from 'react-relay';
import { compose } from 'recompose';
import { createContainer } from 'recompose-relay';

import TermFacet from 'components/Aggregations/TermAggregation';
import type { Bucket } from 'components/Aggregations/types';

type Props = {
  aggregations: {
    demographic__ethnicity: { buckets: [Bucket] },
    demographic__gender: { buckets: [Bucket] },
    demographic__race: { buckets: [Bucket] },
    diagnoses__vital_status: { buckets: [Bucket] },
    project__disease_type: { buckets: [Bucket] },
    project__primary_site: { buckets: [Bucket] },
    project__project_id: { buckets: [Bucket] },
  },
};

const CasesAggregations = (props: Props) => {
  const docType = 'cases';
  const facets = [
    'demographic__ethnicity',
    'demographic__gender',
    'demographic__race',
    'diagnoses__vital_status',
    'project__disease_type',
    'project__primary_site',
    'project__project_id',
  ];
  return (
    <div>
      {facets.map(f => (
        <TermFacet
          key={`${docType}.${f}`}
          pathname={'/files'}
          field={`${docType}.${f}`}
          params={props.relay.route.params}
          buckets={props.aggregations[f].buckets}
        />
      ))}
    </div>
  );
};

const CasesAggregationsQuery = {
  fragments: {
    aggregations: () => Relay.QL`
      fragment on CasesAgg {
        demographic__ethnicity {
          buckets {
            doc_count
            key
          }
        }
        demographic__gender {
          buckets {
            doc_count
            key
          }
        }
        demographic__race {
          buckets {
            doc_count
            key
          }
        }
        diagnoses__vital_status {
          buckets {
            doc_count
            key
          }
        }
        project__disease_type {
          buckets {
            doc_count
            key
          }
        }
        project__primary_site {
          buckets {
            doc_count
            key
          }
        }
        project__project_id {
          buckets {
            doc_count
            key
          }
        }
      }
    `,
  },
};

export default compose(
  createContainer(CasesAggregationsQuery)
)(CasesAggregations);
