// containers/CaseTable.js
import React from 'react';
import Relay from 'react-relay';
import { compose } from 'recompose';
import { createContainer } from 'recompose-relay';

import CaseTr from 'containers/CaseTr';

type Props = {
  hits: {
    pageInfo: {
      hasNextPage: boolean,
      hasPreviousPage: boolean,
      startCursor: string,
      endCursor: string,
    },
    edges: {
      node: {
        id: string,
      }
    }
  }
}

const CaseTable = (props: Props) => (
  <table>
    <thead>
      <tr>
        <th>Case UUID</th>
        <th>Project</th>
        <th>Primary Site</th>
        <th>Gender</th>
      </tr>
    </thead>
    <tbody>
      {
        ((props.hits || {}).edges || []).map(e => (
          <CaseTr key={e.node.id} kase={e.node} />
        ))
      }
    </tbody>
  </table>
);

const CaseTableQuery = {
  fragments: {
    hits: () => Relay.QL`
      fragment on CaseConnection {
        pageInfo {
          hasNextPage
          hasPreviousPage
          startCursor
          endCursor
        }
        edges {
          node {
            id
            ${CaseTr.getFragment('kase')}
          }
        }
      }
    `,
  },
};

export default compose(
  createContainer(CaseTableQuery)
)(CaseTable);
