// containers/FilesAggregations.js
import React from 'react';
import Relay from 'react-relay';
import { compose } from 'recompose';
import { createContainer } from 'recompose-relay';

import TermFacet from 'components/TermFacet';

type Bucket = {
  doc_count: number,
  key: string,
}

type Props = {
  aggregations: {
    access: { buckets: [Bucket] },
    data_category: { buckets: [Bucket] },
    data_format: { buckets: [Bucket] },
    data_type: { buckets: [Bucket] },
    experimental_strategy: { buckets: [Bucket] },
    platform: { buckets: [Bucket] },
  }
}

const FilesAggregations = (props: Props) => {
  const docType = 'files';
  const facets = [
    'access',
    'data_category',
    'data_format',
    'data_type',
    'experimental_strategy',
    'platform',
  ];
  return (
    <div>
      {facets.map(f => (
        <TermFacet
          key={`${docType}.${f}`}
          pathname={`/${docType}`}
          field={`${docType}.${f}`}
          params={props.relay.route.params}
          buckets={props.aggregations[f].buckets}
        />
      ))}
    </div>
  );
};

const FilesAggregationsQuery = {
  fragments: {
    aggregations: () => Relay.QL`
      fragment on FilesAgg {
        access {
          buckets {
            doc_count
            key
          }
        }
        data_category {
          buckets {
            doc_count
            key
          }
        }
        data_format {
          buckets {
            doc_count
            key
          }
        }
        data_type {
          buckets {
            doc_count
            key
          }
        }
        experimental_strategy {
          buckets {
            doc_count
            key
          }
        }
        platform {
          buckets {
            doc_count
            key
          }
        }
      }
    `,
  },
};

export default compose(
  createContainer(FilesAggregationsQuery)
)(FilesAggregations);
