// @ncigdc/components/Cohort/MutationsTab.js
/* global API */
// @flow
import React from 'react';
import LocationSubscriber from '@ncigdc/components/LocationSubscriber';
import { compose, withState, lifecycle, withProps, withPropsOnChange } from 'recompose';

import { fetchApi } from '@ncigdc/utils/ajax';
import { Column, Row } from '@ncigdc/uikit/Flex';
import { parseFilterParam } from '@ncigdc/utils/uri';
import { getDefaultCurve } from '@ncigdc/utils/survivalplot';
import type { TRawQuery } from '@ncigdc/utils/uri/types';
import withFilters from '@ncigdc/utils/withFilters.js';

import FrequentMutationsChart from '@ncigdc/containers/FrequentMutationsChart';
import FrequentMutationsTable from '@ncigdc/containers/FrequentMutationsTable';
import SurvivalPlotWrapper from '@ncigdc/components/SurvivalPlotWrapper';

const styles = {
  heading: {
    flexGrow: 1,
    fontSize: '2rem',
    marginBottom: 7,
    marginTop: 7,
  },
  card: {
    backgroundColor: 'white',
  },
};

const initialState = {
  loading: true,
};

export default
compose(
  withFilters(),
  withState('defaultSurvivalData', 'setDefaultSurvivalData', {}),
  withState('selectedSurvivalData', 'setSelectedSurvivalData', {}),
  withState('state', 'setState', initialState),
  withProps(({
    selectedSurvivalData,
    defaultSurvivalData,
    setDefaultSurvivalData,
    filters,
    setState,
  }) => ({
    survivalData: {
      legend: selectedSurvivalData.legend || defaultSurvivalData.legend,
      rawData: selectedSurvivalData.rawData || defaultSurvivalData.rawData,
    },
    updateData: async () => {
      const survivalData = await getDefaultCurve({
        currentFilters: filters,
        slug: 'Cohort',
      });

      setDefaultSurvivalData(survivalData);

      setState(s => ({
        ...s,
        loading: false,
      }));
    },
  })),
  withPropsOnChange(['filters'], ({ updateData }) => {
    updateData();
  })
)(({
  state,
  totalNumCases,
  projectId = '',
  viewer,
  filters,
  survivalData,
  selectedSurvivalData,
  setSelectedSurvivalData,
}) => (
  <Column style={styles.card}>
    <h1 style={{ ...styles.heading, padding: '1rem' }} id="mutated-genes">
      <i className="fa fa-bar-chart-o" style={{ paddingRight: '10px' }} />
        Somatic Mutations
    </h1>

    <Row>
      <Column flex="1" style={{ width: '50%' }}>
        <FrequentMutationsChart
          defaultFilters={filters}
          cohort={viewer.frequentMutationsChartFragment}
          showSurvivalPlot
        />
      </Column>
      <Column flex="1" style={{ width: '50%' }}>
        <SurvivalPlotWrapper
          {...survivalData}
          onReset={() => setSelectedSurvivalData({})}
          height={240}
          survivalPlotloading={false}
        />
      </Column>
    </Row>

    <FrequentMutationsTable
      defaultFilters={filters}
      cohort={viewer.frequentMutationsTableFragment}
      selectedSurvivalData={selectedSurvivalData}
      setSelectedSurvivalData={setSelectedSurvivalData}
      showSurvivalPlot
    />
  </Column>
));
