// @ncigdc/components/Aggregations/TermAggregation.js
/* @flow */

import React from 'react';
import LocationSubscriber from '@ncigdc/components/LocationSubscriber';
import { compose, withState, mapProps, pure, lifecycle } from 'recompose';
import AngleIcon from 'react-icons/lib/fa/angle-down';

import SearchIcon from '@ncigdc/theme/icons/SearchIcon';
import CloseIcon from '@ncigdc/theme/icons/CloseIcon';
import type { TRawQuery } from '@ncigdc/utils/uri/types';
import { parseFilterParam } from '@ncigdc/utils/uri';
import { inCurrentFilters } from '@ncigdc/utils/filters';

import { Row, Column } from '@ncigdc/uikit/Flex';
import A from '@ncigdc/uikit/A';
import CountBubble from '@ncigdc/uikit/CountBubble';
import styled from '@ncigdc/theme/styled';
import Input from '@ncigdc/uikit/Form/Input';

import Link from '../Links/Link';
import FacetResetButton from './FacetResetButton';

import type { TBucket } from './types';

type TProps = {
  buckets: [TBucket],
  field: string,
  state: {
    collapsed: boolean,
    showingMore: boolean,
    showingValueSearch: boolean,
    filteredBuckets: Array<Object>,
  },
  style: Object,
  title: string,
  toggleCollapsed: Function,
  toggleShowMore: Function,
  toggleValueSearch: Function,
};

const Container = styled(Column, ({
  padding: '1rem 1.2rem',
  backgroundColor: 'white',
}));

const BucketLink = styled(Link, ({
  color: '#3a3a3a',
  ':link': {
    textDecoration: 'none',
  },
}));

const ToggleMoreLink = styled(A, ({
  marginLeft: 'auto',
  color: theme => theme.primaryLight1,
  fontSize: '1.2rem',
  cursor: 'pointer',
  ':link': {
    textDecoration: 'underline',
    color: theme => theme.primaryLight1,
  },
}));

const HeaderRow = styled(Row, ({
  color: ({ theme }) => theme.primary,
  fontSize: '1.7rem',
  marginBottom: '0.5rem',
  cursor: 'pointer',
  alignItems: 'center',
  justifyContent: 'space-between',
}));

const BucketRow = styled(Row, ({
  padding: '0.3rem 0',
}));

const BottomRow = styled(Row, ({
  padding: '0.5rem',
}));

let input;
const TermAggregation = (props: TProps) => {
  const dotField = props.field.replace(/__/g, '.');
  const { filteredBuckets } = props.state;

  const hasValueSearch = filteredBuckets.length >= 20;
  return (
    <LocationSubscriber>{(ctx: {| pathname: string, query: TRawQuery |}) => {
      const currentFilters = ctx.query && parseFilterParam((ctx.query || {}).filters, {}).content || [];
      return (
        <Container style={props.style}>
          <HeaderRow>
            <span style={{ cursor: 'pointer' }} onClick={() => props.toggleCollapsed()}>
              <AngleIcon style={{ transform: `rotate(${props.state.collapsed ? 270 : 0}deg)` }} />
              {props.title}
            </span>
            <span>
              { (hasValueSearch || props.state.showingValueSearch) &&
                <SearchIcon onClick={() => props.toggleValueSearch()} style={{ paddingRight: '5px' }}/>}
              <FacetResetButton field={dotField} currentFilters={currentFilters} />
            </span>
          </HeaderRow>
          {!props.state.collapsed && props.state.showingValueSearch &&
            <Row>
              <Input
                getNode={node => { input = node; }}
                style={{ borderRadius: '4px', marginBottom: '6px' }}
                onChange={() => props.filterBuckets(props.buckets, input.value)}
                placeholder={'Search...'}
              />
              { input && input.value &&
                <CloseIcon
                  style={{
                    position: 'absolute',
                    right: 0,
                    padding: '10px',
                    transition: 'all 0.3s ease',
                    outline: 0,
                  }}
                  onClick={() => {
                    props.filterBuckets(props.buckets, '');
                    input.value = '';
                  }}
                />
              }
            </Row>
          }
          {!props.state.collapsed &&
            <Column>
              {filteredBuckets
                .slice(0, props.state.showingMore ? Infinity : 5)
                .map(b => ({ ...b, name: b.key === '1' ? 'Cancer Gene Sensus' : b.key })) // TODO: this needs to change in the data
                .map(bucket => (
                  <BucketRow key={bucket.key}>
                    <BucketLink
                      merge="toggle"
                      query={{
                        offset: 0,
                        filters: {
                          op: 'and',
                          content: [{
                            op: 'in',
                            content: {
                              field: dotField,
                              value: [bucket.key],
                            },
                          }],
                        },
                      }}
                    >
                      <input
                        readOnly
                        type="checkbox"
                        style={{ pointerEvents: 'none', marginRight: '5px' }}
                        checked={inCurrentFilters({ key: bucket.key, dotField, currentFilters })}
                        id={`input-${bucket.name.replace(/\s/g, '-')}`}
                        name={`input-${bucket.name.replace(/\s/g, '-')}`}
                      />
                      <label htmlFor={`input-${bucket.name.replace(/\s/g, '-')}`} style={{ marginLeft: '0.3rem' }}>{bucket.name}</label>
                    </BucketLink>
                    <CountBubble>{bucket.doc_count.toLocaleString()}</CountBubble>
                  </BucketRow>
                ))
            }
              {filteredBuckets.length > 5 &&
                <BottomRow>
                  <ToggleMoreLink onClick={() => props.toggleShowMore()}>
                    {props.state.showingMore
                      ? 'Less...'
                      : filteredBuckets.length - 5 && 'More...'
                    }
                  </ToggleMoreLink>
                </BottomRow>
              }

              { filteredBuckets.length === 0 &&
                <span>{(input || { value: '' }).value ?
                  'No matching values' :
                    'No data for this field'}
                </span>
              }
            </Column>
          }
        </Container>
      );
    }}
    </LocationSubscriber>
  );
};

const enhance = compose(
  withState('state', 'setState', {
    collapsed: false,
    showingMore: false,
    showingValueSearch: false,
    filteredBuckets: [],
  }),
  mapProps(({ setState, ...rest }) => ({
    toggleCollapsed: () => setState(s => ({ ...s, collapsed: !s.collapsed})),
    toggleShowMore: () => setState(s => ({ ...s, showingMore: !s.showingMore })),
    toggleValueSearch: () => setState(s => ({ ...s, showingValueSearch: !s.showingValueSearch})),
    filterBuckets: (buckets, searchValue) => setState(s => ({
      ...s,
      filteredBuckets: buckets.filter(b =>
                                        b.key !== '_missing' &&
                                        b.key.toLowerCase().indexOf(searchValue.toLowerCase()) !== -1),
    })),
    setState,
    ...rest,
  })),
  lifecycle({
    componentDidMount(): void {
      const { setState, buckets } = this.props;
      setState(s => ({
        ...s,
        filteredBuckets: buckets.filter(b => b.key !== '_missing'),
      }));
    },
  }),
  pure
);

export default enhance(TermAggregation);
