// @ncigdc/utils/generated-relay-query-parts/repositoryCaseAggregations.js
/* @flow */
/* eslint max-len:0, comma-dangle:0 */

import Relay from 'react-relay';

export const initialCaseAggregationsVariables = {
  shouldShow_aliquot_ids: false,
  shouldShow_analyte_ids: false,
  shouldShow_annotations__annotation_id: false,
  shouldShow_annotations__case_id: false,
  shouldShow_annotations__case_submitter_id: false,
  shouldShow_annotations__category: false,
  shouldShow_annotations__classification: false,
  shouldShow_annotations__created_datetime: false,
  shouldShow_annotations__creator: false,
  shouldShow_annotations__entity_id: false,
  shouldShow_annotations__entity_submitter_id: false,
  shouldShow_annotations__entity_type: false,
  shouldShow_annotations__legacy_created_datetime: false,
  shouldShow_annotations__legacy_updated_datetime: false,
  shouldShow_annotations__notes: false,
  shouldShow_annotations__state: false,
  shouldShow_annotations__status: false,
  shouldShow_annotations__submitter_id: false,
  shouldShow_annotations__updated_datetime: false,
  shouldShow_case_id: false,
  shouldShow_created_datetime: false,
  shouldShow_days_to_index: false,
  shouldShow_demographic__created_datetime: false,
  shouldShow_demographic__demographic_id: false,
  shouldShow_demographic__ethnicity: false,
  shouldShow_demographic__gender: false,
  shouldShow_demographic__race: false,
  shouldShow_demographic__state: false,
  shouldShow_demographic__submitter_id: false,
  shouldShow_demographic__updated_datetime: false,
  shouldShow_demographic__year_of_birth: false,
  shouldShow_demographic__year_of_death: false,
  shouldShow_diagnoses__age_at_diagnosis: false,
  shouldShow_diagnoses__ajcc_clinical_m: false,
  shouldShow_diagnoses__ajcc_clinical_n: false,
  shouldShow_diagnoses__ajcc_clinical_stage: false,
  shouldShow_diagnoses__ajcc_clinical_t: false,
  shouldShow_diagnoses__ajcc_pathologic_m: false,
  shouldShow_diagnoses__ajcc_pathologic_n: false,
  shouldShow_diagnoses__ajcc_pathologic_stage: false,
  shouldShow_diagnoses__ajcc_pathologic_t: false,
  shouldShow_diagnoses__ann_arbor_b_symptoms: false,
  shouldShow_diagnoses__ann_arbor_clinical_stage: false,
  shouldShow_diagnoses__ann_arbor_extranodal_involvement: false,
  shouldShow_diagnoses__ann_arbor_pathologic_stage: false,
  shouldShow_diagnoses__burkitt_lymphoma_clinical_variant: false,
  shouldShow_diagnoses__cause_of_death: false,
  shouldShow_diagnoses__circumferential_resection_margin: false,
  shouldShow_diagnoses__classification_of_tumor: false,
  shouldShow_diagnoses__colon_polyps_history: false,
  shouldShow_diagnoses__created_datetime: false,
  shouldShow_diagnoses__days_to_birth: false,
  shouldShow_diagnoses__days_to_death: false,
  shouldShow_diagnoses__days_to_hiv_diagnosis: false,
  shouldShow_diagnoses__days_to_last_follow_up: false,
  shouldShow_diagnoses__days_to_last_known_disease_status: false,
  shouldShow_diagnoses__days_to_new_event: false,
  shouldShow_diagnoses__days_to_recurrence: false,
  shouldShow_diagnoses__diagnosis_id: false,
  shouldShow_diagnoses__figo_stage: false,
  shouldShow_diagnoses__hiv_positive: false,
  shouldShow_diagnoses__hpv_positive_type: false,
  shouldShow_diagnoses__hpv_status: false,
  shouldShow_diagnoses__last_known_disease_status: false,
  shouldShow_diagnoses__laterality: false,
  shouldShow_diagnoses__ldh_level_at_diagnosis: false,
  shouldShow_diagnoses__ldh_normal_range_upper: false,
  shouldShow_diagnoses__lymph_nodes_positive: false,
  shouldShow_diagnoses__lymphatic_invasion_present: false,
  shouldShow_diagnoses__method_of_diagnosis: false,
  shouldShow_diagnoses__morphology: false,
  shouldShow_diagnoses__new_event_anatomic_site: false,
  shouldShow_diagnoses__new_event_type: false,
  shouldShow_diagnoses__perineural_invasion_present: false,
  shouldShow_diagnoses__primary_diagnosis: false,
  shouldShow_diagnoses__prior_malignancy: false,
  shouldShow_diagnoses__prior_treatment: false,
  shouldShow_diagnoses__progression_or_recurrence: false,
  shouldShow_diagnoses__residual_disease: false,
  shouldShow_diagnoses__site_of_resection_or_biopsy: false,
  shouldShow_diagnoses__state: false,
  shouldShow_diagnoses__submitter_id: false,
  shouldShow_diagnoses__tissue_or_organ_of_origin: false,
  shouldShow_diagnoses__treatments__created_datetime: false,
  shouldShow_diagnoses__treatments__days_to_treatment: false,
  shouldShow_diagnoses__treatments__days_to_treatment_end: false,
  shouldShow_diagnoses__treatments__days_to_treatment_start: false,
  shouldShow_diagnoses__treatments__state: false,
  shouldShow_diagnoses__treatments__submitter_id: false,
  shouldShow_diagnoses__treatments__therapeutic_agents: false,
  shouldShow_diagnoses__treatments__treatment_anatomic_site: false,
  shouldShow_diagnoses__treatments__treatment_id: false,
  shouldShow_diagnoses__treatments__treatment_intent_type: false,
  shouldShow_diagnoses__treatments__treatment_or_therapy: false,
  shouldShow_diagnoses__treatments__treatment_outcome: false,
  shouldShow_diagnoses__treatments__treatment_type: false,
  shouldShow_diagnoses__treatments__updated_datetime: false,
  shouldShow_diagnoses__tumor_grade: false,
  shouldShow_diagnoses__tumor_stage: false,
  shouldShow_diagnoses__updated_datetime: false,
  shouldShow_diagnoses__vascular_invasion_present: false,
  shouldShow_diagnoses__vital_status: false,
  shouldShow_diagnoses__year_of_diagnosis: false,
  shouldShow_disease_type: false,
  shouldShow_exposures__alcohol_history: false,
  shouldShow_exposures__alcohol_intensity: false,
  shouldShow_exposures__bmi: false,
  shouldShow_exposures__cigarettes_per_day: false,
  shouldShow_exposures__created_datetime: false,
  shouldShow_exposures__exposure_id: false,
  shouldShow_exposures__height: false,
  shouldShow_exposures__pack_years_smoked: false,
  shouldShow_exposures__state: false,
  shouldShow_exposures__submitter_id: false,
  shouldShow_exposures__tobacco_smoking_onset_year: false,
  shouldShow_exposures__tobacco_smoking_quit_year: false,
  shouldShow_exposures__tobacco_smoking_status: false,
  shouldShow_exposures__updated_datetime: false,
  shouldShow_exposures__weight: false,
  shouldShow_exposures__years_smoked: false,
  shouldShow_family_histories__created_datetime: false,
  shouldShow_family_histories__family_history_id: false,
  shouldShow_family_histories__relationship_age_at_diagnosis: false,
  shouldShow_family_histories__relationship_gender: false,
  shouldShow_family_histories__relationship_primary_diagnosis: false,
  shouldShow_family_histories__relationship_type: false,
  shouldShow_family_histories__relative_with_cancer_history: false,
  shouldShow_family_histories__state: false,
  shouldShow_family_histories__submitter_id: false,
  shouldShow_family_histories__updated_datetime: false,
  shouldShow_portion_ids: false,
  shouldShow_primary_site: false,
  shouldShow_project__dbgap_accession_number: false,
  shouldShow_project__disease_type: false,
  shouldShow_project__intended_release_date: false,
  shouldShow_project__name: false,
  shouldShow_project__primary_site: false,
  shouldShow_project__program__dbgap_accession_number: false,
  shouldShow_project__program__name: false,
  shouldShow_project__program__program_id: false,
  shouldShow_project__project_id: false,
  shouldShow_project__releasable: false,
  shouldShow_project__released: false,
  shouldShow_project__state: false,
  shouldShow_sample_ids: false,
  shouldShow_samples__annotations__annotation_id: false,
  shouldShow_samples__annotations__case_id: false,
  shouldShow_samples__annotations__case_submitter_id: false,
  shouldShow_samples__annotations__category: false,
  shouldShow_samples__annotations__classification: false,
  shouldShow_samples__annotations__created_datetime: false,
  shouldShow_samples__annotations__creator: false,
  shouldShow_samples__annotations__entity_id: false,
  shouldShow_samples__annotations__entity_submitter_id: false,
  shouldShow_samples__annotations__entity_type: false,
  shouldShow_samples__annotations__legacy_created_datetime: false,
  shouldShow_samples__annotations__legacy_updated_datetime: false,
  shouldShow_samples__annotations__notes: false,
  shouldShow_samples__annotations__state: false,
  shouldShow_samples__annotations__status: false,
  shouldShow_samples__annotations__submitter_id: false,
  shouldShow_samples__annotations__updated_datetime: false,
  shouldShow_samples__biospecimen_anatomic_site: false,
  shouldShow_samples__composition: false,
  shouldShow_samples__created_datetime: false,
  shouldShow_samples__current_weight: false,
  shouldShow_samples__days_to_collection: false,
  shouldShow_samples__days_to_sample_procurement: false,
  shouldShow_samples__diagnosis_pathologically_confirmed: false,
  shouldShow_samples__freezing_method: false,
  shouldShow_samples__initial_weight: false,
  shouldShow_samples__intermediate_dimension: false,
  shouldShow_samples__is_ffpe: false,
  shouldShow_samples__longest_dimension: false,
  shouldShow_samples__method_of_sample_procurement: false,
  shouldShow_samples__oct_embedded: false,
  shouldShow_samples__pathology_report_uuid: false,
  shouldShow_samples__portions__analytes__a260_a280_ratio: false,
  shouldShow_samples__portions__analytes__aliquots__aliquot_id: false,
  shouldShow_samples__portions__analytes__aliquots__aliquot_quantity: false,
  shouldShow_samples__portions__analytes__aliquots__aliquot_volume: false,
  shouldShow_samples__portions__analytes__aliquots__amount: false,
  shouldShow_samples__portions__analytes__aliquots__analyte_type: false,
  shouldShow_samples__portions__analytes__aliquots__analyte_type_id: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__annotation_id: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__case_id: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__case_submitter_id: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__category: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__classification: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__created_datetime: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__creator: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__entity_id: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__entity_submitter_id: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__entity_type: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__legacy_created_datetime: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__legacy_updated_datetime: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__notes: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__state: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__status: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__submitter_id: false,
  shouldShow_samples__portions__analytes__aliquots__annotations__updated_datetime: false,
  shouldShow_samples__portions__analytes__aliquots__center__center_id: false,
  shouldShow_samples__portions__analytes__aliquots__center__center_type: false,
  shouldShow_samples__portions__analytes__aliquots__center__code: false,
  shouldShow_samples__portions__analytes__aliquots__center__name: false,
  shouldShow_samples__portions__analytes__aliquots__center__namespace: false,
  shouldShow_samples__portions__analytes__aliquots__center__short_name: false,
  shouldShow_samples__portions__analytes__aliquots__concentration: false,
  shouldShow_samples__portions__analytes__aliquots__created_datetime: false,
  shouldShow_samples__portions__analytes__aliquots__source_center: false,
  shouldShow_samples__portions__analytes__aliquots__state: false,
  shouldShow_samples__portions__analytes__aliquots__submitter_id: false,
  shouldShow_samples__portions__analytes__aliquots__updated_datetime: false,
  shouldShow_samples__portions__analytes__amount: false,
  shouldShow_samples__portions__analytes__analyte_id: false,
  shouldShow_samples__portions__analytes__analyte_quantity: false,
  shouldShow_samples__portions__analytes__analyte_type: false,
  shouldShow_samples__portions__analytes__analyte_type_id: false,
  shouldShow_samples__portions__analytes__analyte_volume: false,
  shouldShow_samples__portions__analytes__annotations__annotation_id: false,
  shouldShow_samples__portions__analytes__annotations__case_id: false,
  shouldShow_samples__portions__analytes__annotations__case_submitter_id: false,
  shouldShow_samples__portions__analytes__annotations__category: false,
  shouldShow_samples__portions__analytes__annotations__classification: false,
  shouldShow_samples__portions__analytes__annotations__created_datetime: false,
  shouldShow_samples__portions__analytes__annotations__creator: false,
  shouldShow_samples__portions__analytes__annotations__entity_id: false,
  shouldShow_samples__portions__analytes__annotations__entity_submitter_id: false,
  shouldShow_samples__portions__analytes__annotations__entity_type: false,
  shouldShow_samples__portions__analytes__annotations__legacy_created_datetime: false,
  shouldShow_samples__portions__analytes__annotations__legacy_updated_datetime: false,
  shouldShow_samples__portions__analytes__annotations__notes: false,
  shouldShow_samples__portions__analytes__annotations__state: false,
  shouldShow_samples__portions__analytes__annotations__status: false,
  shouldShow_samples__portions__analytes__annotations__submitter_id: false,
  shouldShow_samples__portions__analytes__annotations__updated_datetime: false,
  shouldShow_samples__portions__analytes__concentration: false,
  shouldShow_samples__portions__analytes__created_datetime: false,
  shouldShow_samples__portions__analytes__normal_tumor_genotype_snp_match: false,
  shouldShow_samples__portions__analytes__ribosomal_rna_28s_16s_ratio: false,
  shouldShow_samples__portions__analytes__spectrophotometer_method: false,
  shouldShow_samples__portions__analytes__state: false,
  shouldShow_samples__portions__analytes__submitter_id: false,
  shouldShow_samples__portions__analytes__updated_datetime: false,
  shouldShow_samples__portions__analytes__well_number: false,
  shouldShow_samples__portions__annotations__annotation_id: false,
  shouldShow_samples__portions__annotations__case_id: false,
  shouldShow_samples__portions__annotations__case_submitter_id: false,
  shouldShow_samples__portions__annotations__category: false,
  shouldShow_samples__portions__annotations__classification: false,
  shouldShow_samples__portions__annotations__created_datetime: false,
  shouldShow_samples__portions__annotations__creator: false,
  shouldShow_samples__portions__annotations__entity_id: false,
  shouldShow_samples__portions__annotations__entity_submitter_id: false,
  shouldShow_samples__portions__annotations__entity_type: false,
  shouldShow_samples__portions__annotations__legacy_created_datetime: false,
  shouldShow_samples__portions__annotations__legacy_updated_datetime: false,
  shouldShow_samples__portions__annotations__notes: false,
  shouldShow_samples__portions__annotations__state: false,
  shouldShow_samples__portions__annotations__status: false,
  shouldShow_samples__portions__annotations__submitter_id: false,
  shouldShow_samples__portions__annotations__updated_datetime: false,
  shouldShow_samples__portions__center__center_id: false,
  shouldShow_samples__portions__center__center_type: false,
  shouldShow_samples__portions__center__code: false,
  shouldShow_samples__portions__center__name: false,
  shouldShow_samples__portions__center__namespace: false,
  shouldShow_samples__portions__center__short_name: false,
  shouldShow_samples__portions__created_datetime: false,
  shouldShow_samples__portions__creation_datetime: false,
  shouldShow_samples__portions__is_ffpe: false,
  shouldShow_samples__portions__portion_id: false,
  shouldShow_samples__portions__portion_number: false,
  shouldShow_samples__portions__slides__annotations__annotation_id: false,
  shouldShow_samples__portions__slides__annotations__case_id: false,
  shouldShow_samples__portions__slides__annotations__case_submitter_id: false,
  shouldShow_samples__portions__slides__annotations__category: false,
  shouldShow_samples__portions__slides__annotations__classification: false,
  shouldShow_samples__portions__slides__annotations__created_datetime: false,
  shouldShow_samples__portions__slides__annotations__creator: false,
  shouldShow_samples__portions__slides__annotations__entity_id: false,
  shouldShow_samples__portions__slides__annotations__entity_submitter_id: false,
  shouldShow_samples__portions__slides__annotations__entity_type: false,
  shouldShow_samples__portions__slides__annotations__legacy_created_datetime: false,
  shouldShow_samples__portions__slides__annotations__legacy_updated_datetime: false,
  shouldShow_samples__portions__slides__annotations__notes: false,
  shouldShow_samples__portions__slides__annotations__state: false,
  shouldShow_samples__portions__slides__annotations__status: false,
  shouldShow_samples__portions__slides__annotations__submitter_id: false,
  shouldShow_samples__portions__slides__annotations__updated_datetime: false,
  shouldShow_samples__portions__slides__created_datetime: false,
  shouldShow_samples__portions__slides__number_proliferating_cells: false,
  shouldShow_samples__portions__slides__percent_eosinophil_infiltration: false,
  shouldShow_samples__portions__slides__percent_granulocyte_infiltration: false,
  shouldShow_samples__portions__slides__percent_inflam_infiltration: false,
  shouldShow_samples__portions__slides__percent_lymphocyte_infiltration: false,
  shouldShow_samples__portions__slides__percent_monocyte_infiltration: false,
  shouldShow_samples__portions__slides__percent_necrosis: false,
  shouldShow_samples__portions__slides__percent_neutrophil_infiltration: false,
  shouldShow_samples__portions__slides__percent_normal_cells: false,
  shouldShow_samples__portions__slides__percent_stromal_cells: false,
  shouldShow_samples__portions__slides__percent_tumor_cells: false,
  shouldShow_samples__portions__slides__percent_tumor_nuclei: false,
  shouldShow_samples__portions__slides__section_location: false,
  shouldShow_samples__portions__slides__slide_id: false,
  shouldShow_samples__portions__slides__state: false,
  shouldShow_samples__portions__slides__submitter_id: false,
  shouldShow_samples__portions__slides__updated_datetime: false,
  shouldShow_samples__portions__state: false,
  shouldShow_samples__portions__submitter_id: false,
  shouldShow_samples__portions__updated_datetime: false,
  shouldShow_samples__portions__weight: false,
  shouldShow_samples__preservation_method: false,
  shouldShow_samples__sample_id: false,
  shouldShow_samples__sample_type: false,
  shouldShow_samples__sample_type_id: false,
  shouldShow_samples__shortest_dimension: false,
  shouldShow_samples__state: false,
  shouldShow_samples__submitter_id: false,
  shouldShow_samples__time_between_clamping_and_freezing: false,
  shouldShow_samples__time_between_excision_and_freezing: false,
  shouldShow_samples__tissue_type: false,
  shouldShow_samples__tumor_code: false,
  shouldShow_samples__tumor_code_id: false,
  shouldShow_samples__tumor_descriptor: false,
  shouldShow_samples__updated_datetime: false,
  shouldShow_slide_ids: false,
  shouldShow_state: false,
  shouldShow_submitter_aliquot_ids: false,
  shouldShow_submitter_analyte_ids: false,
  shouldShow_submitter_id: false,
  shouldShow_submitter_portion_ids: false,
  shouldShow_submitter_sample_ids: false,
  shouldShow_submitter_slide_ids: false,
  shouldShow_summary__data_categories__data_category: false,
  shouldShow_summary__data_categories__file_count: false,
  shouldShow_summary__experimental_strategies__experimental_strategy: false,
  shouldShow_summary__experimental_strategies__file_count: false,
  shouldShow_summary__file_count: false,
  shouldShow_summary__file_size: false,
  shouldShow_tissue_source_site__bcr_id: false,
  shouldShow_tissue_source_site__code: false,
  shouldShow_tissue_source_site__name: false,
  shouldShow_tissue_source_site__project: false,
  shouldShow_tissue_source_site__tissue_source_site_id: false,
  shouldShow_updated_datetime: false
};

export const repositoryCaseAggregationsFragment = () => Relay.QL`
  fragment on CaseAggregations {

aliquot_ids @include(if: $shouldShow_aliquot_ids) {

      buckets {
        doc_count
        key
      }

}

analyte_ids @include(if: $shouldShow_analyte_ids) {

      buckets {
        doc_count
        key
      }

}

annotations__annotation_id @include(if: $shouldShow_annotations__annotation_id) {

      buckets {
        doc_count
        key
      }

}

annotations__case_id @include(if: $shouldShow_annotations__case_id) {

      buckets {
        doc_count
        key
      }

}

annotations__case_submitter_id @include(if: $shouldShow_annotations__case_submitter_id) {

      buckets {
        doc_count
        key
      }

}

annotations__category @include(if: $shouldShow_annotations__category) {

      buckets {
        doc_count
        key
      }

}

annotations__classification @include(if: $shouldShow_annotations__classification) {

      buckets {
        doc_count
        key
      }

}


annotations__creator @include(if: $shouldShow_annotations__creator) {

      buckets {
        doc_count
        key
      }

}

annotations__entity_id @include(if: $shouldShow_annotations__entity_id) {

      buckets {
        doc_count
        key
      }

}

annotations__entity_submitter_id @include(if: $shouldShow_annotations__entity_submitter_id) {

      buckets {
        doc_count
        key
      }

}

annotations__entity_type @include(if: $shouldShow_annotations__entity_type) {

      buckets {
        doc_count
        key
      }

}



annotations__notes @include(if: $shouldShow_annotations__notes) {

      buckets {
        doc_count
        key
      }

}

annotations__state @include(if: $shouldShow_annotations__state) {

      buckets {
        doc_count
        key
      }

}

annotations__status @include(if: $shouldShow_annotations__status) {

      buckets {
        doc_count
        key
      }

}

annotations__submitter_id @include(if: $shouldShow_annotations__submitter_id) {

      buckets {
        doc_count
        key
      }

}


case_id @include(if: $shouldShow_case_id) {

      buckets {
        doc_count
        key
      }

}


days_to_index @include(if: $shouldShow_days_to_index) {

      count
      max
      avg
      min

}


demographic__demographic_id @include(if: $shouldShow_demographic__demographic_id) {

      buckets {
        doc_count
        key
      }

}

demographic__ethnicity @include(if: $shouldShow_demographic__ethnicity) {

      buckets {
        doc_count
        key
      }

}

demographic__gender @include(if: $shouldShow_demographic__gender) {

      buckets {
        doc_count
        key
      }

}

demographic__race @include(if: $shouldShow_demographic__race) {

      buckets {
        doc_count
        key
      }

}

demographic__state @include(if: $shouldShow_demographic__state) {

      buckets {
        doc_count
        key
      }

}

demographic__submitter_id @include(if: $shouldShow_demographic__submitter_id) {

      buckets {
        doc_count
        key
      }

}


demographic__year_of_birth @include(if: $shouldShow_demographic__year_of_birth) {

      count
      max
      avg
      min

}

demographic__year_of_death @include(if: $shouldShow_demographic__year_of_death) {

      count
      max
      avg
      min

}

diagnoses__age_at_diagnosis @include(if: $shouldShow_diagnoses__age_at_diagnosis) {

      count
      max
      avg
      min

}

diagnoses__ajcc_clinical_m @include(if: $shouldShow_diagnoses__ajcc_clinical_m) {

      buckets {
        doc_count
        key
      }

}

diagnoses__ajcc_clinical_n @include(if: $shouldShow_diagnoses__ajcc_clinical_n) {

      buckets {
        doc_count
        key
      }

}

diagnoses__ajcc_clinical_stage @include(if: $shouldShow_diagnoses__ajcc_clinical_stage) {

      buckets {
        doc_count
        key
      }

}

diagnoses__ajcc_clinical_t @include(if: $shouldShow_diagnoses__ajcc_clinical_t) {

      buckets {
        doc_count
        key
      }

}

diagnoses__ajcc_pathologic_m @include(if: $shouldShow_diagnoses__ajcc_pathologic_m) {

      buckets {
        doc_count
        key
      }

}

diagnoses__ajcc_pathologic_n @include(if: $shouldShow_diagnoses__ajcc_pathologic_n) {

      buckets {
        doc_count
        key
      }

}

diagnoses__ajcc_pathologic_stage @include(if: $shouldShow_diagnoses__ajcc_pathologic_stage) {

      buckets {
        doc_count
        key
      }

}

diagnoses__ajcc_pathologic_t @include(if: $shouldShow_diagnoses__ajcc_pathologic_t) {

      buckets {
        doc_count
        key
      }

}

diagnoses__ann_arbor_b_symptoms @include(if: $shouldShow_diagnoses__ann_arbor_b_symptoms) {

      buckets {
        doc_count
        key
      }

}

diagnoses__ann_arbor_clinical_stage @include(if: $shouldShow_diagnoses__ann_arbor_clinical_stage) {

      buckets {
        doc_count
        key
      }

}

diagnoses__ann_arbor_extranodal_involvement @include(if: $shouldShow_diagnoses__ann_arbor_extranodal_involvement) {

      buckets {
        doc_count
        key
      }

}

diagnoses__ann_arbor_pathologic_stage @include(if: $shouldShow_diagnoses__ann_arbor_pathologic_stage) {

      buckets {
        doc_count
        key
      }

}

diagnoses__burkitt_lymphoma_clinical_variant @include(if: $shouldShow_diagnoses__burkitt_lymphoma_clinical_variant) {

      buckets {
        doc_count
        key
      }

}

diagnoses__cause_of_death @include(if: $shouldShow_diagnoses__cause_of_death) {

      buckets {
        doc_count
        key
      }

}

diagnoses__circumferential_resection_margin @include(if: $shouldShow_diagnoses__circumferential_resection_margin) {

      count
      max
      avg
      min

}

diagnoses__classification_of_tumor @include(if: $shouldShow_diagnoses__classification_of_tumor) {

      buckets {
        doc_count
        key
      }

}

diagnoses__colon_polyps_history @include(if: $shouldShow_diagnoses__colon_polyps_history) {

      buckets {
        doc_count
        key
      }

}


diagnoses__days_to_birth @include(if: $shouldShow_diagnoses__days_to_birth) {

      count
      max
      avg
      min

}

diagnoses__days_to_death @include(if: $shouldShow_diagnoses__days_to_death) {

      count
      max
      avg
      min

}

diagnoses__days_to_hiv_diagnosis @include(if: $shouldShow_diagnoses__days_to_hiv_diagnosis) {

      count
      max
      avg
      min

}

diagnoses__days_to_last_follow_up @include(if: $shouldShow_diagnoses__days_to_last_follow_up) {

      count
      max
      avg
      min

}

diagnoses__days_to_last_known_disease_status @include(if: $shouldShow_diagnoses__days_to_last_known_disease_status) {

      count
      max
      avg
      min

}

diagnoses__days_to_new_event @include(if: $shouldShow_diagnoses__days_to_new_event) {

      count
      max
      avg
      min

}

diagnoses__days_to_recurrence @include(if: $shouldShow_diagnoses__days_to_recurrence) {

      count
      max
      avg
      min

}

diagnoses__diagnosis_id @include(if: $shouldShow_diagnoses__diagnosis_id) {

      buckets {
        doc_count
        key
      }

}

diagnoses__figo_stage @include(if: $shouldShow_diagnoses__figo_stage) {

      buckets {
        doc_count
        key
      }

}

diagnoses__hiv_positive @include(if: $shouldShow_diagnoses__hiv_positive) {

      buckets {
        doc_count
        key
      }

}

diagnoses__hpv_positive_type @include(if: $shouldShow_diagnoses__hpv_positive_type) {

      buckets {
        doc_count
        key
      }

}

diagnoses__hpv_status @include(if: $shouldShow_diagnoses__hpv_status) {

      buckets {
        doc_count
        key
      }

}

diagnoses__last_known_disease_status @include(if: $shouldShow_diagnoses__last_known_disease_status) {

      buckets {
        doc_count
        key
      }

}

diagnoses__laterality @include(if: $shouldShow_diagnoses__laterality) {

      buckets {
        doc_count
        key
      }

}

diagnoses__ldh_level_at_diagnosis @include(if: $shouldShow_diagnoses__ldh_level_at_diagnosis) {

      count
      max
      avg
      min

}

diagnoses__ldh_normal_range_upper @include(if: $shouldShow_diagnoses__ldh_normal_range_upper) {

      count
      max
      avg
      min

}

diagnoses__lymph_nodes_positive @include(if: $shouldShow_diagnoses__lymph_nodes_positive) {

      count
      max
      avg
      min

}

diagnoses__lymphatic_invasion_present @include(if: $shouldShow_diagnoses__lymphatic_invasion_present) {

      buckets {
        doc_count
        key
      }

}

diagnoses__method_of_diagnosis @include(if: $shouldShow_diagnoses__method_of_diagnosis) {

      buckets {
        doc_count
        key
      }

}

diagnoses__morphology @include(if: $shouldShow_diagnoses__morphology) {

      buckets {
        doc_count
        key
      }

}

diagnoses__new_event_anatomic_site @include(if: $shouldShow_diagnoses__new_event_anatomic_site) {

      buckets {
        doc_count
        key
      }

}

diagnoses__new_event_type @include(if: $shouldShow_diagnoses__new_event_type) {

      buckets {
        doc_count
        key
      }

}

diagnoses__perineural_invasion_present @include(if: $shouldShow_diagnoses__perineural_invasion_present) {

      buckets {
        doc_count
        key
      }

}

diagnoses__primary_diagnosis @include(if: $shouldShow_diagnoses__primary_diagnosis) {

      buckets {
        doc_count
        key
      }

}

diagnoses__prior_malignancy @include(if: $shouldShow_diagnoses__prior_malignancy) {

      buckets {
        doc_count
        key
      }

}

diagnoses__prior_treatment @include(if: $shouldShow_diagnoses__prior_treatment) {

      buckets {
        doc_count
        key
      }

}

diagnoses__progression_or_recurrence @include(if: $shouldShow_diagnoses__progression_or_recurrence) {

      buckets {
        doc_count
        key
      }

}

diagnoses__residual_disease @include(if: $shouldShow_diagnoses__residual_disease) {

      buckets {
        doc_count
        key
      }

}

diagnoses__site_of_resection_or_biopsy @include(if: $shouldShow_diagnoses__site_of_resection_or_biopsy) {

      buckets {
        doc_count
        key
      }

}

diagnoses__state @include(if: $shouldShow_diagnoses__state) {

      buckets {
        doc_count
        key
      }

}

diagnoses__submitter_id @include(if: $shouldShow_diagnoses__submitter_id) {

      buckets {
        doc_count
        key
      }

}

diagnoses__tissue_or_organ_of_origin @include(if: $shouldShow_diagnoses__tissue_or_organ_of_origin) {

      buckets {
        doc_count
        key
      }

}


diagnoses__treatments__days_to_treatment @include(if: $shouldShow_diagnoses__treatments__days_to_treatment) {

      count
      max
      avg
      min

}

diagnoses__treatments__days_to_treatment_end @include(if: $shouldShow_diagnoses__treatments__days_to_treatment_end) {

      count
      max
      avg
      min

}

diagnoses__treatments__days_to_treatment_start @include(if: $shouldShow_diagnoses__treatments__days_to_treatment_start) {

      count
      max
      avg
      min

}

diagnoses__treatments__state @include(if: $shouldShow_diagnoses__treatments__state) {

      buckets {
        doc_count
        key
      }

}

diagnoses__treatments__submitter_id @include(if: $shouldShow_diagnoses__treatments__submitter_id) {

      buckets {
        doc_count
        key
      }

}

diagnoses__treatments__therapeutic_agents @include(if: $shouldShow_diagnoses__treatments__therapeutic_agents) {

      buckets {
        doc_count
        key
      }

}

diagnoses__treatments__treatment_anatomic_site @include(if: $shouldShow_diagnoses__treatments__treatment_anatomic_site) {

      buckets {
        doc_count
        key
      }

}

diagnoses__treatments__treatment_id @include(if: $shouldShow_diagnoses__treatments__treatment_id) {

      buckets {
        doc_count
        key
      }

}

diagnoses__treatments__treatment_intent_type @include(if: $shouldShow_diagnoses__treatments__treatment_intent_type) {

      buckets {
        doc_count
        key
      }

}

diagnoses__treatments__treatment_or_therapy @include(if: $shouldShow_diagnoses__treatments__treatment_or_therapy) {

      buckets {
        doc_count
        key
      }

}

diagnoses__treatments__treatment_outcome @include(if: $shouldShow_diagnoses__treatments__treatment_outcome) {

      buckets {
        doc_count
        key
      }

}

diagnoses__treatments__treatment_type @include(if: $shouldShow_diagnoses__treatments__treatment_type) {

      buckets {
        doc_count
        key
      }

}


diagnoses__tumor_grade @include(if: $shouldShow_diagnoses__tumor_grade) {

      buckets {
        doc_count
        key
      }

}

diagnoses__tumor_stage @include(if: $shouldShow_diagnoses__tumor_stage) {

      buckets {
        doc_count
        key
      }

}


diagnoses__vascular_invasion_present @include(if: $shouldShow_diagnoses__vascular_invasion_present) {

      buckets {
        doc_count
        key
      }

}

diagnoses__vital_status @include(if: $shouldShow_diagnoses__vital_status) {

      buckets {
        doc_count
        key
      }

}

diagnoses__year_of_diagnosis @include(if: $shouldShow_diagnoses__year_of_diagnosis) {

      count
      max
      avg
      min

}

disease_type @include(if: $shouldShow_disease_type) {

      buckets {
        doc_count
        key
      }

}

exposures__alcohol_history @include(if: $shouldShow_exposures__alcohol_history) {

      buckets {
        doc_count
        key
      }

}

exposures__alcohol_intensity @include(if: $shouldShow_exposures__alcohol_intensity) {

      buckets {
        doc_count
        key
      }

}

exposures__bmi @include(if: $shouldShow_exposures__bmi) {

      count
      max
      avg
      min

}

exposures__cigarettes_per_day @include(if: $shouldShow_exposures__cigarettes_per_day) {

      count
      max
      avg
      min

}


exposures__exposure_id @include(if: $shouldShow_exposures__exposure_id) {

      buckets {
        doc_count
        key
      }

}

exposures__height @include(if: $shouldShow_exposures__height) {

      count
      max
      avg
      min

}

exposures__pack_years_smoked @include(if: $shouldShow_exposures__pack_years_smoked) {

      count
      max
      avg
      min

}

exposures__state @include(if: $shouldShow_exposures__state) {

      buckets {
        doc_count
        key
      }

}

exposures__submitter_id @include(if: $shouldShow_exposures__submitter_id) {

      buckets {
        doc_count
        key
      }

}

exposures__tobacco_smoking_onset_year @include(if: $shouldShow_exposures__tobacco_smoking_onset_year) {

      count
      max
      avg
      min

}

exposures__tobacco_smoking_quit_year @include(if: $shouldShow_exposures__tobacco_smoking_quit_year) {

      count
      max
      avg
      min

}

exposures__tobacco_smoking_status @include(if: $shouldShow_exposures__tobacco_smoking_status) {

      buckets {
        doc_count
        key
      }

}


exposures__weight @include(if: $shouldShow_exposures__weight) {

      count
      max
      avg
      min

}

exposures__years_smoked @include(if: $shouldShow_exposures__years_smoked) {

      count
      max
      avg
      min

}


family_histories__family_history_id @include(if: $shouldShow_family_histories__family_history_id) {

      buckets {
        doc_count
        key
      }

}

family_histories__relationship_age_at_diagnosis @include(if: $shouldShow_family_histories__relationship_age_at_diagnosis) {

      count
      max
      avg
      min

}

family_histories__relationship_gender @include(if: $shouldShow_family_histories__relationship_gender) {

      buckets {
        doc_count
        key
      }

}

family_histories__relationship_primary_diagnosis @include(if: $shouldShow_family_histories__relationship_primary_diagnosis) {

      buckets {
        doc_count
        key
      }

}

family_histories__relationship_type @include(if: $shouldShow_family_histories__relationship_type) {

      buckets {
        doc_count
        key
      }

}

family_histories__relative_with_cancer_history @include(if: $shouldShow_family_histories__relative_with_cancer_history) {

      buckets {
        doc_count
        key
      }

}

family_histories__state @include(if: $shouldShow_family_histories__state) {

      buckets {
        doc_count
        key
      }

}

family_histories__submitter_id @include(if: $shouldShow_family_histories__submitter_id) {

      buckets {
        doc_count
        key
      }

}

portion_ids @include(if: $shouldShow_portion_ids) {

      buckets {
        doc_count
        key
      }

}

primary_site @include(if: $shouldShow_primary_site) {

      buckets {
        doc_count
        key
      }

}

project__dbgap_accession_number @include(if: $shouldShow_project__dbgap_accession_number) {

      buckets {
        doc_count
        key
      }

}

project__disease_type @include(if: $shouldShow_project__disease_type) {

      buckets {
        doc_count
        key
      }

}

project__intended_release_date @include(if: $shouldShow_project__intended_release_date) {

      buckets {
        doc_count
        key
      }

}

project__name @include(if: $shouldShow_project__name) {

      buckets {
        doc_count
        key
      }

}

project__primary_site @include(if: $shouldShow_project__primary_site) {

      buckets {
        doc_count
        key
      }

}

project__program__dbgap_accession_number @include(if: $shouldShow_project__program__dbgap_accession_number) {

      buckets {
        doc_count
        key
      }

}

project__program__name @include(if: $shouldShow_project__program__name) {

      buckets {
        doc_count
        key
      }

}

project__program__program_id @include(if: $shouldShow_project__program__program_id) {

      buckets {
        doc_count
        key
      }

}

project__project_id @include(if: $shouldShow_project__project_id) {

      buckets {
        doc_count
        key
      }

}

project__releasable @include(if: $shouldShow_project__releasable) {

      buckets {
        doc_count
        key
      }

}

project__released @include(if: $shouldShow_project__released) {

      buckets {
        doc_count
        key
      }

}

project__state @include(if: $shouldShow_project__state) {

      buckets {
        doc_count
        key
      }

}

sample_ids @include(if: $shouldShow_sample_ids) {

      buckets {
        doc_count
        key
      }

}

samples__annotations__annotation_id @include(if: $shouldShow_samples__annotations__annotation_id) {

      buckets {
        doc_count
        key
      }

}

samples__annotations__case_id @include(if: $shouldShow_samples__annotations__case_id) {

      buckets {
        doc_count
        key
      }

}

samples__annotations__case_submitter_id @include(if: $shouldShow_samples__annotations__case_submitter_id) {

      buckets {
        doc_count
        key
      }

}

samples__annotations__category @include(if: $shouldShow_samples__annotations__category) {

      buckets {
        doc_count
        key
      }

}

samples__annotations__classification @include(if: $shouldShow_samples__annotations__classification) {

      buckets {
        doc_count
        key
      }

}


samples__annotations__creator @include(if: $shouldShow_samples__annotations__creator) {

      buckets {
        doc_count
        key
      }

}

samples__annotations__entity_id @include(if: $shouldShow_samples__annotations__entity_id) {

      buckets {
        doc_count
        key
      }

}

samples__annotations__entity_submitter_id @include(if: $shouldShow_samples__annotations__entity_submitter_id) {

      buckets {
        doc_count
        key
      }

}

samples__annotations__entity_type @include(if: $shouldShow_samples__annotations__entity_type) {

      buckets {
        doc_count
        key
      }

}



samples__annotations__notes @include(if: $shouldShow_samples__annotations__notes) {

      buckets {
        doc_count
        key
      }

}

samples__annotations__state @include(if: $shouldShow_samples__annotations__state) {

      buckets {
        doc_count
        key
      }

}

samples__annotations__status @include(if: $shouldShow_samples__annotations__status) {

      buckets {
        doc_count
        key
      }

}

samples__annotations__submitter_id @include(if: $shouldShow_samples__annotations__submitter_id) {

      buckets {
        doc_count
        key
      }

}


samples__biospecimen_anatomic_site @include(if: $shouldShow_samples__biospecimen_anatomic_site) {

      buckets {
        doc_count
        key
      }

}

samples__composition @include(if: $shouldShow_samples__composition) {

      buckets {
        doc_count
        key
      }

}


samples__current_weight @include(if: $shouldShow_samples__current_weight) {

      count
      max
      avg
      min

}

samples__days_to_collection @include(if: $shouldShow_samples__days_to_collection) {

      count
      max
      avg
      min

}

samples__days_to_sample_procurement @include(if: $shouldShow_samples__days_to_sample_procurement) {

      count
      max
      avg
      min

}

samples__diagnosis_pathologically_confirmed @include(if: $shouldShow_samples__diagnosis_pathologically_confirmed) {

      buckets {
        doc_count
        key
      }

}

samples__freezing_method @include(if: $shouldShow_samples__freezing_method) {

      buckets {
        doc_count
        key
      }

}

samples__initial_weight @include(if: $shouldShow_samples__initial_weight) {

      count
      max
      avg
      min

}

samples__intermediate_dimension @include(if: $shouldShow_samples__intermediate_dimension) {

      buckets {
        doc_count
        key
      }

}

samples__is_ffpe @include(if: $shouldShow_samples__is_ffpe) {

      buckets {
        doc_count
        key
      }

}

samples__longest_dimension @include(if: $shouldShow_samples__longest_dimension) {

      buckets {
        doc_count
        key
      }

}

samples__method_of_sample_procurement @include(if: $shouldShow_samples__method_of_sample_procurement) {

      buckets {
        doc_count
        key
      }

}

samples__oct_embedded @include(if: $shouldShow_samples__oct_embedded) {

      buckets {
        doc_count
        key
      }

}

samples__pathology_report_uuid @include(if: $shouldShow_samples__pathology_report_uuid) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__a260_a280_ratio @include(if: $shouldShow_samples__portions__analytes__a260_a280_ratio) {

      count
      max
      avg
      min

}

samples__portions__analytes__aliquots__aliquot_id @include(if: $shouldShow_samples__portions__analytes__aliquots__aliquot_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__aliquot_quantity @include(if: $shouldShow_samples__portions__analytes__aliquots__aliquot_quantity) {

      count
      max
      avg
      min

}

samples__portions__analytes__aliquots__aliquot_volume @include(if: $shouldShow_samples__portions__analytes__aliquots__aliquot_volume) {

      count
      max
      avg
      min

}

samples__portions__analytes__aliquots__amount @include(if: $shouldShow_samples__portions__analytes__aliquots__amount) {

      count
      max
      avg
      min

}

samples__portions__analytes__aliquots__analyte_type @include(if: $shouldShow_samples__portions__analytes__aliquots__analyte_type) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__analyte_type_id @include(if: $shouldShow_samples__portions__analytes__aliquots__analyte_type_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__annotations__annotation_id @include(if: $shouldShow_samples__portions__analytes__aliquots__annotations__annotation_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__annotations__case_id @include(if: $shouldShow_samples__portions__analytes__aliquots__annotations__case_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__annotations__case_submitter_id @include(if: $shouldShow_samples__portions__analytes__aliquots__annotations__case_submitter_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__annotations__category @include(if: $shouldShow_samples__portions__analytes__aliquots__annotations__category) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__annotations__classification @include(if: $shouldShow_samples__portions__analytes__aliquots__annotations__classification) {

      buckets {
        doc_count
        key
      }

}


samples__portions__analytes__aliquots__annotations__creator @include(if: $shouldShow_samples__portions__analytes__aliquots__annotations__creator) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__annotations__entity_id @include(if: $shouldShow_samples__portions__analytes__aliquots__annotations__entity_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__annotations__entity_submitter_id @include(if: $shouldShow_samples__portions__analytes__aliquots__annotations__entity_submitter_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__annotations__entity_type @include(if: $shouldShow_samples__portions__analytes__aliquots__annotations__entity_type) {

      buckets {
        doc_count
        key
      }

}



samples__portions__analytes__aliquots__annotations__notes @include(if: $shouldShow_samples__portions__analytes__aliquots__annotations__notes) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__annotations__state @include(if: $shouldShow_samples__portions__analytes__aliquots__annotations__state) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__annotations__status @include(if: $shouldShow_samples__portions__analytes__aliquots__annotations__status) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__annotations__submitter_id @include(if: $shouldShow_samples__portions__analytes__aliquots__annotations__submitter_id) {

      buckets {
        doc_count
        key
      }

}


samples__portions__analytes__aliquots__center__center_id @include(if: $shouldShow_samples__portions__analytes__aliquots__center__center_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__center__center_type @include(if: $shouldShow_samples__portions__analytes__aliquots__center__center_type) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__center__code @include(if: $shouldShow_samples__portions__analytes__aliquots__center__code) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__center__name @include(if: $shouldShow_samples__portions__analytes__aliquots__center__name) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__center__namespace @include(if: $shouldShow_samples__portions__analytes__aliquots__center__namespace) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__center__short_name @include(if: $shouldShow_samples__portions__analytes__aliquots__center__short_name) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__concentration @include(if: $shouldShow_samples__portions__analytes__aliquots__concentration) {

      count
      max
      avg
      min

}


samples__portions__analytes__aliquots__source_center @include(if: $shouldShow_samples__portions__analytes__aliquots__source_center) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__state @include(if: $shouldShow_samples__portions__analytes__aliquots__state) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__aliquots__submitter_id @include(if: $shouldShow_samples__portions__analytes__aliquots__submitter_id) {

      buckets {
        doc_count
        key
      }

}


samples__portions__analytes__amount @include(if: $shouldShow_samples__portions__analytes__amount) {

      count
      max
      avg
      min

}

samples__portions__analytes__analyte_id @include(if: $shouldShow_samples__portions__analytes__analyte_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__analyte_quantity @include(if: $shouldShow_samples__portions__analytes__analyte_quantity) {

      count
      max
      avg
      min

}

samples__portions__analytes__analyte_type @include(if: $shouldShow_samples__portions__analytes__analyte_type) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__analyte_type_id @include(if: $shouldShow_samples__portions__analytes__analyte_type_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__analyte_volume @include(if: $shouldShow_samples__portions__analytes__analyte_volume) {

      count
      max
      avg
      min

}

samples__portions__analytes__annotations__annotation_id @include(if: $shouldShow_samples__portions__analytes__annotations__annotation_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__annotations__case_id @include(if: $shouldShow_samples__portions__analytes__annotations__case_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__annotations__case_submitter_id @include(if: $shouldShow_samples__portions__analytes__annotations__case_submitter_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__annotations__category @include(if: $shouldShow_samples__portions__analytes__annotations__category) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__annotations__classification @include(if: $shouldShow_samples__portions__analytes__annotations__classification) {

      buckets {
        doc_count
        key
      }

}


samples__portions__analytes__annotations__creator @include(if: $shouldShow_samples__portions__analytes__annotations__creator) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__annotations__entity_id @include(if: $shouldShow_samples__portions__analytes__annotations__entity_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__annotations__entity_submitter_id @include(if: $shouldShow_samples__portions__analytes__annotations__entity_submitter_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__annotations__entity_type @include(if: $shouldShow_samples__portions__analytes__annotations__entity_type) {

      buckets {
        doc_count
        key
      }

}



samples__portions__analytes__annotations__notes @include(if: $shouldShow_samples__portions__analytes__annotations__notes) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__annotations__state @include(if: $shouldShow_samples__portions__analytes__annotations__state) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__annotations__status @include(if: $shouldShow_samples__portions__analytes__annotations__status) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__annotations__submitter_id @include(if: $shouldShow_samples__portions__analytes__annotations__submitter_id) {

      buckets {
        doc_count
        key
      }

}


samples__portions__analytes__concentration @include(if: $shouldShow_samples__portions__analytes__concentration) {

      count
      max
      avg
      min

}


samples__portions__analytes__normal_tumor_genotype_snp_match @include(if: $shouldShow_samples__portions__analytes__normal_tumor_genotype_snp_match) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__ribosomal_rna_28s_16s_ratio @include(if: $shouldShow_samples__portions__analytes__ribosomal_rna_28s_16s_ratio) {

      count
      max
      avg
      min

}

samples__portions__analytes__spectrophotometer_method @include(if: $shouldShow_samples__portions__analytes__spectrophotometer_method) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__state @include(if: $shouldShow_samples__portions__analytes__state) {

      buckets {
        doc_count
        key
      }

}

samples__portions__analytes__submitter_id @include(if: $shouldShow_samples__portions__analytes__submitter_id) {

      buckets {
        doc_count
        key
      }

}


samples__portions__analytes__well_number @include(if: $shouldShow_samples__portions__analytes__well_number) {

      buckets {
        doc_count
        key
      }

}

samples__portions__annotations__annotation_id @include(if: $shouldShow_samples__portions__annotations__annotation_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__annotations__case_id @include(if: $shouldShow_samples__portions__annotations__case_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__annotations__case_submitter_id @include(if: $shouldShow_samples__portions__annotations__case_submitter_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__annotations__category @include(if: $shouldShow_samples__portions__annotations__category) {

      buckets {
        doc_count
        key
      }

}

samples__portions__annotations__classification @include(if: $shouldShow_samples__portions__annotations__classification) {

      buckets {
        doc_count
        key
      }

}


samples__portions__annotations__creator @include(if: $shouldShow_samples__portions__annotations__creator) {

      buckets {
        doc_count
        key
      }

}

samples__portions__annotations__entity_id @include(if: $shouldShow_samples__portions__annotations__entity_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__annotations__entity_submitter_id @include(if: $shouldShow_samples__portions__annotations__entity_submitter_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__annotations__entity_type @include(if: $shouldShow_samples__portions__annotations__entity_type) {

      buckets {
        doc_count
        key
      }

}



samples__portions__annotations__notes @include(if: $shouldShow_samples__portions__annotations__notes) {

      buckets {
        doc_count
        key
      }

}

samples__portions__annotations__state @include(if: $shouldShow_samples__portions__annotations__state) {

      buckets {
        doc_count
        key
      }

}

samples__portions__annotations__status @include(if: $shouldShow_samples__portions__annotations__status) {

      buckets {
        doc_count
        key
      }

}

samples__portions__annotations__submitter_id @include(if: $shouldShow_samples__portions__annotations__submitter_id) {

      buckets {
        doc_count
        key
      }

}


samples__portions__center__center_id @include(if: $shouldShow_samples__portions__center__center_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__center__center_type @include(if: $shouldShow_samples__portions__center__center_type) {

      buckets {
        doc_count
        key
      }

}

samples__portions__center__code @include(if: $shouldShow_samples__portions__center__code) {

      buckets {
        doc_count
        key
      }

}

samples__portions__center__name @include(if: $shouldShow_samples__portions__center__name) {

      buckets {
        doc_count
        key
      }

}

samples__portions__center__namespace @include(if: $shouldShow_samples__portions__center__namespace) {

      buckets {
        doc_count
        key
      }

}

samples__portions__center__short_name @include(if: $shouldShow_samples__portions__center__short_name) {

      buckets {
        doc_count
        key
      }

}



samples__portions__is_ffpe @include(if: $shouldShow_samples__portions__is_ffpe) {

      buckets {
        doc_count
        key
      }

}

samples__portions__portion_id @include(if: $shouldShow_samples__portions__portion_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__portion_number @include(if: $shouldShow_samples__portions__portion_number) {

      buckets {
        doc_count
        key
      }

}

samples__portions__slides__annotations__annotation_id @include(if: $shouldShow_samples__portions__slides__annotations__annotation_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__slides__annotations__case_id @include(if: $shouldShow_samples__portions__slides__annotations__case_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__slides__annotations__case_submitter_id @include(if: $shouldShow_samples__portions__slides__annotations__case_submitter_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__slides__annotations__category @include(if: $shouldShow_samples__portions__slides__annotations__category) {

      buckets {
        doc_count
        key
      }

}

samples__portions__slides__annotations__classification @include(if: $shouldShow_samples__portions__slides__annotations__classification) {

      buckets {
        doc_count
        key
      }

}


samples__portions__slides__annotations__creator @include(if: $shouldShow_samples__portions__slides__annotations__creator) {

      buckets {
        doc_count
        key
      }

}

samples__portions__slides__annotations__entity_id @include(if: $shouldShow_samples__portions__slides__annotations__entity_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__slides__annotations__entity_submitter_id @include(if: $shouldShow_samples__portions__slides__annotations__entity_submitter_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__slides__annotations__entity_type @include(if: $shouldShow_samples__portions__slides__annotations__entity_type) {

      buckets {
        doc_count
        key
      }

}



samples__portions__slides__annotations__notes @include(if: $shouldShow_samples__portions__slides__annotations__notes) {

      buckets {
        doc_count
        key
      }

}

samples__portions__slides__annotations__state @include(if: $shouldShow_samples__portions__slides__annotations__state) {

      buckets {
        doc_count
        key
      }

}

samples__portions__slides__annotations__status @include(if: $shouldShow_samples__portions__slides__annotations__status) {

      buckets {
        doc_count
        key
      }

}

samples__portions__slides__annotations__submitter_id @include(if: $shouldShow_samples__portions__slides__annotations__submitter_id) {

      buckets {
        doc_count
        key
      }

}



samples__portions__slides__number_proliferating_cells @include(if: $shouldShow_samples__portions__slides__number_proliferating_cells) {

      count
      max
      avg
      min

}

samples__portions__slides__percent_eosinophil_infiltration @include(if: $shouldShow_samples__portions__slides__percent_eosinophil_infiltration) {

      count
      max
      avg
      min

}

samples__portions__slides__percent_granulocyte_infiltration @include(if: $shouldShow_samples__portions__slides__percent_granulocyte_infiltration) {

      count
      max
      avg
      min

}

samples__portions__slides__percent_inflam_infiltration @include(if: $shouldShow_samples__portions__slides__percent_inflam_infiltration) {

      count
      max
      avg
      min

}

samples__portions__slides__percent_lymphocyte_infiltration @include(if: $shouldShow_samples__portions__slides__percent_lymphocyte_infiltration) {

      count
      max
      avg
      min

}

samples__portions__slides__percent_monocyte_infiltration @include(if: $shouldShow_samples__portions__slides__percent_monocyte_infiltration) {

      count
      max
      avg
      min

}

samples__portions__slides__percent_necrosis @include(if: $shouldShow_samples__portions__slides__percent_necrosis) {

      count
      max
      avg
      min

}

samples__portions__slides__percent_neutrophil_infiltration @include(if: $shouldShow_samples__portions__slides__percent_neutrophil_infiltration) {

      count
      max
      avg
      min

}

samples__portions__slides__percent_normal_cells @include(if: $shouldShow_samples__portions__slides__percent_normal_cells) {

      count
      max
      avg
      min

}

samples__portions__slides__percent_stromal_cells @include(if: $shouldShow_samples__portions__slides__percent_stromal_cells) {

      count
      max
      avg
      min

}

samples__portions__slides__percent_tumor_cells @include(if: $shouldShow_samples__portions__slides__percent_tumor_cells) {

      count
      max
      avg
      min

}

samples__portions__slides__percent_tumor_nuclei @include(if: $shouldShow_samples__portions__slides__percent_tumor_nuclei) {

      count
      max
      avg
      min

}

samples__portions__slides__section_location @include(if: $shouldShow_samples__portions__slides__section_location) {

      buckets {
        doc_count
        key
      }

}

samples__portions__slides__slide_id @include(if: $shouldShow_samples__portions__slides__slide_id) {

      buckets {
        doc_count
        key
      }

}

samples__portions__slides__state @include(if: $shouldShow_samples__portions__slides__state) {

      buckets {
        doc_count
        key
      }

}

samples__portions__slides__submitter_id @include(if: $shouldShow_samples__portions__slides__submitter_id) {

      buckets {
        doc_count
        key
      }

}


samples__portions__state @include(if: $shouldShow_samples__portions__state) {

      buckets {
        doc_count
        key
      }

}

samples__portions__submitter_id @include(if: $shouldShow_samples__portions__submitter_id) {

      buckets {
        doc_count
        key
      }

}


samples__portions__weight @include(if: $shouldShow_samples__portions__weight) {

      count
      max
      avg
      min

}

samples__preservation_method @include(if: $shouldShow_samples__preservation_method) {

      buckets {
        doc_count
        key
      }

}

samples__sample_id @include(if: $shouldShow_samples__sample_id) {

      buckets {
        doc_count
        key
      }

}

samples__sample_type @include(if: $shouldShow_samples__sample_type) {

      buckets {
        doc_count
        key
      }

}

samples__sample_type_id @include(if: $shouldShow_samples__sample_type_id) {

      buckets {
        doc_count
        key
      }

}

samples__shortest_dimension @include(if: $shouldShow_samples__shortest_dimension) {

      buckets {
        doc_count
        key
      }

}

samples__state @include(if: $shouldShow_samples__state) {

      buckets {
        doc_count
        key
      }

}

samples__submitter_id @include(if: $shouldShow_samples__submitter_id) {

      buckets {
        doc_count
        key
      }

}

samples__time_between_clamping_and_freezing @include(if: $shouldShow_samples__time_between_clamping_and_freezing) {

      buckets {
        doc_count
        key
      }

}

samples__time_between_excision_and_freezing @include(if: $shouldShow_samples__time_between_excision_and_freezing) {

      buckets {
        doc_count
        key
      }

}

samples__tissue_type @include(if: $shouldShow_samples__tissue_type) {

      buckets {
        doc_count
        key
      }

}

samples__tumor_code @include(if: $shouldShow_samples__tumor_code) {

      buckets {
        doc_count
        key
      }

}

samples__tumor_code_id @include(if: $shouldShow_samples__tumor_code_id) {

      buckets {
        doc_count
        key
      }

}

samples__tumor_descriptor @include(if: $shouldShow_samples__tumor_descriptor) {

      buckets {
        doc_count
        key
      }

}


slide_ids @include(if: $shouldShow_slide_ids) {

      buckets {
        doc_count
        key
      }

}

state @include(if: $shouldShow_state) {

      buckets {
        doc_count
        key
      }

}

submitter_aliquot_ids @include(if: $shouldShow_submitter_aliquot_ids) {

      buckets {
        doc_count
        key
      }

}

submitter_analyte_ids @include(if: $shouldShow_submitter_analyte_ids) {

      buckets {
        doc_count
        key
      }

}

submitter_id @include(if: $shouldShow_submitter_id) {

      buckets {
        doc_count
        key
      }

}

submitter_portion_ids @include(if: $shouldShow_submitter_portion_ids) {

      buckets {
        doc_count
        key
      }

}

submitter_sample_ids @include(if: $shouldShow_submitter_sample_ids) {

      buckets {
        doc_count
        key
      }

}

submitter_slide_ids @include(if: $shouldShow_submitter_slide_ids) {

      buckets {
        doc_count
        key
      }

}

summary__data_categories__data_category @include(if: $shouldShow_summary__data_categories__data_category) {

      buckets {
        doc_count
        key
      }

}

summary__data_categories__file_count @include(if: $shouldShow_summary__data_categories__file_count) {

      count
      max
      avg
      min

}

summary__experimental_strategies__experimental_strategy @include(if: $shouldShow_summary__experimental_strategies__experimental_strategy) {

      buckets {
        doc_count
        key
      }

}

summary__experimental_strategies__file_count @include(if: $shouldShow_summary__experimental_strategies__file_count) {

      count
      max
      avg
      min

}

summary__file_count @include(if: $shouldShow_summary__file_count) {

      count
      max
      avg
      min

}

summary__file_size @include(if: $shouldShow_summary__file_size) {

      count
      max
      avg
      min

}

tissue_source_site__bcr_id @include(if: $shouldShow_tissue_source_site__bcr_id) {

      buckets {
        doc_count
        key
      }

}

tissue_source_site__code @include(if: $shouldShow_tissue_source_site__code) {

      buckets {
        doc_count
        key
      }

}

tissue_source_site__name @include(if: $shouldShow_tissue_source_site__name) {

      buckets {
        doc_count
        key
      }

}

tissue_source_site__project @include(if: $shouldShow_tissue_source_site__project) {

      buckets {
        doc_count
        key
      }

}

tissue_source_site__tissue_source_site_id @include(if: $shouldShow_tissue_source_site__tissue_source_site_id) {

      buckets {
        doc_count
        key
      }

}

  }
`;
