// @ncigdc/utils/generated-relay-query-parts/repositoryCaseAggregations.js
/* @flow */
/* eslint max-len:0, comma-dangle:0 */

import Relay from 'react-relay';

export const initialCaseAggregationsVariables = {
  shouldShow_case_id: false,
  shouldShow_created_datetime: false,
  shouldShow_demographic__created_datetime: false,
  shouldShow_demographic__demographic_id: false,
  shouldShow_demographic__ethnicity: false,
  shouldShow_demographic__gender: false,
  shouldShow_demographic__race: false,
  shouldShow_demographic__updated_datetime: false,
  shouldShow_demographic__year_of_birth: false,
  shouldShow_demographic__year_of_death: false,
  shouldShow_diagnoses__age_at_diagnosis: false,
  shouldShow_diagnoses__classification_of_tumor: false,
  shouldShow_diagnoses__created_datetime: false,
  shouldShow_diagnoses__days_to_birth: false,
  shouldShow_diagnoses__days_to_death: false,
  shouldShow_diagnoses__days_to_last_follow_up: false,
  shouldShow_diagnoses__days_to_last_known_disease_status: false,
  shouldShow_diagnoses__days_to_recurrence: false,
  shouldShow_diagnoses__diagnosis_id: false,
  shouldShow_diagnoses__last_known_disease_status: false,
  shouldShow_diagnoses__morphology: false,
  shouldShow_diagnoses__primary_diagnosis: false,
  shouldShow_diagnoses__prior_malignancy: false,
  shouldShow_diagnoses__progression_or_recurrence: false,
  shouldShow_diagnoses__site_of_resection_or_biopsy: false,
  shouldShow_diagnoses__tissue_or_organ_of_origin: false,
  shouldShow_diagnoses__treatments__created_datetime: false,
  shouldShow_diagnoses__treatments__days_to_treatment: false,
  shouldShow_diagnoses__treatments__therapeutic_agents: false,
  shouldShow_diagnoses__treatments__treatment_intent_type: false,
  shouldShow_diagnoses__treatments__treatment_or_therapy: false,
  shouldShow_diagnoses__treatments__updated_datetime: false,
  shouldShow_diagnoses__tumor_grade: false,
  shouldShow_diagnoses__tumor_stage: false,
  shouldShow_diagnoses__updated_datetime: false,
  shouldShow_diagnoses__vital_status: false,
  shouldShow_exposures__alcohol_history: false,
  shouldShow_exposures__alcohol_intensity: false,
  shouldShow_exposures__bmi: false,
  shouldShow_exposures__cigarettes_per_day: false,
  shouldShow_exposures__created_datetime: false,
  shouldShow_exposures__exposure_id: false,
  shouldShow_exposures__height: false,
  shouldShow_exposures__submitter_id: false,
  shouldShow_exposures__updated_datetime: false,
  shouldShow_exposures__weight: false,
  shouldShow_exposures__years_smoked: false,
  shouldShow_family_histories__created_datetime: false,
  shouldShow_family_histories__family_history_id: false,
  shouldShow_family_histories__relationship_age_at_diagnosis: false,
  shouldShow_family_histories__relationship_gender: false,
  shouldShow_family_histories__relationship_primary_diagnosis: false,
  shouldShow_family_histories__relationship_type: false,
  shouldShow_family_histories__relative_with_cancer_history: false,
  shouldShow_family_histories__submitter_id: false,
  shouldShow_family_histories__updated_datetime: false,
  shouldShow_project__dbgap_accession_number: false,
  shouldShow_project__disease_type: false,
  shouldShow_project__name: false,
  shouldShow_project__primary_site: false,
  shouldShow_project__program__dbgap_accession_number: false,
  shouldShow_project__program__name: false,
  shouldShow_project__program__program_id: false,
  shouldShow_project__project_id: false,
  shouldShow_project__released: false,
  shouldShow_samples__composition: false,
  shouldShow_samples__created_datetime: false,
  shouldShow_samples__current_weight: false,
  shouldShow_samples__days_to_collection: false,
  shouldShow_samples__days_to_sample_procurement: false,
  shouldShow_samples__freezing_method: false,
  shouldShow_samples__initial_weight: false,
  shouldShow_samples__intermediate_dimension: false,
  shouldShow_samples__is_ffpe: false,
  shouldShow_samples__longest_dimension: false,
  shouldShow_samples__oct_embedded: false,
  shouldShow_samples__pathology_report_uuid: false,
  shouldShow_samples__portions__analytes__a260_a280_ratio: false,
  shouldShow_samples__portions__analytes__aliquots__aliquot_id: false,
  shouldShow_samples__portions__analytes__aliquots__amount: false,
  shouldShow_samples__portions__analytes__aliquots__analyte_type: false,
  shouldShow_samples__portions__analytes__aliquots__analyte_type_id: false,
  shouldShow_samples__portions__analytes__aliquots__center__center_id: false,
  shouldShow_samples__portions__analytes__aliquots__center__center_type: false,
  shouldShow_samples__portions__analytes__aliquots__center__code: false,
  shouldShow_samples__portions__analytes__aliquots__center__name: false,
  shouldShow_samples__portions__analytes__aliquots__center__namespace: false,
  shouldShow_samples__portions__analytes__aliquots__center__short_name: false,
  shouldShow_samples__portions__analytes__aliquots__concentration: false,
  shouldShow_samples__portions__analytes__aliquots__created_datetime: false,
  shouldShow_samples__portions__analytes__aliquots__source_center: false,
  shouldShow_samples__portions__analytes__aliquots__submitter_id: false,
  shouldShow_samples__portions__analytes__aliquots__updated_datetime: false,
  shouldShow_samples__portions__analytes__amount: false,
  shouldShow_samples__portions__analytes__analyte_id: false,
  shouldShow_samples__portions__analytes__analyte_type: false,
  shouldShow_samples__portions__analytes__analyte_type_id: false,
  shouldShow_samples__portions__analytes__concentration: false,
  shouldShow_samples__portions__analytes__created_datetime: false,
  shouldShow_samples__portions__analytes__spectrophotometer_method: false,
  shouldShow_samples__portions__analytes__submitter_id: false,
  shouldShow_samples__portions__analytes__updated_datetime: false,
  shouldShow_samples__portions__analytes__well_number: false,
  shouldShow_samples__portions__center__center_id: false,
  shouldShow_samples__portions__center__center_type: false,
  shouldShow_samples__portions__center__code: false,
  shouldShow_samples__portions__center__name: false,
  shouldShow_samples__portions__center__namespace: false,
  shouldShow_samples__portions__center__short_name: false,
  shouldShow_samples__portions__created_datetime: false,
  shouldShow_samples__portions__creation_datetime: false,
  shouldShow_samples__portions__is_ffpe: false,
  shouldShow_samples__portions__portion_id: false,
  shouldShow_samples__portions__portion_number: false,
  shouldShow_samples__portions__slides__created_datetime: false,
  shouldShow_samples__portions__slides__number_proliferating_cells: false,
  shouldShow_samples__portions__slides__percent_eosinophil_infiltration: false,
  shouldShow_samples__portions__slides__percent_granulocyte_infiltration: false,
  shouldShow_samples__portions__slides__percent_inflam_infiltration: false,
  shouldShow_samples__portions__slides__percent_lymphocyte_infiltration: false,
  shouldShow_samples__portions__slides__percent_monocyte_infiltration: false,
  shouldShow_samples__portions__slides__percent_necrosis: false,
  shouldShow_samples__portions__slides__percent_neutrophil_infiltration: false,
  shouldShow_samples__portions__slides__percent_normal_cells: false,
  shouldShow_samples__portions__slides__percent_stromal_cells: false,
  shouldShow_samples__portions__slides__percent_tumor_cells: false,
  shouldShow_samples__portions__slides__percent_tumor_nuclei: false,
  shouldShow_samples__portions__slides__section_location: false,
  shouldShow_samples__portions__slides__slide_id: false,
  shouldShow_samples__portions__slides__submitter_id: false,
  shouldShow_samples__portions__slides__updated_datetime: false,
  shouldShow_samples__portions__submitter_id: false,
  shouldShow_samples__portions__updated_datetime: false,
  shouldShow_samples__portions__weight: false,
  shouldShow_samples__preservation_method: false,
  shouldShow_samples__sample_id: false,
  shouldShow_samples__sample_type: false,
  shouldShow_samples__sample_type_id: false,
  shouldShow_samples__shortest_dimension: false,
  shouldShow_samples__submitter_id: false,
  shouldShow_samples__time_between_clamping_and_freezing: false,
  shouldShow_samples__time_between_excision_and_freezing: false,
  shouldShow_samples__tissue_type: false,
  shouldShow_samples__tumor_code: false,
  shouldShow_samples__tumor_code_id: false,
  shouldShow_samples__tumor_descriptor: false,
  shouldShow_samples__updated_datetime: false,
  shouldShow_submitter_id: false,
  shouldShow_updated_datetime: false
};

export const repositoryCaseAggregationsFragment = () => Relay.QL`
  fragment on CaseAggregations {
    
case_id @include(if: $shouldShow_case_id) {
  
      buckets {
        doc_count
        key
      }
      
}



demographic__demographic_id @include(if: $shouldShow_demographic__demographic_id) {
  
      buckets {
        doc_count
        key
      }
      
}

demographic__ethnicity @include(if: $shouldShow_demographic__ethnicity) {
  
      buckets {
        doc_count
        key
      }
      
}

demographic__gender @include(if: $shouldShow_demographic__gender) {
  
      buckets {
        doc_count
        key
      }
      
}

demographic__race @include(if: $shouldShow_demographic__race) {
  
      buckets {
        doc_count
        key
      }
      
}


demographic__year_of_birth @include(if: $shouldShow_demographic__year_of_birth) {
  
      count
      max
      avg
      min
    
}

demographic__year_of_death @include(if: $shouldShow_demographic__year_of_death) {
  
      count
      max
      avg
      min
    
}

diagnoses__age_at_diagnosis @include(if: $shouldShow_diagnoses__age_at_diagnosis) {
  
      count
      max
      avg
      min
    
}

diagnoses__classification_of_tumor @include(if: $shouldShow_diagnoses__classification_of_tumor) {
  
      buckets {
        doc_count
        key
      }
      
}


diagnoses__days_to_birth @include(if: $shouldShow_diagnoses__days_to_birth) {
  
      count
      max
      avg
      min
    
}

diagnoses__days_to_death @include(if: $shouldShow_diagnoses__days_to_death) {
  
      count
      max
      avg
      min
    
}

diagnoses__days_to_last_follow_up @include(if: $shouldShow_diagnoses__days_to_last_follow_up) {
  
      count
      max
      avg
      min
    
}

diagnoses__days_to_last_known_disease_status @include(if: $shouldShow_diagnoses__days_to_last_known_disease_status) {
  
      count
      max
      avg
      min
    
}

diagnoses__days_to_recurrence @include(if: $shouldShow_diagnoses__days_to_recurrence) {
  
      count
      max
      avg
      min
    
}

diagnoses__diagnosis_id @include(if: $shouldShow_diagnoses__diagnosis_id) {
  
      buckets {
        doc_count
        key
      }
      
}

diagnoses__last_known_disease_status @include(if: $shouldShow_diagnoses__last_known_disease_status) {
  
      buckets {
        doc_count
        key
      }
      
}

diagnoses__morphology @include(if: $shouldShow_diagnoses__morphology) {
  
      buckets {
        doc_count
        key
      }
      
}

diagnoses__primary_diagnosis @include(if: $shouldShow_diagnoses__primary_diagnosis) {
  
      buckets {
        doc_count
        key
      }
      
}

diagnoses__prior_malignancy @include(if: $shouldShow_diagnoses__prior_malignancy) {
  
      buckets {
        doc_count
        key
      }
      
}

diagnoses__progression_or_recurrence @include(if: $shouldShow_diagnoses__progression_or_recurrence) {
  
      buckets {
        doc_count
        key
      }
      
}

diagnoses__site_of_resection_or_biopsy @include(if: $shouldShow_diagnoses__site_of_resection_or_biopsy) {
  
      buckets {
        doc_count
        key
      }
      
}

diagnoses__tissue_or_organ_of_origin @include(if: $shouldShow_diagnoses__tissue_or_organ_of_origin) {
  
      buckets {
        doc_count
        key
      }
      
}


diagnoses__treatments__days_to_treatment @include(if: $shouldShow_diagnoses__treatments__days_to_treatment) {
  
      count
      max
      avg
      min
    
}

diagnoses__treatments__therapeutic_agents @include(if: $shouldShow_diagnoses__treatments__therapeutic_agents) {
  
      buckets {
        doc_count
        key
      }
      
}

diagnoses__treatments__treatment_intent_type @include(if: $shouldShow_diagnoses__treatments__treatment_intent_type) {
  
      buckets {
        doc_count
        key
      }
      
}

diagnoses__treatments__treatment_or_therapy @include(if: $shouldShow_diagnoses__treatments__treatment_or_therapy) {
  
      buckets {
        doc_count
        key
      }
      
}


diagnoses__tumor_grade @include(if: $shouldShow_diagnoses__tumor_grade) {
  
      buckets {
        doc_count
        key
      }
      
}

diagnoses__tumor_stage @include(if: $shouldShow_diagnoses__tumor_stage) {
  
      buckets {
        doc_count
        key
      }
      
}


diagnoses__vital_status @include(if: $shouldShow_diagnoses__vital_status) {
  
      buckets {
        doc_count
        key
      }
      
}

exposures__alcohol_history @include(if: $shouldShow_exposures__alcohol_history) {
  
      buckets {
        doc_count
        key
      }
      
}

exposures__alcohol_intensity @include(if: $shouldShow_exposures__alcohol_intensity) {
  
      buckets {
        doc_count
        key
      }
      
}

exposures__bmi @include(if: $shouldShow_exposures__bmi) {
  
      count
      max
      avg
      min
    
}

exposures__cigarettes_per_day @include(if: $shouldShow_exposures__cigarettes_per_day) {
  
      count
      max
      avg
      min
    
}


exposures__exposure_id @include(if: $shouldShow_exposures__exposure_id) {
  
      buckets {
        doc_count
        key
      }
      
}

exposures__height @include(if: $shouldShow_exposures__height) {
  
      count
      max
      avg
      min
    
}

exposures__submitter_id @include(if: $shouldShow_exposures__submitter_id) {
  
      buckets {
        doc_count
        key
      }
      
}


exposures__weight @include(if: $shouldShow_exposures__weight) {
  
      count
      max
      avg
      min
    
}

exposures__years_smoked @include(if: $shouldShow_exposures__years_smoked) {
  
      count
      max
      avg
      min
    
}


family_histories__family_history_id @include(if: $shouldShow_family_histories__family_history_id) {
  
      buckets {
        doc_count
        key
      }
      
}

family_histories__relationship_age_at_diagnosis @include(if: $shouldShow_family_histories__relationship_age_at_diagnosis) {
  
      count
      max
      avg
      min
    
}

family_histories__relationship_gender @include(if: $shouldShow_family_histories__relationship_gender) {
  
      buckets {
        doc_count
        key
      }
      
}

family_histories__relationship_primary_diagnosis @include(if: $shouldShow_family_histories__relationship_primary_diagnosis) {
  
      buckets {
        doc_count
        key
      }
      
}

family_histories__relationship_type @include(if: $shouldShow_family_histories__relationship_type) {
  
      buckets {
        doc_count
        key
      }
      
}

family_histories__relative_with_cancer_history @include(if: $shouldShow_family_histories__relative_with_cancer_history) {
  
      buckets {
        doc_count
        key
      }
      
}

family_histories__submitter_id @include(if: $shouldShow_family_histories__submitter_id) {
  
      buckets {
        doc_count
        key
      }
      
}


project__dbgap_accession_number @include(if: $shouldShow_project__dbgap_accession_number) {
  
      buckets {
        doc_count
        key
      }
      
}

project__disease_type @include(if: $shouldShow_project__disease_type) {
  
      buckets {
        doc_count
        key
      }
      
}

project__name @include(if: $shouldShow_project__name) {
  
      buckets {
        doc_count
        key
      }
      
}

project__primary_site @include(if: $shouldShow_project__primary_site) {
  
      buckets {
        doc_count
        key
      }
      
}

project__program__dbgap_accession_number @include(if: $shouldShow_project__program__dbgap_accession_number) {
  
      buckets {
        doc_count
        key
      }
      
}

project__program__name @include(if: $shouldShow_project__program__name) {
  
      buckets {
        doc_count
        key
      }
      
}

project__program__program_id @include(if: $shouldShow_project__program__program_id) {
  
      buckets {
        doc_count
        key
      }
      
}

project__project_id @include(if: $shouldShow_project__project_id) {
  
      buckets {
        doc_count
        key
      }
      
}

project__released @include(if: $shouldShow_project__released) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__composition @include(if: $shouldShow_samples__composition) {
  
      buckets {
        doc_count
        key
      }
      
}


samples__current_weight @include(if: $shouldShow_samples__current_weight) {
  
      count
      max
      avg
      min
    
}

samples__days_to_collection @include(if: $shouldShow_samples__days_to_collection) {
  
      count
      max
      avg
      min
    
}

samples__days_to_sample_procurement @include(if: $shouldShow_samples__days_to_sample_procurement) {
  
      count
      max
      avg
      min
    
}

samples__freezing_method @include(if: $shouldShow_samples__freezing_method) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__initial_weight @include(if: $shouldShow_samples__initial_weight) {
  
      count
      max
      avg
      min
    
}

samples__intermediate_dimension @include(if: $shouldShow_samples__intermediate_dimension) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__is_ffpe @include(if: $shouldShow_samples__is_ffpe) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__longest_dimension @include(if: $shouldShow_samples__longest_dimension) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__oct_embedded @include(if: $shouldShow_samples__oct_embedded) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__pathology_report_uuid @include(if: $shouldShow_samples__pathology_report_uuid) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__analytes__a260_a280_ratio @include(if: $shouldShow_samples__portions__analytes__a260_a280_ratio) {
  
      count
      max
      avg
      min
    
}

samples__portions__analytes__aliquots__aliquot_id @include(if: $shouldShow_samples__portions__analytes__aliquots__aliquot_id) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__analytes__aliquots__amount @include(if: $shouldShow_samples__portions__analytes__aliquots__amount) {
  
      count
      max
      avg
      min
    
}

samples__portions__analytes__aliquots__analyte_type @include(if: $shouldShow_samples__portions__analytes__aliquots__analyte_type) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__analytes__aliquots__analyte_type_id @include(if: $shouldShow_samples__portions__analytes__aliquots__analyte_type_id) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__analytes__aliquots__center__center_id @include(if: $shouldShow_samples__portions__analytes__aliquots__center__center_id) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__analytes__aliquots__center__center_type @include(if: $shouldShow_samples__portions__analytes__aliquots__center__center_type) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__analytes__aliquots__center__code @include(if: $shouldShow_samples__portions__analytes__aliquots__center__code) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__analytes__aliquots__center__name @include(if: $shouldShow_samples__portions__analytes__aliquots__center__name) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__analytes__aliquots__center__namespace @include(if: $shouldShow_samples__portions__analytes__aliquots__center__namespace) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__analytes__aliquots__center__short_name @include(if: $shouldShow_samples__portions__analytes__aliquots__center__short_name) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__analytes__aliquots__concentration @include(if: $shouldShow_samples__portions__analytes__aliquots__concentration) {
  
      count
      max
      avg
      min
    
}


samples__portions__analytes__aliquots__source_center @include(if: $shouldShow_samples__portions__analytes__aliquots__source_center) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__analytes__aliquots__submitter_id @include(if: $shouldShow_samples__portions__analytes__aliquots__submitter_id) {
  
      buckets {
        doc_count
        key
      }
      
}


samples__portions__analytes__amount @include(if: $shouldShow_samples__portions__analytes__amount) {
  
      count
      max
      avg
      min
    
}

samples__portions__analytes__analyte_id @include(if: $shouldShow_samples__portions__analytes__analyte_id) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__analytes__analyte_type @include(if: $shouldShow_samples__portions__analytes__analyte_type) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__analytes__analyte_type_id @include(if: $shouldShow_samples__portions__analytes__analyte_type_id) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__analytes__concentration @include(if: $shouldShow_samples__portions__analytes__concentration) {
  
      count
      max
      avg
      min
    
}


samples__portions__analytes__spectrophotometer_method @include(if: $shouldShow_samples__portions__analytes__spectrophotometer_method) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__analytes__submitter_id @include(if: $shouldShow_samples__portions__analytes__submitter_id) {
  
      buckets {
        doc_count
        key
      }
      
}


samples__portions__analytes__well_number @include(if: $shouldShow_samples__portions__analytes__well_number) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__center__center_id @include(if: $shouldShow_samples__portions__center__center_id) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__center__center_type @include(if: $shouldShow_samples__portions__center__center_type) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__center__code @include(if: $shouldShow_samples__portions__center__code) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__center__name @include(if: $shouldShow_samples__portions__center__name) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__center__namespace @include(if: $shouldShow_samples__portions__center__namespace) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__center__short_name @include(if: $shouldShow_samples__portions__center__short_name) {
  
      buckets {
        doc_count
        key
      }
      
}



samples__portions__is_ffpe @include(if: $shouldShow_samples__portions__is_ffpe) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__portion_id @include(if: $shouldShow_samples__portions__portion_id) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__portion_number @include(if: $shouldShow_samples__portions__portion_number) {
  
      buckets {
        doc_count
        key
      }
      
}


samples__portions__slides__number_proliferating_cells @include(if: $shouldShow_samples__portions__slides__number_proliferating_cells) {
  
      count
      max
      avg
      min
    
}

samples__portions__slides__percent_eosinophil_infiltration @include(if: $shouldShow_samples__portions__slides__percent_eosinophil_infiltration) {
  
      count
      max
      avg
      min
    
}

samples__portions__slides__percent_granulocyte_infiltration @include(if: $shouldShow_samples__portions__slides__percent_granulocyte_infiltration) {
  
      count
      max
      avg
      min
    
}

samples__portions__slides__percent_inflam_infiltration @include(if: $shouldShow_samples__portions__slides__percent_inflam_infiltration) {
  
      count
      max
      avg
      min
    
}

samples__portions__slides__percent_lymphocyte_infiltration @include(if: $shouldShow_samples__portions__slides__percent_lymphocyte_infiltration) {
  
      count
      max
      avg
      min
    
}

samples__portions__slides__percent_monocyte_infiltration @include(if: $shouldShow_samples__portions__slides__percent_monocyte_infiltration) {
  
      count
      max
      avg
      min
    
}

samples__portions__slides__percent_necrosis @include(if: $shouldShow_samples__portions__slides__percent_necrosis) {
  
      count
      max
      avg
      min
    
}

samples__portions__slides__percent_neutrophil_infiltration @include(if: $shouldShow_samples__portions__slides__percent_neutrophil_infiltration) {
  
      count
      max
      avg
      min
    
}

samples__portions__slides__percent_normal_cells @include(if: $shouldShow_samples__portions__slides__percent_normal_cells) {
  
      count
      max
      avg
      min
    
}

samples__portions__slides__percent_stromal_cells @include(if: $shouldShow_samples__portions__slides__percent_stromal_cells) {
  
      count
      max
      avg
      min
    
}

samples__portions__slides__percent_tumor_cells @include(if: $shouldShow_samples__portions__slides__percent_tumor_cells) {
  
      count
      max
      avg
      min
    
}

samples__portions__slides__percent_tumor_nuclei @include(if: $shouldShow_samples__portions__slides__percent_tumor_nuclei) {
  
      count
      max
      avg
      min
    
}

samples__portions__slides__section_location @include(if: $shouldShow_samples__portions__slides__section_location) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__slides__slide_id @include(if: $shouldShow_samples__portions__slides__slide_id) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__portions__slides__submitter_id @include(if: $shouldShow_samples__portions__slides__submitter_id) {
  
      buckets {
        doc_count
        key
      }
      
}


samples__portions__submitter_id @include(if: $shouldShow_samples__portions__submitter_id) {
  
      buckets {
        doc_count
        key
      }
      
}


samples__portions__weight @include(if: $shouldShow_samples__portions__weight) {
  
      count
      max
      avg
      min
    
}

samples__preservation_method @include(if: $shouldShow_samples__preservation_method) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__sample_id @include(if: $shouldShow_samples__sample_id) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__sample_type @include(if: $shouldShow_samples__sample_type) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__sample_type_id @include(if: $shouldShow_samples__sample_type_id) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__shortest_dimension @include(if: $shouldShow_samples__shortest_dimension) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__submitter_id @include(if: $shouldShow_samples__submitter_id) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__time_between_clamping_and_freezing @include(if: $shouldShow_samples__time_between_clamping_and_freezing) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__time_between_excision_and_freezing @include(if: $shouldShow_samples__time_between_excision_and_freezing) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__tissue_type @include(if: $shouldShow_samples__tissue_type) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__tumor_code @include(if: $shouldShow_samples__tumor_code) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__tumor_code_id @include(if: $shouldShow_samples__tumor_code_id) {
  
      buckets {
        doc_count
        key
      }
      
}

samples__tumor_descriptor @include(if: $shouldShow_samples__tumor_descriptor) {
  
      buckets {
        doc_count
        key
      }
      
}


submitter_id @include(if: $shouldShow_submitter_id) {
  
      buckets {
        doc_count
        key
      }
      
}

  }
`;
