// @ncigdc/utils/generated-relay-query-parts/repositoryFileAggregations.js
/* @flow */
/* eslint max-len:0, comma-dangle:0 */

import Relay from 'react-relay';

export const initialFileAggregationsVariables = {
  shouldShow_access: false,
  shouldShow_analysis__analysis_id: false,
  shouldShow_analysis__analysis_type: false,
  shouldShow_analysis__created_datetime: false,
  shouldShow_analysis__input_files__access: false,
  shouldShow_analysis__input_files__created_datetime: false,
  shouldShow_analysis__input_files__data_category: false,
  shouldShow_analysis__input_files__data_format: false,
  shouldShow_analysis__input_files__data_type: false,
  shouldShow_analysis__input_files__experimental_strategy: false,
  shouldShow_analysis__input_files__file_id: false,
  shouldShow_analysis__input_files__file_name: false,
  shouldShow_analysis__input_files__file_size: false,
  shouldShow_analysis__input_files__file_state: false,
  shouldShow_analysis__input_files__md5sum: false,
  shouldShow_analysis__input_files__platform: false,
  shouldShow_analysis__input_files__revision: false,
  shouldShow_analysis__input_files__updated_datetime: false,
  shouldShow_analysis__metadata__read_groups__RIN: false,
  shouldShow_analysis__metadata__read_groups__adapter_name: false,
  shouldShow_analysis__metadata__read_groups__adapter_sequence: false,
  shouldShow_analysis__metadata__read_groups__base_caller_name: false,
  shouldShow_analysis__metadata__read_groups__base_caller_version: false,
  shouldShow_analysis__metadata__read_groups__created_datetime: false,
  shouldShow_analysis__metadata__read_groups__experiment_name: false,
  shouldShow_analysis__metadata__read_groups__flow_cell_barcode: false,
  shouldShow_analysis__metadata__read_groups__includes_spike_ins: false,
  shouldShow_analysis__metadata__read_groups__instrument_model: false,
  shouldShow_analysis__metadata__read_groups__is_paired_end: false,
  shouldShow_analysis__metadata__read_groups__library_name: false,
  shouldShow_analysis__metadata__read_groups__library_preparation_kit_catalog_number: false,
  shouldShow_analysis__metadata__read_groups__library_preparation_kit_name: false,
  shouldShow_analysis__metadata__read_groups__library_preparation_kit_vendor: false,
  shouldShow_analysis__metadata__read_groups__library_preparation_kit_version: false,
  shouldShow_analysis__metadata__read_groups__library_selection: false,
  shouldShow_analysis__metadata__read_groups__library_strand: false,
  shouldShow_analysis__metadata__read_groups__library_strategy: false,
  shouldShow_analysis__metadata__read_groups__platform: false,
  shouldShow_analysis__metadata__read_groups__read_group_id: false,
  shouldShow_analysis__metadata__read_groups__read_group_name: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__adapter_content: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__basic_statistics: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__created_datetime: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__encoding: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__fastq_name: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__kmer_content: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__overrepresented_sequences: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__per_base_n_content: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__per_base_sequence_content: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__per_base_sequence_quality: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__per_sequence_gc_content: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__per_sequence_quality_score: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__per_tile_sequence_quality: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__percent_gc_content: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__read_group_qc_id: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__sequence_duplication_levels: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__sequence_length_distribution: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__state: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__submitter_id: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__total_sequences: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__updated_datetime: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__workflow_end_datetime: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__workflow_link: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__workflow_start_datetime: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__workflow_type: false,
  shouldShow_analysis__metadata__read_groups__read_group_qcs__workflow_version: false,
  shouldShow_analysis__metadata__read_groups__read_length: false,
  shouldShow_analysis__metadata__read_groups__sequencing_center: false,
  shouldShow_analysis__metadata__read_groups__sequencing_date: false,
  shouldShow_analysis__metadata__read_groups__size_selection_range: false,
  shouldShow_analysis__metadata__read_groups__spike_ins_concentration: false,
  shouldShow_analysis__metadata__read_groups__spike_ins_fasta: false,
  shouldShow_analysis__metadata__read_groups__submitter_id: false,
  shouldShow_analysis__metadata__read_groups__target_capture_kit_catalog_number: false,
  shouldShow_analysis__metadata__read_groups__target_capture_kit_name: false,
  shouldShow_analysis__metadata__read_groups__target_capture_kit_target_region: false,
  shouldShow_analysis__metadata__read_groups__target_capture_kit_vendor: false,
  shouldShow_analysis__metadata__read_groups__target_capture_kit_version: false,
  shouldShow_analysis__metadata__read_groups__to_trim_adapter_sequence: false,
  shouldShow_analysis__metadata__read_groups__updated_datetime: false,
  shouldShow_analysis__updated_datetime: false,
  shouldShow_analysis__workflow_end_datetime: false,
  shouldShow_analysis__workflow_link: false,
  shouldShow_analysis__workflow_start_datetime: false,
  shouldShow_analysis__workflow_type: false,
  shouldShow_analysis__workflow_version: false,
  shouldShow_archive__created_datetime: false,
  shouldShow_archive__data_category: false,
  shouldShow_archive__data_format: false,
  shouldShow_archive__data_type: false,
  shouldShow_archive__error_type: false,
  shouldShow_archive__file_name: false,
  shouldShow_archive__file_size: false,
  shouldShow_archive__file_state: false,
  shouldShow_archive__md5sum: false,
  shouldShow_archive__revision: false,
  shouldShow_archive__state: false,
  shouldShow_archive__state_comment: false,
  shouldShow_archive__submitter_id: false,
  shouldShow_archive__updated_datetime: false,
  shouldShow_center__center_id: false,
  shouldShow_center__center_type: false,
  shouldShow_center__code: false,
  shouldShow_center__name: false,
  shouldShow_center__namespace: false,
  shouldShow_center__short_name: false,
  shouldShow_created_datetime: false,
  shouldShow_data_category: false,
  shouldShow_data_format: false,
  shouldShow_data_type: false,
  shouldShow_downstream_analyses__analysis_id: false,
  shouldShow_downstream_analyses__analysis_type: false,
  shouldShow_downstream_analyses__created_datetime: false,
  shouldShow_downstream_analyses__output_files__access: false,
  shouldShow_downstream_analyses__output_files__created_datetime: false,
  shouldShow_downstream_analyses__output_files__data_category: false,
  shouldShow_downstream_analyses__output_files__data_format: false,
  shouldShow_downstream_analyses__output_files__data_type: false,
  shouldShow_downstream_analyses__output_files__error_type: false,
  shouldShow_downstream_analyses__output_files__experimental_strategy: false,
  shouldShow_downstream_analyses__output_files__file_id: false,
  shouldShow_downstream_analyses__output_files__file_name: false,
  shouldShow_downstream_analyses__output_files__file_size: false,
  shouldShow_downstream_analyses__output_files__file_state: false,
  shouldShow_downstream_analyses__output_files__md5sum: false,
  shouldShow_downstream_analyses__output_files__platform: false,
  shouldShow_downstream_analyses__output_files__revision: false,
  shouldShow_downstream_analyses__output_files__state: false,
  shouldShow_downstream_analyses__output_files__state_comment: false,
  shouldShow_downstream_analyses__output_files__submitter_id: false,
  shouldShow_downstream_analyses__output_files__updated_datetime: false,
  shouldShow_downstream_analyses__submitter_id: false,
  shouldShow_downstream_analyses__updated_datetime: false,
  shouldShow_downstream_analyses__workflow_end_datetime: false,
  shouldShow_downstream_analyses__workflow_link: false,
  shouldShow_downstream_analyses__workflow_start_datetime: false,
  shouldShow_downstream_analyses__workflow_type: false,
  shouldShow_downstream_analyses__workflow_version: false,
  shouldShow_experimental_strategy: false,
  shouldShow_file_id: false,
  shouldShow_file_name: false,
  shouldShow_file_size: false,
  shouldShow_file_state: false,
  shouldShow_index_files__access: false,
  shouldShow_index_files__created_datetime: false,
  shouldShow_index_files__data_category: false,
  shouldShow_index_files__data_format: false,
  shouldShow_index_files__data_type: false,
  shouldShow_index_files__error_type: false,
  shouldShow_index_files__experimental_strategy: false,
  shouldShow_index_files__file_id: false,
  shouldShow_index_files__file_name: false,
  shouldShow_index_files__file_size: false,
  shouldShow_index_files__file_state: false,
  shouldShow_index_files__md5sum: false,
  shouldShow_index_files__platform: false,
  shouldShow_index_files__revision: false,
  shouldShow_index_files__updated_datetime: false,
  shouldShow_md5sum: false,
  shouldShow_metadata_files__access: false,
  shouldShow_metadata_files__created_datetime: false,
  shouldShow_metadata_files__data_category: false,
  shouldShow_metadata_files__data_format: false,
  shouldShow_metadata_files__data_type: false,
  shouldShow_metadata_files__error_type: false,
  shouldShow_metadata_files__file_id: false,
  shouldShow_metadata_files__file_name: false,
  shouldShow_metadata_files__file_size: false,
  shouldShow_metadata_files__file_state: false,
  shouldShow_metadata_files__md5sum: false,
  shouldShow_metadata_files__submitter_id: false,
  shouldShow_metadata_files__updated_datetime: false,
  shouldShow_origin: false,
  shouldShow_platform: false,
  shouldShow_revision: false,
  shouldShow_tags: false,
  shouldShow_updated_datetime: false
};

export const repositoryFileAggregationsFragment = () => Relay.QL`
  fragment on FileAggregations {
    
access @include(if: $shouldShow_access) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__analysis_id @include(if: $shouldShow_analysis__analysis_id) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__analysis_type @include(if: $shouldShow_analysis__analysis_type) {
  
      buckets {
        doc_count
        key
      }
      
}


analysis__input_files__access @include(if: $shouldShow_analysis__input_files__access) {
  
      buckets {
        doc_count
        key
      }
      
}


analysis__input_files__data_category @include(if: $shouldShow_analysis__input_files__data_category) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__input_files__data_format @include(if: $shouldShow_analysis__input_files__data_format) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__input_files__data_type @include(if: $shouldShow_analysis__input_files__data_type) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__input_files__experimental_strategy @include(if: $shouldShow_analysis__input_files__experimental_strategy) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__input_files__file_id @include(if: $shouldShow_analysis__input_files__file_id) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__input_files__file_name @include(if: $shouldShow_analysis__input_files__file_name) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__input_files__file_size @include(if: $shouldShow_analysis__input_files__file_size) {
  
      count
      max
      avg
      min
    
}

analysis__input_files__file_state @include(if: $shouldShow_analysis__input_files__file_state) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__input_files__md5sum @include(if: $shouldShow_analysis__input_files__md5sum) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__input_files__platform @include(if: $shouldShow_analysis__input_files__platform) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__input_files__revision @include(if: $shouldShow_analysis__input_files__revision) {
  
      count
      max
      avg
      min
    
}


analysis__metadata__read_groups__RIN @include(if: $shouldShow_analysis__metadata__read_groups__RIN) {
  
      count
      max
      avg
      min
    
}

analysis__metadata__read_groups__adapter_name @include(if: $shouldShow_analysis__metadata__read_groups__adapter_name) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__adapter_sequence @include(if: $shouldShow_analysis__metadata__read_groups__adapter_sequence) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__base_caller_name @include(if: $shouldShow_analysis__metadata__read_groups__base_caller_name) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__base_caller_version @include(if: $shouldShow_analysis__metadata__read_groups__base_caller_version) {
  
      buckets {
        doc_count
        key
      }
      
}


analysis__metadata__read_groups__experiment_name @include(if: $shouldShow_analysis__metadata__read_groups__experiment_name) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__flow_cell_barcode @include(if: $shouldShow_analysis__metadata__read_groups__flow_cell_barcode) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__includes_spike_ins @include(if: $shouldShow_analysis__metadata__read_groups__includes_spike_ins) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__instrument_model @include(if: $shouldShow_analysis__metadata__read_groups__instrument_model) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__is_paired_end @include(if: $shouldShow_analysis__metadata__read_groups__is_paired_end) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__library_name @include(if: $shouldShow_analysis__metadata__read_groups__library_name) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__library_preparation_kit_catalog_number @include(if: $shouldShow_analysis__metadata__read_groups__library_preparation_kit_catalog_number) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__library_preparation_kit_name @include(if: $shouldShow_analysis__metadata__read_groups__library_preparation_kit_name) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__library_preparation_kit_vendor @include(if: $shouldShow_analysis__metadata__read_groups__library_preparation_kit_vendor) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__library_preparation_kit_version @include(if: $shouldShow_analysis__metadata__read_groups__library_preparation_kit_version) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__library_selection @include(if: $shouldShow_analysis__metadata__read_groups__library_selection) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__library_strand @include(if: $shouldShow_analysis__metadata__read_groups__library_strand) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__library_strategy @include(if: $shouldShow_analysis__metadata__read_groups__library_strategy) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__platform @include(if: $shouldShow_analysis__metadata__read_groups__platform) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_id @include(if: $shouldShow_analysis__metadata__read_groups__read_group_id) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_name @include(if: $shouldShow_analysis__metadata__read_groups__read_group_name) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__adapter_content @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__adapter_content) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__basic_statistics @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__basic_statistics) {
  
      buckets {
        doc_count
        key
      }
      
}


analysis__metadata__read_groups__read_group_qcs__encoding @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__encoding) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__fastq_name @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__fastq_name) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__kmer_content @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__kmer_content) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__overrepresented_sequences @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__overrepresented_sequences) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__per_base_n_content @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__per_base_n_content) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__per_base_sequence_content @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__per_base_sequence_content) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__per_base_sequence_quality @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__per_base_sequence_quality) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__per_sequence_gc_content @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__per_sequence_gc_content) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__per_sequence_quality_score @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__per_sequence_quality_score) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__per_tile_sequence_quality @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__per_tile_sequence_quality) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__percent_gc_content @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__percent_gc_content) {
  
      count
      max
      avg
      min
    
}

analysis__metadata__read_groups__read_group_qcs__read_group_qc_id @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__read_group_qc_id) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__sequence_duplication_levels @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__sequence_duplication_levels) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__sequence_length_distribution @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__sequence_length_distribution) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__state @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__state) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__submitter_id @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__submitter_id) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__total_sequences @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__total_sequences) {
  
      count
      max
      avg
      min
    
}



analysis__metadata__read_groups__read_group_qcs__workflow_link @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__workflow_link) {
  
      buckets {
        doc_count
        key
      }
      
}


analysis__metadata__read_groups__read_group_qcs__workflow_type @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__workflow_type) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_group_qcs__workflow_version @include(if: $shouldShow_analysis__metadata__read_groups__read_group_qcs__workflow_version) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__read_length @include(if: $shouldShow_analysis__metadata__read_groups__read_length) {
  
      count
      max
      avg
      min
    
}

analysis__metadata__read_groups__sequencing_center @include(if: $shouldShow_analysis__metadata__read_groups__sequencing_center) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__sequencing_date @include(if: $shouldShow_analysis__metadata__read_groups__sequencing_date) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__size_selection_range @include(if: $shouldShow_analysis__metadata__read_groups__size_selection_range) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__spike_ins_concentration @include(if: $shouldShow_analysis__metadata__read_groups__spike_ins_concentration) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__spike_ins_fasta @include(if: $shouldShow_analysis__metadata__read_groups__spike_ins_fasta) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__submitter_id @include(if: $shouldShow_analysis__metadata__read_groups__submitter_id) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__target_capture_kit_catalog_number @include(if: $shouldShow_analysis__metadata__read_groups__target_capture_kit_catalog_number) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__target_capture_kit_name @include(if: $shouldShow_analysis__metadata__read_groups__target_capture_kit_name) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__target_capture_kit_target_region @include(if: $shouldShow_analysis__metadata__read_groups__target_capture_kit_target_region) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__target_capture_kit_vendor @include(if: $shouldShow_analysis__metadata__read_groups__target_capture_kit_vendor) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__target_capture_kit_version @include(if: $shouldShow_analysis__metadata__read_groups__target_capture_kit_version) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__metadata__read_groups__to_trim_adapter_sequence @include(if: $shouldShow_analysis__metadata__read_groups__to_trim_adapter_sequence) {
  
      buckets {
        doc_count
        key
      }
      
}




analysis__workflow_link @include(if: $shouldShow_analysis__workflow_link) {
  
      buckets {
        doc_count
        key
      }
      
}


analysis__workflow_type @include(if: $shouldShow_analysis__workflow_type) {
  
      buckets {
        doc_count
        key
      }
      
}

analysis__workflow_version @include(if: $shouldShow_analysis__workflow_version) {
  
      buckets {
        doc_count
        key
      }
      
}


archive__data_category @include(if: $shouldShow_archive__data_category) {
  
      buckets {
        doc_count
        key
      }
      
}

archive__data_format @include(if: $shouldShow_archive__data_format) {
  
      buckets {
        doc_count
        key
      }
      
}

archive__data_type @include(if: $shouldShow_archive__data_type) {
  
      buckets {
        doc_count
        key
      }
      
}

archive__error_type @include(if: $shouldShow_archive__error_type) {
  
      buckets {
        doc_count
        key
      }
      
}

archive__file_name @include(if: $shouldShow_archive__file_name) {
  
      buckets {
        doc_count
        key
      }
      
}

archive__file_size @include(if: $shouldShow_archive__file_size) {
  
      count
      max
      avg
      min
    
}

archive__file_state @include(if: $shouldShow_archive__file_state) {
  
      buckets {
        doc_count
        key
      }
      
}

archive__md5sum @include(if: $shouldShow_archive__md5sum) {
  
      buckets {
        doc_count
        key
      }
      
}

archive__revision @include(if: $shouldShow_archive__revision) {
  
      count
      max
      avg
      min
    
}

archive__state @include(if: $shouldShow_archive__state) {
  
      buckets {
        doc_count
        key
      }
      
}

archive__state_comment @include(if: $shouldShow_archive__state_comment) {
  
      buckets {
        doc_count
        key
      }
      
}

archive__submitter_id @include(if: $shouldShow_archive__submitter_id) {
  
      buckets {
        doc_count
        key
      }
      
}


center__center_id @include(if: $shouldShow_center__center_id) {
  
      buckets {
        doc_count
        key
      }
      
}

center__center_type @include(if: $shouldShow_center__center_type) {
  
      buckets {
        doc_count
        key
      }
      
}

center__code @include(if: $shouldShow_center__code) {
  
      buckets {
        doc_count
        key
      }
      
}

center__name @include(if: $shouldShow_center__name) {
  
      buckets {
        doc_count
        key
      }
      
}

center__namespace @include(if: $shouldShow_center__namespace) {
  
      buckets {
        doc_count
        key
      }
      
}

center__short_name @include(if: $shouldShow_center__short_name) {
  
      buckets {
        doc_count
        key
      }
      
}


data_category @include(if: $shouldShow_data_category) {
  
      buckets {
        doc_count
        key
      }
      
}

data_format @include(if: $shouldShow_data_format) {
  
      buckets {
        doc_count
        key
      }
      
}

data_type @include(if: $shouldShow_data_type) {
  
      buckets {
        doc_count
        key
      }
      
}

downstream_analyses__analysis_id @include(if: $shouldShow_downstream_analyses__analysis_id) {
  
      buckets {
        doc_count
        key
      }
      
}

downstream_analyses__analysis_type @include(if: $shouldShow_downstream_analyses__analysis_type) {
  
      buckets {
        doc_count
        key
      }
      
}


downstream_analyses__output_files__access @include(if: $shouldShow_downstream_analyses__output_files__access) {
  
      buckets {
        doc_count
        key
      }
      
}


downstream_analyses__output_files__data_category @include(if: $shouldShow_downstream_analyses__output_files__data_category) {
  
      buckets {
        doc_count
        key
      }
      
}

downstream_analyses__output_files__data_format @include(if: $shouldShow_downstream_analyses__output_files__data_format) {
  
      buckets {
        doc_count
        key
      }
      
}

downstream_analyses__output_files__data_type @include(if: $shouldShow_downstream_analyses__output_files__data_type) {
  
      buckets {
        doc_count
        key
      }
      
}

downstream_analyses__output_files__error_type @include(if: $shouldShow_downstream_analyses__output_files__error_type) {
  
      buckets {
        doc_count
        key
      }
      
}

downstream_analyses__output_files__experimental_strategy @include(if: $shouldShow_downstream_analyses__output_files__experimental_strategy) {
  
      buckets {
        doc_count
        key
      }
      
}

downstream_analyses__output_files__file_id @include(if: $shouldShow_downstream_analyses__output_files__file_id) {
  
      buckets {
        doc_count
        key
      }
      
}

downstream_analyses__output_files__file_name @include(if: $shouldShow_downstream_analyses__output_files__file_name) {
  
      buckets {
        doc_count
        key
      }
      
}

downstream_analyses__output_files__file_size @include(if: $shouldShow_downstream_analyses__output_files__file_size) {
  
      count
      max
      avg
      min
    
}

downstream_analyses__output_files__file_state @include(if: $shouldShow_downstream_analyses__output_files__file_state) {
  
      buckets {
        doc_count
        key
      }
      
}

downstream_analyses__output_files__md5sum @include(if: $shouldShow_downstream_analyses__output_files__md5sum) {
  
      buckets {
        doc_count
        key
      }
      
}

downstream_analyses__output_files__platform @include(if: $shouldShow_downstream_analyses__output_files__platform) {
  
      buckets {
        doc_count
        key
      }
      
}

downstream_analyses__output_files__revision @include(if: $shouldShow_downstream_analyses__output_files__revision) {
  
      count
      max
      avg
      min
    
}

downstream_analyses__output_files__state @include(if: $shouldShow_downstream_analyses__output_files__state) {
  
      buckets {
        doc_count
        key
      }
      
}

downstream_analyses__output_files__state_comment @include(if: $shouldShow_downstream_analyses__output_files__state_comment) {
  
      buckets {
        doc_count
        key
      }
      
}

downstream_analyses__output_files__submitter_id @include(if: $shouldShow_downstream_analyses__output_files__submitter_id) {
  
      buckets {
        doc_count
        key
      }
      
}


downstream_analyses__submitter_id @include(if: $shouldShow_downstream_analyses__submitter_id) {
  
      buckets {
        doc_count
        key
      }
      
}



downstream_analyses__workflow_link @include(if: $shouldShow_downstream_analyses__workflow_link) {
  
      buckets {
        doc_count
        key
      }
      
}


downstream_analyses__workflow_type @include(if: $shouldShow_downstream_analyses__workflow_type) {
  
      buckets {
        doc_count
        key
      }
      
}

downstream_analyses__workflow_version @include(if: $shouldShow_downstream_analyses__workflow_version) {
  
      buckets {
        doc_count
        key
      }
      
}

experimental_strategy @include(if: $shouldShow_experimental_strategy) {
  
      buckets {
        doc_count
        key
      }
      
}

file_id @include(if: $shouldShow_file_id) {
  
      buckets {
        doc_count
        key
      }
      
}

file_name @include(if: $shouldShow_file_name) {
  
      buckets {
        doc_count
        key
      }
      
}

file_size @include(if: $shouldShow_file_size) {
  
      count
      max
      avg
      min
    
}

file_state @include(if: $shouldShow_file_state) {
  
      buckets {
        doc_count
        key
      }
      
}

index_files__access @include(if: $shouldShow_index_files__access) {
  
      buckets {
        doc_count
        key
      }
      
}


index_files__data_category @include(if: $shouldShow_index_files__data_category) {
  
      buckets {
        doc_count
        key
      }
      
}

index_files__data_format @include(if: $shouldShow_index_files__data_format) {
  
      buckets {
        doc_count
        key
      }
      
}

index_files__data_type @include(if: $shouldShow_index_files__data_type) {
  
      buckets {
        doc_count
        key
      }
      
}

index_files__error_type @include(if: $shouldShow_index_files__error_type) {
  
      buckets {
        doc_count
        key
      }
      
}

index_files__experimental_strategy @include(if: $shouldShow_index_files__experimental_strategy) {
  
      buckets {
        doc_count
        key
      }
      
}

index_files__file_id @include(if: $shouldShow_index_files__file_id) {
  
      buckets {
        doc_count
        key
      }
      
}

index_files__file_name @include(if: $shouldShow_index_files__file_name) {
  
      buckets {
        doc_count
        key
      }
      
}

index_files__file_size @include(if: $shouldShow_index_files__file_size) {
  
      count
      max
      avg
      min
    
}

index_files__file_state @include(if: $shouldShow_index_files__file_state) {
  
      buckets {
        doc_count
        key
      }
      
}

index_files__md5sum @include(if: $shouldShow_index_files__md5sum) {
  
      buckets {
        doc_count
        key
      }
      
}

index_files__platform @include(if: $shouldShow_index_files__platform) {
  
      buckets {
        doc_count
        key
      }
      
}

index_files__revision @include(if: $shouldShow_index_files__revision) {
  
      count
      max
      avg
      min
    
}


md5sum @include(if: $shouldShow_md5sum) {
  
      buckets {
        doc_count
        key
      }
      
}

metadata_files__access @include(if: $shouldShow_metadata_files__access) {
  
      buckets {
        doc_count
        key
      }
      
}


metadata_files__data_category @include(if: $shouldShow_metadata_files__data_category) {
  
      buckets {
        doc_count
        key
      }
      
}

metadata_files__data_format @include(if: $shouldShow_metadata_files__data_format) {
  
      buckets {
        doc_count
        key
      }
      
}

metadata_files__data_type @include(if: $shouldShow_metadata_files__data_type) {
  
      buckets {
        doc_count
        key
      }
      
}

metadata_files__error_type @include(if: $shouldShow_metadata_files__error_type) {
  
      buckets {
        doc_count
        key
      }
      
}

metadata_files__file_id @include(if: $shouldShow_metadata_files__file_id) {
  
      buckets {
        doc_count
        key
      }
      
}

metadata_files__file_name @include(if: $shouldShow_metadata_files__file_name) {
  
      buckets {
        doc_count
        key
      }
      
}

metadata_files__file_size @include(if: $shouldShow_metadata_files__file_size) {
  
      count
      max
      avg
      min
    
}

metadata_files__file_state @include(if: $shouldShow_metadata_files__file_state) {
  
      buckets {
        doc_count
        key
      }
      
}

metadata_files__md5sum @include(if: $shouldShow_metadata_files__md5sum) {
  
      buckets {
        doc_count
        key
      }
      
}

metadata_files__submitter_id @include(if: $shouldShow_metadata_files__submitter_id) {
  
      buckets {
        doc_count
        key
      }
      
}


origin @include(if: $shouldShow_origin) {
  
      buckets {
        doc_count
        key
      }
      
}

platform @include(if: $shouldShow_platform) {
  
      buckets {
        doc_count
        key
      }
      
}

revision @include(if: $shouldShow_revision) {
  
      count
      max
      avg
      min
    
}

tags @include(if: $shouldShow_tags) {
  
      buckets {
        doc_count
        key
      }
      
}

  }
`;
