// @ncigdc/tableColumns/projectsTableColumns.js
// @flow
import React from 'react';
import { RepositoryCasesLink, RepositoryFilesLink } from '@ncigdc/components/Links/RepositoryLink';
import ProjectLink from '@ncigdc/components/Links/ProjectLink';
import { Th, Td } from '@ncigdc/uikit/Table';
import { findDataCategory, CATEGORY_MAP } from '@ncigdc/utils/data';
import { makeFilter } from '@ncigdc/utils/filters';

type TLinkProps = { node: Object, fields?: Array<Object>, children?: mixed };
type TLink = (props: TLinkProps) => any;

const CasesLink: TLink = ({ node, fields = [], children }) => (
  <RepositoryCasesLink
    query={{
      filters: makeFilter([{ field: 'cases.project.project_id', value: [node.project_id] }, ...fields], false),
    }}
  >
    {children}
  </RepositoryCasesLink>
);

const projectsTableColumns = [
  {
    name: 'ID',
    id: 'project_id',
    th: <Th key="project_id" rowSpan="2">ID</Th>,
    td: node => (
      <Td key="project_id">
        <ProjectLink id={node.project_id} />
      </Td>
    ),
  },
  {
    name: 'Disease Type',
    id: 'disease_type',
    th: <Th key="disease_type" rowSpan="2">Disease Type</Th>,
    td: node => (
      <Td key={node.disease_type} style={{ whiteSpace: 'normal' }}>
        {node.disease_type}
      </Td>
    ),
  },
  {
    name: 'Primary Site',
    id: 'primary_site',
    th: <Th key="primary_site" rowSpan="2">Primary Site</Th>,
    td: node => (
      <Td key="primary_site">{node.primary_site}</Td>
    ),
  },
  {
    name: 'Program',
    id: 'program',
    th: <Th key="program" rowSpan="2">Program</Th>,
    td: node => (
      <Td key="program">{node.program.name}</Td>
    ),
  },
  {
    name: 'Cases',
    id: 'cases',
    th: <Th key="cases" rowSpan="2">Cases</Th>,
    td: node => (
      <Td key="cases">
        <CasesLink
          node={node}
          fields={[{ field: 'files.data_category', value: CATEGORY_MAP.Seq }]}
        >
          {node.summary.case_count.toLocaleString()}
        </CasesLink>
      </Td>
    ),
  },
  {
    name: 'Data Categories',
    id: 'data_category',
    th: <Th key="data_category" colSpan="6">Available Cases per Data Category</Th>,
    subHeadingIds: ['seq', 'exp', 'snv', 'cnv', 'clinical', 'bio'],
  },
  {
    name: 'Seq',
    id: 'seq',
    subHeading: true,
    th: <Th key="seq">Seq</Th>,
    td: node => (
      <Td key="seq">
        <CasesLink
          node={node}
          fields={[{ field: 'files.data_category', value: CATEGORY_MAP.Seq }]}
        >
          {findDataCategory('Seq', node.summary.data_categories).case_count.toLocaleString()}
        </CasesLink>
      </Td>
    ),
  },
  {
    name: 'Exp',
    id: 'exp',
    subHeading: true,
    th: <Th key="exp">Exp</Th>,
    td: node => (
      <Td key="exp">
        <CasesLink
          node={node}
          fields={[{ field: 'files.data_category', value: CATEGORY_MAP.Exp }]}
        >
          {findDataCategory('Exp', node.summary.data_categories).case_count.toLocaleString()}
        </CasesLink>
      </Td>
    ),
  },
  {
    name: 'Snv',
    id: 'snv',
    subHeading: true,
    th: <Th key="snv">SNV</Th>,
    td: node => (
      <Td key="snv">
        <CasesLink
          node={node}
          fields={[{ field: 'files.data_category', value: CATEGORY_MAP.SNV }]}
        >
          {findDataCategory('SNV', node.summary.data_categories).case_count.toLocaleString()}
        </CasesLink>
      </Td>
    ),
  },
  {
    name: 'Cnv',
    id: 'cnv',
    subHeading: true,
    th: <Th key="cnv">CNV</Th>,
    td: node => (
      <Td key="cnv">
        <CasesLink
          node={node}
          fields={[{ field: 'files.data_category', value: CATEGORY_MAP.CNV }]}
        >
          {findDataCategory('CNV', node.summary.data_categories).case_count.toLocaleString()}
        </CasesLink>
      </Td>
    ),
  },
  {
    name: 'Clinical',
    id: 'clinical',
    subHeading: true,
    th: <Th key="clinical">Clinical</Th>,
    td: node => (
      <Td key="clinical">
        <CasesLink
          node={node}
          fields={[{ field: 'files.data_category', value: CATEGORY_MAP.Clinical }]}
        >
          {findDataCategory('Clinical', node.summary.data_categories).case_count.toLocaleString()}
        </CasesLink>
      </Td>
    ),
  },
  {
    name: 'Bio',
    id: 'bio',
    subHeading: true,
    th: <Th key="bio">Bio</Th>,
    td: node => (
      <Td key="bio">
        <CasesLink
          node={node}
          fields={[{ field: 'files.data_category', value: CATEGORY_MAP.Bio }]}
        >
          {findDataCategory('Bio', node.summary.data_categories).case_count.toLocaleString()}
        </CasesLink>
      </Td>
    ),
  },
  {
    name: 'Files',
    id: 'files',
    th: <Th key="files" rowSpan="2">Files</Th>,
    td: node => (
      <Td key="files">
        <RepositoryFilesLink
          query={{
            filters: makeFilter([{ field: 'cases.project.project_id', value: node.project_id }], false),
          }}
        >
          {node.summary.file_count.toLocaleString()}
        </RepositoryFilesLink>
      </Td>
    ),
  },
];

export default projectsTableColumns;
