// @ncigdc/containers/__generated__/FrequentlyMutatedGenesTable_explore.graphql.js
/**
 * This file was generated by:
 *   relay-compiler
 *
 * @providesModule FrequentlyMutatedGenesTable_explore.graphql
 * @generated SignedSource<<2d7583ebb4dcc51b7cab8d770aaaa138>>
 * @flow
 * @nogrep
 */

'use strict';

/*::
import type {ConcreteFragment} from 'relay-runtime';
export type FrequentlyMutatedGenesTable_explore = {
  ssms?: ?any;
  filteredCases?: ?FrequentlyMutatedGenesTable_explore_filteredCases;
  genes?: ?FrequentlyMutatedGenesTable_explore_genes;
};

export type FrequentlyMutatedGenesTable_explore_filteredCases_hits = {
  total: number;
};

export type FrequentlyMutatedGenesTable_explore_filteredCases = {
  hits?: ?FrequentlyMutatedGenesTable_explore_filteredCases_hits;
};

export type FrequentlyMutatedGenesTable_explore_genes_hits_edges_node_case_hits = {
  total: number;
};

export type FrequentlyMutatedGenesTable_explore_genes_hits_edges_node_case = {
  hits?: ?FrequentlyMutatedGenesTable_explore_genes_hits_edges_node_case_hits;
};

export type FrequentlyMutatedGenesTable_explore_genes_hits_edges_node = {
  numCases?: ?number;
  symbol?: ?string;
  name?: ?string;
  cytoband?: ?Array<?string>;
  biotype?: ?string;
  gene_id?: ?string;
  is_cancer_gene_census?: ?boolean;
  case?: ?FrequentlyMutatedGenesTable_explore_genes_hits_edges_node_case;
};

export type FrequentlyMutatedGenesTable_explore_genes_hits_edges = {
  node?: ?FrequentlyMutatedGenesTable_explore_genes_hits_edges_node;
};

export type FrequentlyMutatedGenesTable_explore_genes_hits = {
  total: number;
  edges?: ?Array<?FrequentlyMutatedGenesTable_explore_genes_hits_edges>;
};

export type FrequentlyMutatedGenesTable_explore_genes = {
  hits?: ?FrequentlyMutatedGenesTable_explore_genes_hits;
};
*/

/* eslint-disable comma-dangle, quotes */

const fragment /*: ConcreteFragment*/ = {
  "argumentDefinitions": [
    {
      "kind": "LocalArgument",
      "name": "fetchData",
      "type": "Boolean",
      "defaultValue": false
    },
    {
      "kind": "LocalArgument",
      "name": "fmgTable_filters",
      "type": "FiltersArgument",
      "defaultValue": null
    },
    {
      "kind": "LocalArgument",
      "name": "fmgTable_size",
      "type": "Int",
      "defaultValue": 10
    },
    {
      "kind": "LocalArgument",
      "name": "fmgTable_offset",
      "type": "Int",
      "defaultValue": 0
    },
    {
      "kind": "LocalArgument",
      "name": "score",
      "type": "String",
      "defaultValue": "case.project.project_id"
    },
    {
      "kind": "LocalArgument",
      "name": "geneCaseFilter",
      "type": "FiltersArgument",
      "defaultValue": null
    },
    {
      "kind": "LocalArgument",
      "name": "ssmTested",
      "type": "FiltersArgument",
      "defaultValue": null
    }
  ],
  "kind": "Fragment",
  "metadata": null,
  "name": "FrequentlyMutatedGenesTable_explore",
  "selections": [
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "Ssms",
      "name": "ssms",
      "plural": false,
      "selections": [
        {
          "kind": "FragmentSpread",
          "name": "MutationsCount_ssms",
          "args": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "Condition",
      "passingValue": true,
      "condition": "fetchData",
      "selections": [
        {
          "kind": "LinkedField",
          "alias": "filteredCases",
          "args": null,
          "concreteType": "ExploreCases",
          "name": "cases",
          "plural": false,
          "selections": [
            {
              "kind": "LinkedField",
              "alias": null,
              "args": [
                {
                  "kind": "Variable",
                  "name": "filters",
                  "variableName": "geneCaseFilter",
                  "type": "FiltersArgument"
                },
                {
                  "kind": "Literal",
                  "name": "first",
                  "value": 0,
                  "type": "Int"
                }
              ],
              "concreteType": "ECaseConnection",
              "name": "hits",
              "plural": false,
              "selections": [
                {
                  "kind": "ScalarField",
                  "alias": null,
                  "args": null,
                  "name": "total",
                  "storageKey": null
                }
              ],
              "storageKey": null
            }
          ],
          "storageKey": null
        },
        {
          "kind": "LinkedField",
          "alias": null,
          "args": null,
          "concreteType": "Genes",
          "name": "genes",
          "plural": false,
          "selections": [
            {
              "kind": "LinkedField",
              "alias": null,
              "args": [
                {
                  "kind": "Variable",
                  "name": "filters",
                  "variableName": "fmgTable_filters",
                  "type": "FiltersArgument"
                },
                {
                  "kind": "Variable",
                  "name": "first",
                  "variableName": "fmgTable_size",
                  "type": "Int"
                },
                {
                  "kind": "Variable",
                  "name": "offset",
                  "variableName": "fmgTable_offset",
                  "type": "Int"
                },
                {
                  "kind": "Variable",
                  "name": "score",
                  "variableName": "score",
                  "type": "String"
                }
              ],
              "concreteType": "GeneConnection",
              "name": "hits",
              "plural": false,
              "selections": [
                {
                  "kind": "ScalarField",
                  "alias": null,
                  "args": null,
                  "name": "total",
                  "storageKey": null
                },
                {
                  "kind": "LinkedField",
                  "alias": null,
                  "args": null,
                  "concreteType": "GeneEdge",
                  "name": "edges",
                  "plural": true,
                  "selections": [
                    {
                      "kind": "LinkedField",
                      "alias": null,
                      "args": null,
                      "concreteType": "Gene",
                      "name": "node",
                      "plural": false,
                      "selections": [
                        {
                          "kind": "ScalarField",
                          "alias": "numCases",
                          "args": null,
                          "name": "score",
                          "storageKey": null
                        },
                        {
                          "kind": "ScalarField",
                          "alias": null,
                          "args": null,
                          "name": "symbol",
                          "storageKey": null
                        },
                        {
                          "kind": "ScalarField",
                          "alias": null,
                          "args": null,
                          "name": "name",
                          "storageKey": null
                        },
                        {
                          "kind": "ScalarField",
                          "alias": null,
                          "args": null,
                          "name": "cytoband",
                          "storageKey": null
                        },
                        {
                          "kind": "ScalarField",
                          "alias": null,
                          "args": null,
                          "name": "biotype",
                          "storageKey": null
                        },
                        {
                          "kind": "ScalarField",
                          "alias": null,
                          "args": null,
                          "name": "gene_id",
                          "storageKey": null
                        },
                        {
                          "kind": "ScalarField",
                          "alias": null,
                          "args": null,
                          "name": "is_cancer_gene_census",
                          "storageKey": null
                        },
                        {
                          "kind": "LinkedField",
                          "alias": null,
                          "args": null,
                          "concreteType": "GeneCases",
                          "name": "case",
                          "plural": false,
                          "selections": [
                            {
                              "kind": "LinkedField",
                              "alias": null,
                              "args": [
                                {
                                  "kind": "Variable",
                                  "name": "filters",
                                  "variableName": "ssmTested",
                                  "type": "FiltersArgument"
                                },
                                {
                                  "kind": "Literal",
                                  "name": "first",
                                  "value": 0,
                                  "type": "Int"
                                }
                              ],
                              "concreteType": "GeneCaseConnection",
                              "name": "hits",
                              "plural": false,
                              "selections": [
                                {
                                  "kind": "ScalarField",
                                  "alias": null,
                                  "args": null,
                                  "name": "total",
                                  "storageKey": null
                                }
                              ],
                              "storageKey": null
                            }
                          ],
                          "storageKey": null
                        }
                      ],
                      "storageKey": null
                    }
                  ],
                  "storageKey": null
                }
              ],
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ]
    }
  ],
  "type": "Explore"
};

module.exports = fragment;
