// @ncigdc/containers/__generated__/ProjectsPage_viewer.graphql.js
/**
 * This file was generated by:
 *   relay-compiler
 *
 * @providesModule ProjectsPage_viewer.graphql
 * @generated SignedSource<<3184d83b3ad7246ebda935d0e390c46f>>
 * @flow
 * @nogrep
 */

'use strict';

/*::
import type {ConcreteFragment} from 'relay-runtime';
export type ProjectsPage_viewer = {
  explore?: ?any;
  autocomplete?: ?ProjectsPage_viewer_autocomplete;
  projects?: ?ProjectsPage_viewer_projects;
};

export type ProjectsPage_viewer_autocomplete_hits = {
  id: string;
  project_id?: ?string;
  name?: ?string;
  primary_site?: ?Array<?string>;
};

export type ProjectsPage_viewer_autocomplete = {
  hits?: ?Array<?ProjectsPage_viewer_autocomplete_hits>;
};

export type ProjectsPage_viewer_projects = {
  aggregations?: ?any;
  hits?: ?any;
};
*/

/* eslint-disable comma-dangle, quotes */

const fragment /*: ConcreteFragment*/ = {
  "argumentDefinitions": [
    {
      "kind": "LocalArgument",
      "name": "size",
      "type": "Int",
      "defaultValue": 1000
    },
    {
      "kind": "LocalArgument",
      "name": "offset",
      "type": "Int",
      "defaultValue": null
    },
    {
      "kind": "LocalArgument",
      "name": "filters",
      "type": "FiltersArgument",
      "defaultValue": null
    },
    {
      "kind": "LocalArgument",
      "name": "idAutocomplete",
      "type": "String",
      "defaultValue": null
    },
    {
      "kind": "LocalArgument",
      "name": "runAutocomplete",
      "type": "Boolean",
      "defaultValue": false
    },
    {
      "kind": "LocalArgument",
      "name": "projects_sort",
      "type": "[Sort]",
      "defaultValue": null
    }
  ],
  "kind": "Fragment",
  "metadata": null,
  "name": "ProjectsPage_viewer",
  "selections": [
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "Explore",
      "name": "explore",
      "plural": false,
      "selections": [
        {
          "kind": "FragmentSpread",
          "name": "ProjectsCharts_explore",
          "args": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "Projects",
      "name": "projects",
      "plural": false,
      "selections": [
        {
          "kind": "LinkedField",
          "alias": null,
          "args": [
            {
              "kind": "Variable",
              "name": "filters",
              "variableName": "filters",
              "type": "FiltersArgument"
            }
          ],
          "concreteType": "ProjectAggregations",
          "name": "aggregations",
          "plural": false,
          "selections": [
            {
              "kind": "FragmentSpread",
              "name": "ProjectAggregations_aggregations",
              "args": null
            }
          ],
          "storageKey": null
        },
        {
          "kind": "LinkedField",
          "alias": null,
          "args": [
            {
              "kind": "Variable",
              "name": "filters",
              "variableName": "filters",
              "type": "FiltersArgument"
            },
            {
              "kind": "Variable",
              "name": "first",
              "variableName": "size",
              "type": "Int"
            },
            {
              "kind": "Variable",
              "name": "offset",
              "variableName": "offset",
              "type": "Int"
            },
            {
              "kind": "Variable",
              "name": "sort",
              "variableName": "projects_sort",
              "type": "[Sort]"
            }
          ],
          "concreteType": "ProjectConnection",
          "name": "hits",
          "plural": false,
          "selections": [
            {
              "kind": "FragmentSpread",
              "name": "ProjectTable_projects",
              "args": null
            },
            {
              "kind": "FragmentSpread",
              "name": "ProjectsCharts_hits",
              "args": null
            }
          ],
          "storageKey": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "Condition",
      "passingValue": true,
      "condition": "runAutocomplete",
      "selections": [
        {
          "kind": "LinkedField",
          "alias": "autocomplete",
          "args": [
            {
              "kind": "Variable",
              "name": "query",
              "variableName": "idAutocomplete",
              "type": "String"
            },
            {
              "kind": "Literal",
              "name": "types",
              "value": [
                "project"
              ],
              "type": "[String]"
            }
          ],
          "concreteType": "QueryResults",
          "name": "query",
          "plural": false,
          "selections": [
            {
              "kind": "LinkedField",
              "alias": null,
              "args": null,
              "concreteType": null,
              "name": "hits",
              "plural": true,
              "selections": [
                {
                  "kind": "ScalarField",
                  "alias": null,
                  "args": null,
                  "name": "id",
                  "storageKey": null
                },
                {
                  "kind": "InlineFragment",
                  "type": "Project",
                  "selections": [
                    {
                      "kind": "ScalarField",
                      "alias": null,
                      "args": null,
                      "name": "project_id",
                      "storageKey": null
                    },
                    {
                      "kind": "ScalarField",
                      "alias": null,
                      "args": null,
                      "name": "name",
                      "storageKey": null
                    },
                    {
                      "kind": "ScalarField",
                      "alias": null,
                      "args": null,
                      "name": "primary_site",
                      "storageKey": null
                    }
                  ]
                }
              ],
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ]
    }
  ],
  "type": "Root"
};

module.exports = fragment;
