// @ncigdc/containers/__generated__/ProjectsPageRefetchQuery.graphql.js
/**
 * This file was generated by:
 *   relay-compiler
 *
 * @providesModule ProjectsPageRefetchQuery.graphql
 * @generated SignedSource<<11329dac95608438b67249bb71a90e07>>
 * @relayHash 70e436e48a5702015f5439f7eacc46d6
 * @flow
 * @nogrep
 */

'use strict';

/*::
import type {ConcreteBatch} from 'relay-runtime';

*/

/* eslint-disable comma-dangle, quotes */

/*
query ProjectsPageRefetchQuery(
  $size: Int
  $offset: Int
  $filters: FiltersArgument
  $idAutocomplete: String
  $runAutocomplete: Boolean
  $projects_sort: [Sort]
) {
  viewer {
    ...ProjectsPage_viewer_YwgOG
  }
}

fragment ProjectsPage_viewer_YwgOG on Root {
  explore {
    ...ProjectsCharts_explore
  }
  autocomplete: query(query: $idAutocomplete, types: ["project"]) @include(if: $runAutocomplete) {
    hits {
      __typename
      id
      ... on Project {
        project_id
        name
        primary_site
      }
    }
  }
  projects {
    aggregations(filters: $filters) {
      ...ProjectAggregations_aggregations
    }
    hits(first: $size, offset: $offset, sort: $projects_sort, filters: $filters) {
      ...ProjectTable_projects
      ...ProjectsCharts_hits
    }
  }
}

fragment ProjectsCharts_explore on Explore {
  cases {
    hits(first: 0) {
      total
    }
  }
  genes {
    h: hits {
      total
    }
  }
}

fragment ProjectAggregations_aggregations on ProjectAggregations {
  primary_site {
    buckets {
      doc_count
      key
    }
  }
  program__name {
    buckets {
      doc_count
      key
    }
  }
  disease_type {
    buckets {
      doc_count
      key
    }
  }
  project_id {
    buckets {
      doc_count
      key
    }
  }
  summary__experimental_strategies__experimental_strategy {
    buckets {
      doc_count
      key
    }
  }
  summary__data_categories__data_category {
    buckets {
      doc_count
      key
    }
  }
}

fragment ProjectTable_projects on ProjectConnection {
  total
  edges @relay(plural: true) {
    node {
      id
      project_id
      disease_type
      program {
        name
      }
      primary_site
      summary {
        case_count
        data_categories {
          case_count
          data_category
        }
        file_count
      }
    }
  }
}

fragment ProjectsCharts_hits on ProjectConnection {
  total
  edges @relay(plural: true) {
    node {
      id
      project_id
      name
      disease_type
      program {
        name
      }
      primary_site
      summary {
        case_count
        data_categories {
          case_count
          data_category
        }
        file_count
      }
    }
  }
}
*/

const batch /*: ConcreteBatch*/ = {
  "fragment": {
    "argumentDefinitions": [
      {
        "kind": "LocalArgument",
        "name": "size",
        "type": "Int",
        "defaultValue": null
      },
      {
        "kind": "LocalArgument",
        "name": "offset",
        "type": "Int",
        "defaultValue": null
      },
      {
        "kind": "LocalArgument",
        "name": "filters",
        "type": "FiltersArgument",
        "defaultValue": null
      },
      {
        "kind": "LocalArgument",
        "name": "idAutocomplete",
        "type": "String",
        "defaultValue": null
      },
      {
        "kind": "LocalArgument",
        "name": "runAutocomplete",
        "type": "Boolean",
        "defaultValue": null
      },
      {
        "kind": "LocalArgument",
        "name": "projects_sort",
        "type": "[Sort]",
        "defaultValue": null
      }
    ],
    "kind": "Fragment",
    "metadata": null,
    "name": "ProjectsPageRefetchQuery",
    "selections": [
      {
        "kind": "LinkedField",
        "alias": null,
        "args": null,
        "concreteType": "Root",
        "name": "viewer",
        "plural": false,
        "selections": [
          {
            "kind": "FragmentSpread",
            "name": "ProjectsPage_viewer",
            "args": [
              {
                "kind": "Variable",
                "name": "filters",
                "variableName": "filters",
                "type": null
              },
              {
                "kind": "Variable",
                "name": "idAutocomplete",
                "variableName": "idAutocomplete",
                "type": null
              },
              {
                "kind": "Variable",
                "name": "offset",
                "variableName": "offset",
                "type": null
              },
              {
                "kind": "Variable",
                "name": "projects_sort",
                "variableName": "projects_sort",
                "type": null
              },
              {
                "kind": "Variable",
                "name": "runAutocomplete",
                "variableName": "runAutocomplete",
                "type": null
              },
              {
                "kind": "Variable",
                "name": "size",
                "variableName": "size",
                "type": null
              }
            ]
          }
        ],
        "storageKey": null
      }
    ],
    "type": "Root"
  },
  "id": null,
  "kind": "Batch",
  "metadata": {},
  "name": "ProjectsPageRefetchQuery",
  "query": {
    "argumentDefinitions": [
      {
        "kind": "LocalArgument",
        "name": "size",
        "type": "Int",
        "defaultValue": null
      },
      {
        "kind": "LocalArgument",
        "name": "offset",
        "type": "Int",
        "defaultValue": null
      },
      {
        "kind": "LocalArgument",
        "name": "filters",
        "type": "FiltersArgument",
        "defaultValue": null
      },
      {
        "kind": "LocalArgument",
        "name": "idAutocomplete",
        "type": "String",
        "defaultValue": null
      },
      {
        "kind": "LocalArgument",
        "name": "runAutocomplete",
        "type": "Boolean",
        "defaultValue": null
      },
      {
        "kind": "LocalArgument",
        "name": "projects_sort",
        "type": "[Sort]",
        "defaultValue": null
      }
    ],
    "kind": "Root",
    "name": "ProjectsPageRefetchQuery",
    "operation": "query",
    "selections": [
      {
        "kind": "LinkedField",
        "alias": null,
        "args": null,
        "concreteType": "Root",
        "name": "viewer",
        "plural": false,
        "selections": [
          {
            "kind": "LinkedField",
            "alias": null,
            "args": null,
            "concreteType": "Explore",
            "name": "explore",
            "plural": false,
            "selections": [
              {
                "kind": "LinkedField",
                "alias": null,
                "args": null,
                "concreteType": "ExploreCases",
                "name": "cases",
                "plural": false,
                "selections": [
                  {
                    "kind": "LinkedField",
                    "alias": null,
                    "args": [
                      {
                        "kind": "Literal",
                        "name": "first",
                        "value": 0,
                        "type": "Int"
                      }
                    ],
                    "concreteType": "ECaseConnection",
                    "name": "hits",
                    "plural": false,
                    "selections": [
                      {
                        "kind": "ScalarField",
                        "alias": null,
                        "args": null,
                        "name": "total",
                        "storageKey": null
                      }
                    ],
                    "storageKey": "hits{\"first\":0}"
                  }
                ],
                "storageKey": null
              },
              {
                "kind": "LinkedField",
                "alias": null,
                "args": null,
                "concreteType": "Genes",
                "name": "genes",
                "plural": false,
                "selections": [
                  {
                    "kind": "LinkedField",
                    "alias": "h",
                    "args": null,
                    "concreteType": "GeneConnection",
                    "name": "hits",
                    "plural": false,
                    "selections": [
                      {
                        "kind": "ScalarField",
                        "alias": null,
                        "args": null,
                        "name": "total",
                        "storageKey": null
                      }
                    ],
                    "storageKey": null
                  }
                ],
                "storageKey": null
              }
            ],
            "storageKey": null
          },
          {
            "kind": "LinkedField",
            "alias": null,
            "args": null,
            "concreteType": "Projects",
            "name": "projects",
            "plural": false,
            "selections": [
              {
                "kind": "LinkedField",
                "alias": null,
                "args": [
                  {
                    "kind": "Variable",
                    "name": "filters",
                    "variableName": "filters",
                    "type": "FiltersArgument"
                  }
                ],
                "concreteType": "ProjectAggregations",
                "name": "aggregations",
                "plural": false,
                "selections": [
                  {
                    "kind": "LinkedField",
                    "alias": null,
                    "args": null,
                    "concreteType": "Aggregations",
                    "name": "primary_site",
                    "plural": false,
                    "selections": [
                      {
                        "kind": "LinkedField",
                        "alias": null,
                        "args": null,
                        "concreteType": "Bucket",
                        "name": "buckets",
                        "plural": true,
                        "selections": [
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "doc_count",
                            "storageKey": null
                          },
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "key",
                            "storageKey": null
                          }
                        ],
                        "storageKey": null
                      }
                    ],
                    "storageKey": null
                  },
                  {
                    "kind": "LinkedField",
                    "alias": null,
                    "args": null,
                    "concreteType": "Aggregations",
                    "name": "program__name",
                    "plural": false,
                    "selections": [
                      {
                        "kind": "LinkedField",
                        "alias": null,
                        "args": null,
                        "concreteType": "Bucket",
                        "name": "buckets",
                        "plural": true,
                        "selections": [
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "doc_count",
                            "storageKey": null
                          },
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "key",
                            "storageKey": null
                          }
                        ],
                        "storageKey": null
                      }
                    ],
                    "storageKey": null
                  },
                  {
                    "kind": "LinkedField",
                    "alias": null,
                    "args": null,
                    "concreteType": "Aggregations",
                    "name": "disease_type",
                    "plural": false,
                    "selections": [
                      {
                        "kind": "LinkedField",
                        "alias": null,
                        "args": null,
                        "concreteType": "Bucket",
                        "name": "buckets",
                        "plural": true,
                        "selections": [
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "doc_count",
                            "storageKey": null
                          },
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "key",
                            "storageKey": null
                          }
                        ],
                        "storageKey": null
                      }
                    ],
                    "storageKey": null
                  },
                  {
                    "kind": "LinkedField",
                    "alias": null,
                    "args": null,
                    "concreteType": "Aggregations",
                    "name": "project_id",
                    "plural": false,
                    "selections": [
                      {
                        "kind": "LinkedField",
                        "alias": null,
                        "args": null,
                        "concreteType": "Bucket",
                        "name": "buckets",
                        "plural": true,
                        "selections": [
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "doc_count",
                            "storageKey": null
                          },
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "key",
                            "storageKey": null
                          }
                        ],
                        "storageKey": null
                      }
                    ],
                    "storageKey": null
                  },
                  {
                    "kind": "LinkedField",
                    "alias": null,
                    "args": null,
                    "concreteType": "Aggregations",
                    "name": "summary__experimental_strategies__experimental_strategy",
                    "plural": false,
                    "selections": [
                      {
                        "kind": "LinkedField",
                        "alias": null,
                        "args": null,
                        "concreteType": "Bucket",
                        "name": "buckets",
                        "plural": true,
                        "selections": [
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "doc_count",
                            "storageKey": null
                          },
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "key",
                            "storageKey": null
                          }
                        ],
                        "storageKey": null
                      }
                    ],
                    "storageKey": null
                  },
                  {
                    "kind": "LinkedField",
                    "alias": null,
                    "args": null,
                    "concreteType": "Aggregations",
                    "name": "summary__data_categories__data_category",
                    "plural": false,
                    "selections": [
                      {
                        "kind": "LinkedField",
                        "alias": null,
                        "args": null,
                        "concreteType": "Bucket",
                        "name": "buckets",
                        "plural": true,
                        "selections": [
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "doc_count",
                            "storageKey": null
                          },
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "key",
                            "storageKey": null
                          }
                        ],
                        "storageKey": null
                      }
                    ],
                    "storageKey": null
                  }
                ],
                "storageKey": null
              },
              {
                "kind": "LinkedField",
                "alias": null,
                "args": [
                  {
                    "kind": "Variable",
                    "name": "filters",
                    "variableName": "filters",
                    "type": "FiltersArgument"
                  },
                  {
                    "kind": "Variable",
                    "name": "first",
                    "variableName": "size",
                    "type": "Int"
                  },
                  {
                    "kind": "Variable",
                    "name": "offset",
                    "variableName": "offset",
                    "type": "Int"
                  },
                  {
                    "kind": "Variable",
                    "name": "sort",
                    "variableName": "projects_sort",
                    "type": "[Sort]"
                  }
                ],
                "concreteType": "ProjectConnection",
                "name": "hits",
                "plural": false,
                "selections": [
                  {
                    "kind": "ScalarField",
                    "alias": null,
                    "args": null,
                    "name": "total",
                    "storageKey": null
                  },
                  {
                    "kind": "LinkedField",
                    "alias": null,
                    "args": null,
                    "concreteType": "ProjectEdge",
                    "name": "edges",
                    "plural": true,
                    "selections": [
                      {
                        "kind": "LinkedField",
                        "alias": null,
                        "args": null,
                        "concreteType": "Project",
                        "name": "node",
                        "plural": false,
                        "selections": [
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "id",
                            "storageKey": null
                          },
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "project_id",
                            "storageKey": null
                          },
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "disease_type",
                            "storageKey": null
                          },
                          {
                            "kind": "LinkedField",
                            "alias": null,
                            "args": null,
                            "concreteType": "Program",
                            "name": "program",
                            "plural": false,
                            "selections": [
                              {
                                "kind": "ScalarField",
                                "alias": null,
                                "args": null,
                                "name": "name",
                                "storageKey": null
                              }
                            ],
                            "storageKey": null
                          },
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "primary_site",
                            "storageKey": null
                          },
                          {
                            "kind": "LinkedField",
                            "alias": null,
                            "args": null,
                            "concreteType": "Summary",
                            "name": "summary",
                            "plural": false,
                            "selections": [
                              {
                                "kind": "ScalarField",
                                "alias": null,
                                "args": null,
                                "name": "case_count",
                                "storageKey": null
                              },
                              {
                                "kind": "LinkedField",
                                "alias": null,
                                "args": null,
                                "concreteType": "DataCategories",
                                "name": "data_categories",
                                "plural": true,
                                "selections": [
                                  {
                                    "kind": "ScalarField",
                                    "alias": null,
                                    "args": null,
                                    "name": "case_count",
                                    "storageKey": null
                                  },
                                  {
                                    "kind": "ScalarField",
                                    "alias": null,
                                    "args": null,
                                    "name": "data_category",
                                    "storageKey": null
                                  }
                                ],
                                "storageKey": null
                              },
                              {
                                "kind": "ScalarField",
                                "alias": null,
                                "args": null,
                                "name": "file_count",
                                "storageKey": null
                              }
                            ],
                            "storageKey": null
                          },
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "name",
                            "storageKey": null
                          }
                        ],
                        "storageKey": null
                      }
                    ],
                    "storageKey": null
                  }
                ],
                "storageKey": null
              }
            ],
            "storageKey": null
          },
          {
            "kind": "Condition",
            "passingValue": true,
            "condition": "runAutocomplete",
            "selections": [
              {
                "kind": "LinkedField",
                "alias": "autocomplete",
                "args": [
                  {
                    "kind": "Variable",
                    "name": "query",
                    "variableName": "idAutocomplete",
                    "type": "String"
                  },
                  {
                    "kind": "Literal",
                    "name": "types",
                    "value": [
                      "project"
                    ],
                    "type": "[String]"
                  }
                ],
                "concreteType": "QueryResults",
                "name": "query",
                "plural": false,
                "selections": [
                  {
                    "kind": "LinkedField",
                    "alias": null,
                    "args": null,
                    "concreteType": null,
                    "name": "hits",
                    "plural": true,
                    "selections": [
                      {
                        "kind": "ScalarField",
                        "alias": null,
                        "args": null,
                        "name": "__typename",
                        "storageKey": null
                      },
                      {
                        "kind": "ScalarField",
                        "alias": null,
                        "args": null,
                        "name": "id",
                        "storageKey": null
                      },
                      {
                        "kind": "InlineFragment",
                        "type": "Project",
                        "selections": [
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "project_id",
                            "storageKey": null
                          },
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "name",
                            "storageKey": null
                          },
                          {
                            "kind": "ScalarField",
                            "alias": null,
                            "args": null,
                            "name": "primary_site",
                            "storageKey": null
                          }
                        ]
                      }
                    ],
                    "storageKey": null
                  }
                ],
                "storageKey": null
              }
            ]
          }
        ],
        "storageKey": null
      }
    ]
  },
  "text": "query ProjectsPageRefetchQuery(\n  $size: Int\n  $offset: Int\n  $filters: FiltersArgument\n  $idAutocomplete: String\n  $runAutocomplete: Boolean\n  $projects_sort: [Sort]\n) {\n  viewer {\n    ...ProjectsPage_viewer_YwgOG\n  }\n}\n\nfragment ProjectsPage_viewer_YwgOG on Root {\n  explore {\n    ...ProjectsCharts_explore\n  }\n  autocomplete: query(query: $idAutocomplete, types: [\"project\"]) @include(if: $runAutocomplete) {\n    hits {\n      __typename\n      id\n      ... on Project {\n        project_id\n        name\n        primary_site\n      }\n    }\n  }\n  projects {\n    aggregations(filters: $filters) {\n      ...ProjectAggregations_aggregations\n    }\n    hits(first: $size, offset: $offset, sort: $projects_sort, filters: $filters) {\n      ...ProjectTable_projects\n      ...ProjectsCharts_hits\n    }\n  }\n}\n\nfragment ProjectsCharts_explore on Explore {\n  cases {\n    hits(first: 0) {\n      total\n    }\n  }\n  genes {\n    h: hits {\n      total\n    }\n  }\n}\n\nfragment ProjectAggregations_aggregations on ProjectAggregations {\n  primary_site {\n    buckets {\n      doc_count\n      key\n    }\n  }\n  program__name {\n    buckets {\n      doc_count\n      key\n    }\n  }\n  disease_type {\n    buckets {\n      doc_count\n      key\n    }\n  }\n  project_id {\n    buckets {\n      doc_count\n      key\n    }\n  }\n  summary__experimental_strategies__experimental_strategy {\n    buckets {\n      doc_count\n      key\n    }\n  }\n  summary__data_categories__data_category {\n    buckets {\n      doc_count\n      key\n    }\n  }\n}\n\nfragment ProjectTable_projects on ProjectConnection {\n  total\n  edges @relay(plural: true) {\n    node {\n      id\n      project_id\n      disease_type\n      program {\n        name\n      }\n      primary_site\n      summary {\n        case_count\n        data_categories {\n          case_count\n          data_category\n        }\n        file_count\n      }\n    }\n  }\n}\n\nfragment ProjectsCharts_hits on ProjectConnection {\n  total\n  edges @relay(plural: true) {\n    node {\n      id\n      project_id\n      name\n      disease_type\n      program {\n        name\n      }\n      primary_site\n      summary {\n        case_count\n        data_categories {\n          case_count\n          data_category\n        }\n        file_count\n      }\n    }\n  }\n}\n"
};

module.exports = batch;
