// @ncigdc/containers/__generated__/ProjectAggregations_aggregations.graphql.js
/**
 * This file was generated by:
 *   relay-compiler
 *
 * @providesModule ProjectAggregations_aggregations.graphql
 * @generated SignedSource<<6366f1257b5d21351e9f4216ec28a54c>>
 * @flow
 * @nogrep
 */

'use strict';

/*::
import type {ConcreteFragment} from 'relay-runtime';
export type ProjectAggregations_aggregations = {
  primary_site?: ?ProjectAggregations_aggregations_primary_site;
  program__name?: ?ProjectAggregations_aggregations_program__name;
  disease_type?: ?ProjectAggregations_aggregations_disease_type;
  project_id?: ?ProjectAggregations_aggregations_project_id;
  summary__experimental_strategies__experimental_strategy?: ?ProjectAggregations_aggregations_summary__experimental_strategies__experimental_strategy;
  summary__data_categories__data_category?: ?ProjectAggregations_aggregations_summary__data_categories__data_category;
};

export type ProjectAggregations_aggregations_primary_site_buckets = {
  doc_count?: ?number;
  key?: ?string;
};

export type ProjectAggregations_aggregations_primary_site = {
  buckets?: ?Array<?ProjectAggregations_aggregations_primary_site_buckets>;
};

export type ProjectAggregations_aggregations_program__name_buckets = {
  doc_count?: ?number;
  key?: ?string;
};

export type ProjectAggregations_aggregations_program__name = {
  buckets?: ?Array<?ProjectAggregations_aggregations_program__name_buckets>;
};

export type ProjectAggregations_aggregations_disease_type_buckets = {
  doc_count?: ?number;
  key?: ?string;
};

export type ProjectAggregations_aggregations_disease_type = {
  buckets?: ?Array<?ProjectAggregations_aggregations_disease_type_buckets>;
};

export type ProjectAggregations_aggregations_project_id_buckets = {
  doc_count?: ?number;
  key?: ?string;
};

export type ProjectAggregations_aggregations_project_id = {
  buckets?: ?Array<?ProjectAggregations_aggregations_project_id_buckets>;
};

export type ProjectAggregations_aggregations_summary__experimental_strategies__experimental_strategy_buckets = {
  doc_count?: ?number;
  key?: ?string;
};

export type ProjectAggregations_aggregations_summary__experimental_strategies__experimental_strategy = {
  buckets?: ?Array<?ProjectAggregations_aggregations_summary__experimental_strategies__experimental_strategy_buckets>;
};

export type ProjectAggregations_aggregations_summary__data_categories__data_category_buckets = {
  doc_count?: ?number;
  key?: ?string;
};

export type ProjectAggregations_aggregations_summary__data_categories__data_category = {
  buckets?: ?Array<?ProjectAggregations_aggregations_summary__data_categories__data_category_buckets>;
};
*/

/* eslint-disable comma-dangle, quotes */

const fragment /*: ConcreteFragment*/ = {
  "argumentDefinitions": [],
  "kind": "Fragment",
  "metadata": null,
  "name": "ProjectAggregations_aggregations",
  "selections": [
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "Aggregations",
      "name": "primary_site",
      "plural": false,
      "selections": [
        {
          "kind": "LinkedField",
          "alias": null,
          "args": null,
          "concreteType": "Bucket",
          "name": "buckets",
          "plural": true,
          "selections": [
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "doc_count",
              "storageKey": null
            },
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "key",
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "Aggregations",
      "name": "program__name",
      "plural": false,
      "selections": [
        {
          "kind": "LinkedField",
          "alias": null,
          "args": null,
          "concreteType": "Bucket",
          "name": "buckets",
          "plural": true,
          "selections": [
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "doc_count",
              "storageKey": null
            },
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "key",
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "Aggregations",
      "name": "disease_type",
      "plural": false,
      "selections": [
        {
          "kind": "LinkedField",
          "alias": null,
          "args": null,
          "concreteType": "Bucket",
          "name": "buckets",
          "plural": true,
          "selections": [
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "doc_count",
              "storageKey": null
            },
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "key",
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "Aggregations",
      "name": "project_id",
      "plural": false,
      "selections": [
        {
          "kind": "LinkedField",
          "alias": null,
          "args": null,
          "concreteType": "Bucket",
          "name": "buckets",
          "plural": true,
          "selections": [
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "doc_count",
              "storageKey": null
            },
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "key",
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "Aggregations",
      "name": "summary__experimental_strategies__experimental_strategy",
      "plural": false,
      "selections": [
        {
          "kind": "LinkedField",
          "alias": null,
          "args": null,
          "concreteType": "Bucket",
          "name": "buckets",
          "plural": true,
          "selections": [
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "doc_count",
              "storageKey": null
            },
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "key",
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "Aggregations",
      "name": "summary__data_categories__data_category",
      "plural": false,
      "selections": [
        {
          "kind": "LinkedField",
          "alias": null,
          "args": null,
          "concreteType": "Bucket",
          "name": "buckets",
          "plural": true,
          "selections": [
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "doc_count",
              "storageKey": null
            },
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "key",
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ],
      "storageKey": null
    }
  ],
  "type": "ProjectAggregations"
};

module.exports = fragment;
