// @ncigdc/uikit/Button.js
// @flow

// Vendor
import React from 'react';
import { css, merge } from 'glamor';
import Color from 'color';

// Custom
import { withTheme } from '@ncigdc/theme';
import { center, margin } from '@ncigdc/theme/mixins';

/*----------------------------------------------------------------------------*/

const buttonStyles = theme => css({
  ...center,
  position: 'relative',
  cursor: 'pointer',
  padding: '6px 12px',
  fontSize: '14px',
  borderRadius: '4px',
  borderColor: 'transparent',
  borderWidth: 1,
  borderStyle: 'solid',
  backgroundColor: theme.primary,
  color: 'white',
  outline: 'none',
  transition: '0.25s ease',
  ':hover': {
    backgroundColor: Color(theme.primary).lighten(0.7).rgbString(),
  },
});

const centerStyles = css(center);

const Button = ({ style, children, rightIcon, leftIcon, theme, ...props }) => (
  <button {...merge(buttonStyles(theme), css(style))} {...props}>
    {leftIcon}
    <span {...merge(css(margin(leftIcon, rightIcon)), centerStyles)}>{children}</span>
    {rightIcon}
  </button>
);

Button.propTypes = {
  children: React.PropTypes.node,
  style: React.PropTypes.object,
  leftIcon: React.PropTypes.node,
  rightIcon: React.PropTypes.node,
  onClick: React.PropTypes.func,
};

export default withTheme(Button);
