// @ncigdc/containers/cohort/CaseTable.js
/* @flow */

import React from 'react';
import Relay from 'react-relay';

import { Row } from '@ncigdc/uikit/Flex';
import Showing from '@ncigdc/components/Pagination/Showing';

import Pagination from '@ncigdc/components/Pagination';

import TableActions from '@ncigdc/components/TableActions';

import { Tr, Th } from '@ncigdc/uikit/Table';

import CaseTr from './CaseTr';

import type { TTableProps } from '../types';

const styles = {
  table: {
    width: '100%',
    borderCollapse: 'collapse',
    borderSpacing: 0,
  },
};

export const CaseTableComponent = (props: TTableProps) => {
  const prefix = 'cases';

  return (
    <div>
      <Row style={{ backgroundColor: 'white', padding: '1rem', justifyContent: 'space-between' }}>
        <Showing
          docType="cases"
          prefix={prefix}
          params={props.relay.route.params}
          total={props.hits.total}
        />
        <TableActions
          prefix={prefix}
          total={props.hits.total}
          sortKey="cases_sort"
          endpoint={`${API}/cases`}
          downloadFields={[
            'case_id',
            'project.project_id',
            'project.primary_site',
            'demographic.gender',
            'summary.data_categories.file_count',
            'summary.data_categories.data_category',
          ]}
          sortOptions={[
            {
              id: 'project.project_id',
              name: 'Project',
            },
            {
              id: 'project.primary_site',
              name: 'Primary Site',
            },
            {
              id: 'demographic.gender',
              name: 'Gender',
            },
          ]}
        />
      </Row>
      <div style={{ overflowX: 'auto' }}>
        <table style={styles.table}>
          <thead>
            <Tr>
              <Th rowSpan="2">Case UUID</Th>
              <Th rowSpan="2">Project</Th>
              <Th rowSpan="2">Primary Site</Th>
              <Th rowSpan="2">Gender</Th>
              <Th rowSpan="2">Files</Th>
              <Th colSpan="6">Available Cases per Data Category</Th>
            </Tr>

            <Tr>
              <Th>Seq</Th>
              <Th>Exp</Th>
              <Th>SNV</Th>
              <Th>CNV</Th>
              <Th>Clinical</Th>
              <Th>Bio</Th>
            </Tr>
          </thead>
          <tbody>
            {props.hits.edges.map((e, i) => (
              <CaseTr {...e} key={e.node.id} index={i} />
            ))}
          </tbody>
        </table>
      </div>
      <Pagination prefix={prefix} params={props.relay.route.params} total={props.hits.total} />
    </div>
  );
};

export const CaseTableQuery = {
  fragments: {
    hits: () => Relay.QL`
      fragment on CaseConnection {
        total
        edges {
          node {
            id
            ${CaseTr.getFragment('node')}
          }
        }
      }
    `,
  },
};

const CaseTable = Relay.createContainer(
  CaseTableComponent,
  CaseTableQuery
);

export default CaseTable;
