// @ncigdc/utils/withSelectableList.js
// @flow
import _ from 'lodash';
import { compose, withState, withHandlers, withProps, withPropsOnChange } from 'recompose';

const getPreviousItem = (items, reference) => (reference ? _.nth(items, items.indexOf(reference) - 1) : _.last(items));
const getNextItem = (items, reference) => (reference ? _.nth(items, items.indexOf(reference) + 1) : _.first(items));

const withKeyboardSelection = (
  { keyHandlerName, listSourcePropName },
  { onFocusItem = _.noop, onSelectItem = _.noop }
) => compose(
  withState('focusedItem', 'setFocusedItem', undefined),
  withState('selectedItem', 'setSelectedItem', undefined),
  withProps(({
    [listSourcePropName]: list,
    focusedItem, setFocusedItem,
    setSelectedItem,
    ...props
  }) => ({
    focusPreviousItem: () => {
      if (_.isEmpty(list)) return;
      const nextFocus = getPreviousItem(list, focusedItem);
      setFocusedItem(nextFocus);
    },
    focusNextItem: () => {
      if (_.isEmpty(list)) return;
      const nextFocus = getNextItem(list, focusedItem);
      setFocusedItem(nextFocus);
    },
    selectItem: (item) => {
      setSelectedItem(item);
      onSelectItem(item, { focusedItem, list, ...props });
    },
    reset: () => {
      setSelectedItem(undefined);
      setFocusedItem(undefined);
    },
  })),
  withPropsOnChange(['focusedItem'], ({ focusedItem, ...props }) => onFocusItem(focusedItem, props)),
  withHandlers({
    [keyHandlerName]: ({
      focusedItem,
      selectItem,
      focusPreviousItem,
      focusNextItem,
      reset,
    }) => (event) => ({
      ArrowUp: () => {
        event.preventDefault();
        focusPreviousItem();
      },
      ArrowDown: () => {
        event.preventDefault();
        focusNextItem();
      },
      Enter: () => selectItem(focusedItem),
      Escape: () => reset(),
    }[event.key] || _.noop)(),
  })
);

export default withKeyboardSelection;
