// @ncigdc/containers/explore/__generated__/CaseTr_node.graphql.js
/**
 * This file was generated by:
 *   relay-compiler
 *
 * @providesModule CaseTr_node.graphql
 * @generated SignedSource<<e9cfb831791fd4360d214f1a7c6e063c>>
 * @flow
 * @nogrep
 */

'use strict';

/*::
import type {ConcreteFragment} from 'relay-runtime';
export type CaseTr_node = {
  case_id?: ?string;
  primary_site?: ?string;
  submitter_id?: ?string;
  project?: ?CaseTr_node_project;
  score?: ?number;
  demographic?: ?CaseTr_node_demographic;
  summary?: ?CaseTr_node_summary;
};

export type CaseTr_node_project = {
  project_id?: ?string;
};

export type CaseTr_node_demographic = {
  gender?: ?string;
};

export type CaseTr_node_summary_data_categories = {
  file_count?: ?number;
  data_category?: ?string;
};

export type CaseTr_node_summary = {
  data_categories?: ?Array<?CaseTr_node_summary_data_categories>;
};
*/

/* eslint-disable comma-dangle, quotes */

const fragment /*: ConcreteFragment*/ = {
  "argumentDefinitions": [],
  "kind": "Fragment",
  "metadata": null,
  "name": "CaseTr_node",
  "selections": [
    {
      "kind": "ScalarField",
      "alias": null,
      "args": null,
      "name": "case_id",
      "storageKey": null
    },
    {
      "kind": "ScalarField",
      "alias": null,
      "args": null,
      "name": "primary_site",
      "storageKey": null
    },
    {
      "kind": "ScalarField",
      "alias": null,
      "args": null,
      "name": "submitter_id",
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "Project",
      "name": "project",
      "plural": false,
      "selections": [
        {
          "kind": "ScalarField",
          "alias": null,
          "args": null,
          "name": "project_id",
          "storageKey": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "ScalarField",
      "alias": null,
      "args": null,
      "name": "score",
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "EDemographic",
      "name": "demographic",
      "plural": false,
      "selections": [
        {
          "kind": "ScalarField",
          "alias": null,
          "args": null,
          "name": "gender",
          "storageKey": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "Summary",
      "name": "summary",
      "plural": false,
      "selections": [
        {
          "kind": "LinkedField",
          "alias": null,
          "args": null,
          "concreteType": "DataCategories",
          "name": "data_categories",
          "plural": true,
          "selections": [
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "file_count",
              "storageKey": null
            },
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "data_category",
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ],
      "storageKey": null
    }
  ],
  "type": "ECase"
};

module.exports = fragment;
