// @ncigdc/containers/explore/__generated__/CaseTable_hits.graphql.js
/**
 * This file was generated by:
 *   relay-compiler
 *
 * @providesModule CaseTable_hits.graphql
 * @generated SignedSource<<947ad49812671f7a4a204ebe5c0aac08>>
 * @flow
 * @nogrep
 */

'use strict';

/*::
import type {ConcreteFragment} from 'relay-runtime';
export type CaseTable_hits = {
  total: number;
  edges?: ?Array<?CaseTable_hits_edges>;
};

export type CaseTable_hits_edges_node = {
  id: string;
};

export type CaseTable_hits_edges = {
  node?: ?CaseTable_hits_edges_node;
};
*/

/* eslint-disable comma-dangle, quotes */

const fragment /*: ConcreteFragment*/ = {
  "argumentDefinitions": [],
  "kind": "Fragment",
  "metadata": null,
  "name": "CaseTable_hits",
  "selections": [
    {
      "kind": "ScalarField",
      "alias": null,
      "args": null,
      "name": "total",
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "ECaseEdge",
      "name": "edges",
      "plural": true,
      "selections": [
        {
          "kind": "LinkedField",
          "alias": null,
          "args": null,
          "concreteType": "ECase",
          "name": "node",
          "plural": false,
          "selections": [
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "id",
              "storageKey": null
            },
            {
              "kind": "FragmentSpread",
              "name": "CaseTr_node",
              "args": null
            }
          ],
          "storageKey": null
        }
      ],
      "storageKey": null
    }
  ],
  "type": "ECaseConnection"
};

module.exports = fragment;
