// @ncigdc/containers/explore/__generated__/ExplorePage_viewer.graphql.js
/**
 * This file was generated by:
 *   relay-compiler
 *
 * @providesModule ExplorePage_viewer.graphql
 * @generated SignedSource<<511a7e7a137d0187a77178d589bd404e>>
 * @flow
 * @nogrep
 */

'use strict';

/*::
import type {ConcreteFragment} from 'relay-runtime';
export type ExplorePage_viewer = {
  autocomplete_cases?: ?ExplorePage_viewer_autocomplete_cases;
  autocomplete_genes?: ?ExplorePage_viewer_autocomplete_genes;
  autocomplete_ssms?: ?ExplorePage_viewer_autocomplete_ssms;
  frequentMutationsTableFragment?: ?any;
  frequentMutationsChartFragment?: ?any;
  frequentlyMutatedGenesChartFragment?: ?any;
  frequentlyMutatedGenesTableFragment?: ?any;
  projectBreakdownFragment?: ?any;
  explore?: ?ExplorePage_viewer_explore;
};

export type ExplorePage_viewer_autocomplete_cases_hits_project = {
  project_id?: ?string;
};

export type ExplorePage_viewer_autocomplete_cases_hits = {
  id: string;
  case_id?: ?string;
  project?: ?ExplorePage_viewer_autocomplete_cases_hits_project;
  submitter_id?: ?string;
};

export type ExplorePage_viewer_autocomplete_cases = {
  hits?: ?Array<?ExplorePage_viewer_autocomplete_cases_hits>;
};

export type ExplorePage_viewer_autocomplete_genes_hits = {
  id: string;
  symbol?: ?string;
  name?: ?string;
  gene_id?: ?string;
};

export type ExplorePage_viewer_autocomplete_genes = {
  hits?: ?Array<?ExplorePage_viewer_autocomplete_genes_hits>;
};

export type ExplorePage_viewer_autocomplete_ssms_hits = {
  id: string;
  ssm_id?: ?string;
};

export type ExplorePage_viewer_autocomplete_ssms = {
  hits?: ?Array<?ExplorePage_viewer_autocomplete_ssms_hits>;
};

export type ExplorePage_viewer_explore_cases_hits = {
  total: number;
};

export type ExplorePage_viewer_explore_cases = {
  aggregations?: ?any;
  hits?: ?ExplorePage_viewer_explore_cases_hits;
};

export type ExplorePage_viewer_explore_genes_hits = {
  total: number;
};

export type ExplorePage_viewer_explore_genes = {
  aggregations?: ?any;
  hits?: ?ExplorePage_viewer_explore_genes_hits;
};

export type ExplorePage_viewer_explore_ssms_hits = {
  total: number;
};

export type ExplorePage_viewer_explore_ssms = {
  aggregations?: ?any;
  hits?: ?ExplorePage_viewer_explore_ssms_hits;
};

export type ExplorePage_viewer_explore = {
  mutationsCountFragment?: ?any;
  cases?: ?ExplorePage_viewer_explore_cases;
  genes?: ?ExplorePage_viewer_explore_genes;
  ssms?: ?ExplorePage_viewer_explore_ssms;
};
*/

/* eslint-disable comma-dangle, quotes */

const fragment /*: ConcreteFragment*/ = {
  "argumentDefinitions": [
    {
      "kind": "RootArgument",
      "name": "idAutocompleteCases",
      "type": "String"
    },
    {
      "kind": "RootArgument",
      "name": "runAutocompleteCases",
      "type": "Boolean"
    },
    {
      "kind": "RootArgument",
      "name": "idAutocompleteGenes",
      "type": "String"
    },
    {
      "kind": "RootArgument",
      "name": "runAutocompleteGenes",
      "type": "Boolean"
    },
    {
      "kind": "RootArgument",
      "name": "idAutocompleteSsms",
      "type": "String"
    },
    {
      "kind": "RootArgument",
      "name": "runAutocompleteSsms",
      "type": "Boolean"
    },
    {
      "kind": "RootArgument",
      "name": "filters",
      "type": "FiltersArgument"
    },
    {
      "kind": "RootArgument",
      "name": "cases_size",
      "type": "Int"
    },
    {
      "kind": "RootArgument",
      "name": "cases_offset",
      "type": "Int"
    },
    {
      "kind": "RootArgument",
      "name": "genes_size",
      "type": "Int"
    },
    {
      "kind": "RootArgument",
      "name": "genes_offset",
      "type": "Int"
    },
    {
      "kind": "RootArgument",
      "name": "ssms_size",
      "type": "Int"
    },
    {
      "kind": "RootArgument",
      "name": "ssms_offset",
      "type": "Int"
    }
  ],
  "kind": "Fragment",
  "metadata": null,
  "name": "ExplorePage_viewer",
  "selections": [
    {
      "kind": "LinkedField",
      "alias": "frequentMutationsTableFragment",
      "args": null,
      "concreteType": "Explore",
      "name": "explore",
      "plural": false,
      "selections": [
        {
          "kind": "FragmentSpread",
          "name": "FrequentMutationsTable_explore",
          "args": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": "frequentMutationsChartFragment",
      "args": null,
      "concreteType": "Explore",
      "name": "explore",
      "plural": false,
      "selections": [
        {
          "kind": "FragmentSpread",
          "name": "FrequentMutationsChart_explore",
          "args": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": "frequentlyMutatedGenesChartFragment",
      "args": null,
      "concreteType": "Explore",
      "name": "explore",
      "plural": false,
      "selections": [
        {
          "kind": "FragmentSpread",
          "name": "FrequentlyMutatedGenesChart_explore",
          "args": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": "frequentlyMutatedGenesTableFragment",
      "args": null,
      "concreteType": "Explore",
      "name": "explore",
      "plural": false,
      "selections": [
        {
          "kind": "FragmentSpread",
          "name": "FrequentlyMutatedGenesTable_explore",
          "args": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": "projectBreakdownFragment",
      "args": null,
      "concreteType": "Explore",
      "name": "explore",
      "plural": false,
      "selections": [
        {
          "kind": "FragmentSpread",
          "name": "ProjectBreakdown_explore",
          "args": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "Explore",
      "name": "explore",
      "plural": false,
      "selections": [
        {
          "kind": "LinkedField",
          "alias": "mutationsCountFragment",
          "args": null,
          "concreteType": "Ssms",
          "name": "ssms",
          "plural": false,
          "selections": [
            {
              "kind": "FragmentSpread",
              "name": "MutationsCount_ssms",
              "args": null
            }
          ],
          "storageKey": null
        },
        {
          "kind": "LinkedField",
          "alias": null,
          "args": null,
          "concreteType": "ExploreCases",
          "name": "cases",
          "plural": false,
          "selections": [
            {
              "kind": "LinkedField",
              "alias": null,
              "args": [
                {
                  "kind": "Variable",
                  "name": "filters",
                  "variableName": "filters",
                  "type": "FiltersArgument"
                }
              ],
              "concreteType": "ECaseAggregations",
              "name": "aggregations",
              "plural": false,
              "selections": [
                {
                  "kind": "FragmentSpread",
                  "name": "CaseAggregations_aggregations",
                  "args": null
                },
                {
                  "kind": "FragmentSpread",
                  "name": "ExploreCasesPies_aggregations",
                  "args": null
                }
              ],
              "storageKey": null
            },
            {
              "kind": "LinkedField",
              "alias": null,
              "args": [
                {
                  "kind": "Variable",
                  "name": "filters",
                  "variableName": "filters",
                  "type": "FiltersArgument"
                },
                {
                  "kind": "Variable",
                  "name": "first",
                  "variableName": "cases_size",
                  "type": "Int"
                },
                {
                  "kind": "Variable",
                  "name": "offset",
                  "variableName": "cases_offset",
                  "type": "Int"
                },
                {
                  "kind": "Literal",
                  "name": "score",
                  "value": "gene.gene_id",
                  "type": "String"
                }
              ],
              "concreteType": "ECaseConnection",
              "name": "hits",
              "plural": false,
              "selections": [
                {
                  "kind": "FragmentSpread",
                  "name": "CaseTable_hits",
                  "args": null
                },
                {
                  "kind": "ScalarField",
                  "alias": null,
                  "args": null,
                  "name": "total",
                  "storageKey": null
                }
              ],
              "storageKey": null
            }
          ],
          "storageKey": null
        },
        {
          "kind": "LinkedField",
          "alias": null,
          "args": null,
          "concreteType": "Genes",
          "name": "genes",
          "plural": false,
          "selections": [
            {
              "kind": "LinkedField",
              "alias": null,
              "args": [
                {
                  "kind": "Variable",
                  "name": "filters",
                  "variableName": "filters",
                  "type": "FiltersArgument"
                }
              ],
              "concreteType": "GeneAggregations",
              "name": "aggregations",
              "plural": false,
              "selections": [
                {
                  "kind": "FragmentSpread",
                  "name": "GeneAggregations_aggregations",
                  "args": null
                }
              ],
              "storageKey": null
            },
            {
              "kind": "LinkedField",
              "alias": null,
              "args": [
                {
                  "kind": "Variable",
                  "name": "filters",
                  "variableName": "filters",
                  "type": "FiltersArgument"
                },
                {
                  "kind": "Variable",
                  "name": "first",
                  "variableName": "genes_size",
                  "type": "Int"
                },
                {
                  "kind": "Variable",
                  "name": "offset",
                  "variableName": "genes_offset",
                  "type": "Int"
                }
              ],
              "concreteType": "GeneConnection",
              "name": "hits",
              "plural": false,
              "selections": [
                {
                  "kind": "ScalarField",
                  "alias": null,
                  "args": null,
                  "name": "total",
                  "storageKey": null
                }
              ],
              "storageKey": null
            }
          ],
          "storageKey": null
        },
        {
          "kind": "LinkedField",
          "alias": null,
          "args": null,
          "concreteType": "Ssms",
          "name": "ssms",
          "plural": false,
          "selections": [
            {
              "kind": "LinkedField",
              "alias": null,
              "args": [
                {
                  "kind": "Variable",
                  "name": "filters",
                  "variableName": "filters",
                  "type": "FiltersArgument"
                }
              ],
              "concreteType": "SsmAggregations",
              "name": "aggregations",
              "plural": false,
              "selections": [
                {
                  "kind": "FragmentSpread",
                  "name": "SSMAggregations_aggregations",
                  "args": null
                }
              ],
              "storageKey": null
            },
            {
              "kind": "LinkedField",
              "alias": null,
              "args": [
                {
                  "kind": "Variable",
                  "name": "filters",
                  "variableName": "filters",
                  "type": "FiltersArgument"
                },
                {
                  "kind": "Variable",
                  "name": "first",
                  "variableName": "ssms_size",
                  "type": "Int"
                },
                {
                  "kind": "Variable",
                  "name": "offset",
                  "variableName": "ssms_offset",
                  "type": "Int"
                }
              ],
              "concreteType": "SsmConnection",
              "name": "hits",
              "plural": false,
              "selections": [
                {
                  "kind": "ScalarField",
                  "alias": null,
                  "args": null,
                  "name": "total",
                  "storageKey": null
                }
              ],
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "Condition",
      "passingValue": true,
      "condition": "runAutocompleteSsms",
      "selections": [
        {
          "kind": "LinkedField",
          "alias": "autocomplete_ssms",
          "args": [
            {
              "kind": "Variable",
              "name": "query",
              "variableName": "idAutocompleteSsms",
              "type": "String"
            },
            {
              "kind": "Literal",
              "name": "types",
              "value": [
                "ssm_centric"
              ],
              "type": "[String]"
            }
          ],
          "concreteType": "QueryResults",
          "name": "query",
          "plural": false,
          "selections": [
            {
              "kind": "LinkedField",
              "alias": null,
              "args": null,
              "concreteType": null,
              "name": "hits",
              "plural": true,
              "selections": [
                {
                  "kind": "ScalarField",
                  "alias": null,
                  "args": null,
                  "name": "id",
                  "storageKey": null
                },
                {
                  "kind": "InlineFragment",
                  "type": "Ssm",
                  "selections": [
                    {
                      "kind": "ScalarField",
                      "alias": null,
                      "args": null,
                      "name": "ssm_id",
                      "storageKey": null
                    }
                  ]
                }
              ],
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ]
    },
    {
      "kind": "Condition",
      "passingValue": true,
      "condition": "runAutocompleteGenes",
      "selections": [
        {
          "kind": "LinkedField",
          "alias": "autocomplete_genes",
          "args": [
            {
              "kind": "Variable",
              "name": "query",
              "variableName": "idAutocompleteGenes",
              "type": "String"
            },
            {
              "kind": "Literal",
              "name": "types",
              "value": [
                "gene_centric"
              ],
              "type": "[String]"
            }
          ],
          "concreteType": "QueryResults",
          "name": "query",
          "plural": false,
          "selections": [
            {
              "kind": "LinkedField",
              "alias": null,
              "args": null,
              "concreteType": null,
              "name": "hits",
              "plural": true,
              "selections": [
                {
                  "kind": "ScalarField",
                  "alias": null,
                  "args": null,
                  "name": "id",
                  "storageKey": null
                },
                {
                  "kind": "InlineFragment",
                  "type": "Gene",
                  "selections": [
                    {
                      "kind": "ScalarField",
                      "alias": null,
                      "args": null,
                      "name": "symbol",
                      "storageKey": null
                    },
                    {
                      "kind": "ScalarField",
                      "alias": null,
                      "args": null,
                      "name": "name",
                      "storageKey": null
                    },
                    {
                      "kind": "ScalarField",
                      "alias": null,
                      "args": null,
                      "name": "gene_id",
                      "storageKey": null
                    }
                  ]
                }
              ],
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ]
    },
    {
      "kind": "Condition",
      "passingValue": true,
      "condition": "runAutocompleteCases",
      "selections": [
        {
          "kind": "LinkedField",
          "alias": "autocomplete_cases",
          "args": [
            {
              "kind": "Variable",
              "name": "query",
              "variableName": "idAutocompleteCases",
              "type": "String"
            },
            {
              "kind": "Literal",
              "name": "types",
              "value": [
                "case"
              ],
              "type": "[String]"
            }
          ],
          "concreteType": "QueryResults",
          "name": "query",
          "plural": false,
          "selections": [
            {
              "kind": "LinkedField",
              "alias": null,
              "args": null,
              "concreteType": null,
              "name": "hits",
              "plural": true,
              "selections": [
                {
                  "kind": "ScalarField",
                  "alias": null,
                  "args": null,
                  "name": "id",
                  "storageKey": null
                },
                {
                  "kind": "InlineFragment",
                  "type": "Case",
                  "selections": [
                    {
                      "kind": "ScalarField",
                      "alias": null,
                      "args": null,
                      "name": "case_id",
                      "storageKey": null
                    },
                    {
                      "kind": "LinkedField",
                      "alias": null,
                      "args": null,
                      "concreteType": "Project",
                      "name": "project",
                      "plural": false,
                      "selections": [
                        {
                          "kind": "ScalarField",
                          "alias": null,
                          "args": null,
                          "name": "project_id",
                          "storageKey": null
                        }
                      ],
                      "storageKey": null
                    },
                    {
                      "kind": "ScalarField",
                      "alias": null,
                      "args": null,
                      "name": "submitter_id",
                      "storageKey": null
                    }
                  ]
                }
              ],
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ]
    }
  ],
  "type": "Root"
};

module.exports = fragment;
