// @ncigdc/components/__generated__/ProjectsCharts_explore.graphql.js
/**
 * This file was generated by:
 *   relay-compiler
 *
 * @providesModule ProjectsCharts_explore.graphql
 * @generated SignedSource<<559369bdb766d8cf4ff97b41c55f1c64>>
 * @flow
 * @nogrep
 */

'use strict';

/*::
import type {ConcreteFragment} from 'relay-runtime';
export type ProjectsCharts_explore = {
  cases?: ?ProjectsCharts_explore_cases;
  genes?: ?ProjectsCharts_explore_genes;
  genes?: ?ProjectsCharts_explore_genes;
};

export type ProjectsCharts_explore_cases_hits = {
  total: number;
};

export type ProjectsCharts_explore_cases = {
  hits?: ?ProjectsCharts_explore_cases_hits;
};

export type ProjectsCharts_explore_genes_h = {
  total: number;
};

export type ProjectsCharts_explore_genes = {
  h?: ?ProjectsCharts_explore_genes_h;
};

export type ProjectsCharts_explore_genes_hits_edges_node_filteredCases_hits = {
  total: number;
};

export type ProjectsCharts_explore_genes_hits_edges_node_filteredCases = {
  hits?: ?ProjectsCharts_explore_genes_hits_edges_node_filteredCases_hits;
};

export type ProjectsCharts_explore_genes_hits_edges_node_allCases_hits = {
  total: number;
};

export type ProjectsCharts_explore_genes_hits_edges_node_allCases = {
  hits?: ?ProjectsCharts_explore_genes_hits_edges_node_allCases_hits;
};

export type ProjectsCharts_explore_genes_hits_edges_node = {
  score?: ?number;
  symbol?: ?string;
  gene_id?: ?string;
  filteredCases?: ?ProjectsCharts_explore_genes_hits_edges_node_filteredCases;
  allCases?: ?ProjectsCharts_explore_genes_hits_edges_node_allCases;
};

export type ProjectsCharts_explore_genes_hits_edges = {
  node?: ?ProjectsCharts_explore_genes_hits_edges_node;
};

export type ProjectsCharts_explore_genes_hits = {
  total: number;
  edges?: ?Array<?ProjectsCharts_explore_genes_hits_edges>;
};

export type ProjectsCharts_explore_genes = {
  hits?: ?ProjectsCharts_explore_genes_hits;
};
*/

/* eslint-disable comma-dangle, quotes */

const fragment /*: ConcreteFragment*/ = {
  "argumentDefinitions": [
    {
      "kind": "LocalArgument",
      "name": "fetchGeneData",
      "type": "Boolean",
      "defaultValue": false
    },
    {
      "kind": "LocalArgument",
      "name": "fmgCaseCount_filters",
      "type": "FiltersArgument",
      "defaultValue": null
    },
    {
      "kind": "LocalArgument",
      "name": "fmgChart_filters",
      "type": "FiltersArgument",
      "defaultValue": null
    }
  ],
  "kind": "Fragment",
  "metadata": null,
  "name": "ProjectsCharts_explore",
  "selections": [
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "ExploreCases",
      "name": "cases",
      "plural": false,
      "selections": [
        {
          "kind": "LinkedField",
          "alias": null,
          "args": [
            {
              "kind": "Variable",
              "name": "filters",
              "variableName": "fmgCaseCount_filters",
              "type": "FiltersArgument"
            },
            {
              "kind": "Literal",
              "name": "first",
              "value": 0,
              "type": "Int"
            }
          ],
          "concreteType": "ECaseConnection",
          "name": "hits",
          "plural": false,
          "selections": [
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "total",
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "Genes",
      "name": "genes",
      "plural": false,
      "selections": [
        {
          "kind": "LinkedField",
          "alias": "h",
          "args": null,
          "concreteType": "GeneConnection",
          "name": "hits",
          "plural": false,
          "selections": [
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "total",
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ],
      "storageKey": null
    },
    {
      "kind": "Condition",
      "passingValue": true,
      "condition": "fetchGeneData",
      "selections": [
        {
          "kind": "LinkedField",
          "alias": null,
          "args": null,
          "concreteType": "Genes",
          "name": "genes",
          "plural": false,
          "selections": [
            {
              "kind": "LinkedField",
              "alias": null,
              "args": [
                {
                  "kind": "Variable",
                  "name": "filters",
                  "variableName": "fmgChart_filters",
                  "type": "FiltersArgument"
                },
                {
                  "kind": "Literal",
                  "name": "first",
                  "value": 20,
                  "type": "Int"
                },
                {
                  "kind": "Literal",
                  "name": "score",
                  "value": "case.project.project_id",
                  "type": "String"
                }
              ],
              "concreteType": "GeneConnection",
              "name": "hits",
              "plural": false,
              "selections": [
                {
                  "kind": "ScalarField",
                  "alias": null,
                  "args": null,
                  "name": "total",
                  "storageKey": null
                },
                {
                  "kind": "LinkedField",
                  "alias": null,
                  "args": null,
                  "concreteType": "GeneEdge",
                  "name": "edges",
                  "plural": true,
                  "selections": [
                    {
                      "kind": "LinkedField",
                      "alias": null,
                      "args": null,
                      "concreteType": "Gene",
                      "name": "node",
                      "plural": false,
                      "selections": [
                        {
                          "kind": "ScalarField",
                          "alias": null,
                          "args": null,
                          "name": "score",
                          "storageKey": null
                        },
                        {
                          "kind": "ScalarField",
                          "alias": null,
                          "args": null,
                          "name": "symbol",
                          "storageKey": null
                        },
                        {
                          "kind": "ScalarField",
                          "alias": null,
                          "args": null,
                          "name": "gene_id",
                          "storageKey": null
                        },
                        {
                          "kind": "LinkedField",
                          "alias": "filteredCases",
                          "args": null,
                          "concreteType": "GeneCases",
                          "name": "case",
                          "plural": false,
                          "selections": [
                            {
                              "kind": "LinkedField",
                              "alias": null,
                              "args": [
                                {
                                  "kind": "Variable",
                                  "name": "filters",
                                  "variableName": "fmgChart_filters",
                                  "type": "FiltersArgument"
                                },
                                {
                                  "kind": "Literal",
                                  "name": "first",
                                  "value": 1,
                                  "type": "Int"
                                }
                              ],
                              "concreteType": "GeneCaseConnection",
                              "name": "hits",
                              "plural": false,
                              "selections": [
                                {
                                  "kind": "ScalarField",
                                  "alias": null,
                                  "args": null,
                                  "name": "total",
                                  "storageKey": null
                                }
                              ],
                              "storageKey": null
                            }
                          ],
                          "storageKey": null
                        },
                        {
                          "kind": "LinkedField",
                          "alias": "allCases",
                          "args": null,
                          "concreteType": "GeneCases",
                          "name": "case",
                          "plural": false,
                          "selections": [
                            {
                              "kind": "LinkedField",
                              "alias": null,
                              "args": [
                                {
                                  "kind": "Literal",
                                  "name": "first",
                                  "value": 1,
                                  "type": "Int"
                                }
                              ],
                              "concreteType": "GeneCaseConnection",
                              "name": "hits",
                              "plural": false,
                              "selections": [
                                {
                                  "kind": "ScalarField",
                                  "alias": null,
                                  "args": null,
                                  "name": "total",
                                  "storageKey": null
                                }
                              ],
                              "storageKey": "hits{\"first\":1}"
                            }
                          ],
                          "storageKey": null
                        }
                      ],
                      "storageKey": null
                    }
                  ],
                  "storageKey": null
                }
              ],
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ]
    }
  ],
  "type": "Explore"
};

module.exports = fragment;
