// @ncigdc/components/__generated__/ProjectsChartsRefetchQuery.graphql.js
/**
 * This file was generated by:
 *   relay-compiler
 *
 * @providesModule ProjectsChartsRefetchQuery.graphql
 * @generated SignedSource<<a11065eca71e41ce3100634e64208205>>
 * @relayHash ee7a583b475c71ea00971f7a1760c6dd
 * @flow
 * @nogrep
 */

'use strict';

/*::
import type {ConcreteBatch} from 'relay-runtime';

*/

/* eslint-disable comma-dangle, quotes */

/*
query ProjectsChartsRefetchQuery(
  $fetchGeneData: Boolean
  $fmgCaseCount_filters: FiltersArgument
  $fmgChart_filters: FiltersArgument
) {
  explore {
    ...ProjectsCharts_explore_2uAtVa
  }
}

fragment ProjectsCharts_explore_2uAtVa on Explore {
  cases {
    hits(first: 0, filters: $fmgCaseCount_filters) {
      total
    }
  }
  genes {
    h: hits {
      total
    }
  }
  genes @include(if: $fetchGeneData) {
    hits(first: 20, filters: $fmgChart_filters, score: "case.project.project_id") {
      total
      edges {
        node {
          score
          symbol
          gene_id
          filteredCases: case {
            hits(first: 1, filters: $fmgChart_filters) {
              total
            }
          }
          allCases: case {
            hits(first: 1) {
              total
            }
          }
          id
        }
      }
    }
  }
}
*/

const batch /*: ConcreteBatch*/ = {
  "fragment": {
    "argumentDefinitions": [
      {
        "kind": "LocalArgument",
        "name": "fetchGeneData",
        "type": "Boolean",
        "defaultValue": null
      },
      {
        "kind": "LocalArgument",
        "name": "fmgCaseCount_filters",
        "type": "FiltersArgument",
        "defaultValue": null
      },
      {
        "kind": "LocalArgument",
        "name": "fmgChart_filters",
        "type": "FiltersArgument",
        "defaultValue": null
      }
    ],
    "kind": "Fragment",
    "metadata": null,
    "name": "ProjectsChartsRefetchQuery",
    "selections": [
      {
        "kind": "LinkedField",
        "alias": null,
        "args": null,
        "concreteType": "Explore",
        "name": "explore",
        "plural": false,
        "selections": [
          {
            "kind": "FragmentSpread",
            "name": "ProjectsCharts_explore",
            "args": [
              {
                "kind": "Variable",
                "name": "fetchGeneData",
                "variableName": "fetchGeneData",
                "type": null
              },
              {
                "kind": "Variable",
                "name": "fmgCaseCount_filters",
                "variableName": "fmgCaseCount_filters",
                "type": null
              },
              {
                "kind": "Variable",
                "name": "fmgChart_filters",
                "variableName": "fmgChart_filters",
                "type": null
              }
            ]
          }
        ],
        "storageKey": null
      }
    ],
    "type": "Root"
  },
  "id": null,
  "kind": "Batch",
  "metadata": {},
  "name": "ProjectsChartsRefetchQuery",
  "query": {
    "argumentDefinitions": [
      {
        "kind": "LocalArgument",
        "name": "fetchGeneData",
        "type": "Boolean",
        "defaultValue": null
      },
      {
        "kind": "LocalArgument",
        "name": "fmgCaseCount_filters",
        "type": "FiltersArgument",
        "defaultValue": null
      },
      {
        "kind": "LocalArgument",
        "name": "fmgChart_filters",
        "type": "FiltersArgument",
        "defaultValue": null
      }
    ],
    "kind": "Root",
    "name": "ProjectsChartsRefetchQuery",
    "operation": "query",
    "selections": [
      {
        "kind": "LinkedField",
        "alias": null,
        "args": null,
        "concreteType": "Explore",
        "name": "explore",
        "plural": false,
        "selections": [
          {
            "kind": "LinkedField",
            "alias": null,
            "args": null,
            "concreteType": "ExploreCases",
            "name": "cases",
            "plural": false,
            "selections": [
              {
                "kind": "LinkedField",
                "alias": null,
                "args": [
                  {
                    "kind": "Variable",
                    "name": "filters",
                    "variableName": "fmgCaseCount_filters",
                    "type": "FiltersArgument"
                  },
                  {
                    "kind": "Literal",
                    "name": "first",
                    "value": 0,
                    "type": "Int"
                  }
                ],
                "concreteType": "ECaseConnection",
                "name": "hits",
                "plural": false,
                "selections": [
                  {
                    "kind": "ScalarField",
                    "alias": null,
                    "args": null,
                    "name": "total",
                    "storageKey": null
                  }
                ],
                "storageKey": null
              }
            ],
            "storageKey": null
          },
          {
            "kind": "LinkedField",
            "alias": null,
            "args": null,
            "concreteType": "Genes",
            "name": "genes",
            "plural": false,
            "selections": [
              {
                "kind": "LinkedField",
                "alias": "h",
                "args": null,
                "concreteType": "GeneConnection",
                "name": "hits",
                "plural": false,
                "selections": [
                  {
                    "kind": "ScalarField",
                    "alias": null,
                    "args": null,
                    "name": "total",
                    "storageKey": null
                  }
                ],
                "storageKey": null
              }
            ],
            "storageKey": null
          },
          {
            "kind": "Condition",
            "passingValue": true,
            "condition": "fetchGeneData",
            "selections": [
              {
                "kind": "LinkedField",
                "alias": null,
                "args": null,
                "concreteType": "Genes",
                "name": "genes",
                "plural": false,
                "selections": [
                  {
                    "kind": "LinkedField",
                    "alias": null,
                    "args": [
                      {
                        "kind": "Variable",
                        "name": "filters",
                        "variableName": "fmgChart_filters",
                        "type": "FiltersArgument"
                      },
                      {
                        "kind": "Literal",
                        "name": "first",
                        "value": 20,
                        "type": "Int"
                      },
                      {
                        "kind": "Literal",
                        "name": "score",
                        "value": "case.project.project_id",
                        "type": "String"
                      }
                    ],
                    "concreteType": "GeneConnection",
                    "name": "hits",
                    "plural": false,
                    "selections": [
                      {
                        "kind": "ScalarField",
                        "alias": null,
                        "args": null,
                        "name": "total",
                        "storageKey": null
                      },
                      {
                        "kind": "LinkedField",
                        "alias": null,
                        "args": null,
                        "concreteType": "GeneEdge",
                        "name": "edges",
                        "plural": true,
                        "selections": [
                          {
                            "kind": "LinkedField",
                            "alias": null,
                            "args": null,
                            "concreteType": "Gene",
                            "name": "node",
                            "plural": false,
                            "selections": [
                              {
                                "kind": "ScalarField",
                                "alias": null,
                                "args": null,
                                "name": "score",
                                "storageKey": null
                              },
                              {
                                "kind": "ScalarField",
                                "alias": null,
                                "args": null,
                                "name": "symbol",
                                "storageKey": null
                              },
                              {
                                "kind": "ScalarField",
                                "alias": null,
                                "args": null,
                                "name": "gene_id",
                                "storageKey": null
                              },
                              {
                                "kind": "LinkedField",
                                "alias": "filteredCases",
                                "args": null,
                                "concreteType": "GeneCases",
                                "name": "case",
                                "plural": false,
                                "selections": [
                                  {
                                    "kind": "LinkedField",
                                    "alias": null,
                                    "args": [
                                      {
                                        "kind": "Variable",
                                        "name": "filters",
                                        "variableName": "fmgChart_filters",
                                        "type": "FiltersArgument"
                                      },
                                      {
                                        "kind": "Literal",
                                        "name": "first",
                                        "value": 1,
                                        "type": "Int"
                                      }
                                    ],
                                    "concreteType": "GeneCaseConnection",
                                    "name": "hits",
                                    "plural": false,
                                    "selections": [
                                      {
                                        "kind": "ScalarField",
                                        "alias": null,
                                        "args": null,
                                        "name": "total",
                                        "storageKey": null
                                      }
                                    ],
                                    "storageKey": null
                                  }
                                ],
                                "storageKey": null
                              },
                              {
                                "kind": "LinkedField",
                                "alias": "allCases",
                                "args": null,
                                "concreteType": "GeneCases",
                                "name": "case",
                                "plural": false,
                                "selections": [
                                  {
                                    "kind": "LinkedField",
                                    "alias": null,
                                    "args": [
                                      {
                                        "kind": "Literal",
                                        "name": "first",
                                        "value": 1,
                                        "type": "Int"
                                      }
                                    ],
                                    "concreteType": "GeneCaseConnection",
                                    "name": "hits",
                                    "plural": false,
                                    "selections": [
                                      {
                                        "kind": "ScalarField",
                                        "alias": null,
                                        "args": null,
                                        "name": "total",
                                        "storageKey": null
                                      }
                                    ],
                                    "storageKey": "hits{\"first\":1}"
                                  }
                                ],
                                "storageKey": null
                              },
                              {
                                "kind": "ScalarField",
                                "alias": null,
                                "args": null,
                                "name": "id",
                                "storageKey": null
                              }
                            ],
                            "storageKey": null
                          }
                        ],
                        "storageKey": null
                      }
                    ],
                    "storageKey": null
                  }
                ],
                "storageKey": null
              }
            ]
          }
        ],
        "storageKey": null
      }
    ]
  },
  "text": "query ProjectsChartsRefetchQuery(\n  $fetchGeneData: Boolean\n  $fmgCaseCount_filters: FiltersArgument\n  $fmgChart_filters: FiltersArgument\n) {\n  explore {\n    ...ProjectsCharts_explore_2uAtVa\n  }\n}\n\nfragment ProjectsCharts_explore_2uAtVa on Explore {\n  cases {\n    hits(first: 0, filters: $fmgCaseCount_filters) {\n      total\n    }\n  }\n  genes {\n    h: hits {\n      total\n    }\n  }\n  genes @include(if: $fetchGeneData) {\n    hits(first: 20, filters: $fmgChart_filters, score: \"case.project.project_id\") {\n      total\n      edges {\n        node {\n          score\n          symbol\n          gene_id\n          filteredCases: case {\n            hits(first: 1, filters: $fmgChart_filters) {\n              total\n            }\n          }\n          allCases: case {\n            hits(first: 1) {\n              total\n            }\n          }\n          id\n        }\n      }\n    }\n  }\n}\n"
};

module.exports = batch;
