// @ncigdc/components/__generated__/ProjectsCharts_hits.graphql.js
/**
 * This file was generated by:
 *   relay-compiler
 *
 * @providesModule ProjectsCharts_hits.graphql
 * @generated SignedSource<<fbbb2a02d3e63b3f546d5c9191f27d22>>
 * @flow
 * @nogrep
 */

'use strict';

/*::
import type {ConcreteFragment} from 'relay-runtime';
export type ProjectsCharts_hits = {
  total: number;
  edges?: ?Array<?ProjectsCharts_hits_edges>;
};

export type ProjectsCharts_hits_edges_node_program = {
  name?: ?string;
};

export type ProjectsCharts_hits_edges_node_summary_data_categories = {
  case_count?: ?number;
  data_category?: ?string;
};

export type ProjectsCharts_hits_edges_node_summary = {
  case_count?: ?number;
  data_categories?: ?Array<?ProjectsCharts_hits_edges_node_summary_data_categories>;
  file_count?: ?number;
};

export type ProjectsCharts_hits_edges_node = {
  id: string;
  project_id?: ?string;
  name?: ?string;
  disease_type?: ?Array<?string>;
  program?: ?ProjectsCharts_hits_edges_node_program;
  primary_site?: ?Array<?string>;
  summary?: ?ProjectsCharts_hits_edges_node_summary;
};

export type ProjectsCharts_hits_edges = {
  node?: ?ProjectsCharts_hits_edges_node;
};
*/

/* eslint-disable comma-dangle, quotes */

const fragment /*: ConcreteFragment*/ = {
  "argumentDefinitions": [],
  "kind": "Fragment",
  "metadata": null,
  "name": "ProjectsCharts_hits",
  "selections": [
    {
      "kind": "ScalarField",
      "alias": null,
      "args": null,
      "name": "total",
      "storageKey": null
    },
    {
      "kind": "LinkedField",
      "alias": null,
      "args": null,
      "concreteType": "ProjectEdge",
      "name": "edges",
      "plural": true,
      "selections": [
        {
          "kind": "LinkedField",
          "alias": null,
          "args": null,
          "concreteType": "Project",
          "name": "node",
          "plural": false,
          "selections": [
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "id",
              "storageKey": null
            },
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "project_id",
              "storageKey": null
            },
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "name",
              "storageKey": null
            },
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "disease_type",
              "storageKey": null
            },
            {
              "kind": "LinkedField",
              "alias": null,
              "args": null,
              "concreteType": "Program",
              "name": "program",
              "plural": false,
              "selections": [
                {
                  "kind": "ScalarField",
                  "alias": null,
                  "args": null,
                  "name": "name",
                  "storageKey": null
                }
              ],
              "storageKey": null
            },
            {
              "kind": "ScalarField",
              "alias": null,
              "args": null,
              "name": "primary_site",
              "storageKey": null
            },
            {
              "kind": "LinkedField",
              "alias": null,
              "args": null,
              "concreteType": "Summary",
              "name": "summary",
              "plural": false,
              "selections": [
                {
                  "kind": "ScalarField",
                  "alias": null,
                  "args": null,
                  "name": "case_count",
                  "storageKey": null
                },
                {
                  "kind": "LinkedField",
                  "alias": null,
                  "args": null,
                  "concreteType": "DataCategories",
                  "name": "data_categories",
                  "plural": true,
                  "selections": [
                    {
                      "kind": "ScalarField",
                      "alias": null,
                      "args": null,
                      "name": "case_count",
                      "storageKey": null
                    },
                    {
                      "kind": "ScalarField",
                      "alias": null,
                      "args": null,
                      "name": "data_category",
                      "storageKey": null
                    }
                  ],
                  "storageKey": null
                },
                {
                  "kind": "ScalarField",
                  "alias": null,
                  "args": null,
                  "name": "file_count",
                  "storageKey": null
                }
              ],
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ],
      "storageKey": null
    }
  ],
  "type": "ProjectConnection"
};

module.exports = fragment;
