// @ncigdc/containers/legacy/FilesPage.js
/* @flow */

import React from 'react';
import Relay from 'react-relay';

import SearchPage from '@ncigdc/components/SearchPage';
import TabbedLinks from '@ncigdc/components/TabbedLinks';

import FileTable from '../FileTable';
import CaseTable from '../CaseTable';
import FileAggregations from '../FileAggregations';
import CaseAggregations from '../CaseAggregations';

export type TProps = {
  viewer: {
    files: {
      aggregations: string,
      hits: string,
    },
  },
};

export const FilesPageComponent = (props: TProps) => (
  <SearchPage
    Facets={
      <TabbedLinks
        queryParam="facetTab"
        defaultIndex={0}
        links={[
          {
            text: 'cases',
            component: <CaseAggregations aggregations={props.viewer.cases.aggregations} />,
          },
          {
            text: 'files',
            component: <FileAggregations aggregations={props.viewer.files.aggregations} />,
          },
        ]}
      />
    }
    Results={<FileTable hits={props.viewer.files.hits} />}
  />
);

export const FilesPageQuery = {
  initialVariables: {
    files_offset: null,
    files_size: null,
    files_sort: null,
    filters: null,
    quicksearch: '',
    runQuicksearch: false,
  },
  fragments: {
    viewer: () => Relay.QL`
      fragment on Root {
        cases {
          aggregations(filters: $filters) {
            ${CaseAggregations.getFragment('aggregations')}
          }
          hits(first: 5, quicksearch: $quicksearch) @include(if: $runQuicksearch) {
            ${CaseTable.getFragment('hits')}
          }
        }
        files {
          aggregations(filters: $filters) {
            ${FileAggregations.getFragment('aggregations')}
          }
          hits(first: $files_size offset: $files_offset, sort: $files_sort filters: $filters) {
            ${FileTable.getFragment('hits')}
          }
        }
      }
    `,
  },
};

const FilesPage = Relay.createContainer(
  FilesPageComponent,
  FilesPageQuery
);

export default FilesPage;
