// @ncigdc/dux/cart.js
// @flow

// Vendor
import React from 'react';
import { REHYDRATE } from 'redux-persist/constants';
import _ from 'lodash';
import UndoIcon from 'react-icons/lib/md/undo';

// Custom
import { notify } from '@ncigdc/dux/notification';
import { Column } from '@ncigdc/uikit/Flex';
import { center } from '@ncigdc/theme/mixins';
import styled from '@ncigdc/theme/styled';

/*----------------------------------------------------------------------------*/

const UPDATE_CART = 'UPDATE_CART';
const ADD_CART = 'ADD_CART';

const NotificationColumn = styled(Column, ({
  textAlign: 'center',
  wordBreak: 'break-word'
}));

function toggleFilesInCart(incomingFile) {
  return (dispatch, getState) => {
    const incomingFileArray = incomingFile.constructor === Array ? incomingFile : [incomingFile];
    const existingFiles = getState().cart.files;
    const nextFiles = _.xorBy(existingFiles, incomingFileArray, 'file_id');

    if (nextFiles.length === existingFiles.length + 1) {
      dispatch(notify({
        action: 'add',
        id: `add/${incomingFile.file_name}`,
        component:
          <NotificationColumn>
            <span>Added <strong>{incomingFile.file_name}</strong> to the cart.</span>
            <span style={center}>
              <UndoIcon />
              <a onClick={() => dispatch(toggleFilesInCart(incomingFile))}>Undo</a>
            </span>
          </NotificationColumn>,
      }));
    }

    if (nextFiles.length === existingFiles.length - 1) {
      dispatch(notify({
        action: 'remove',
        id: `remove/${incomingFile.file_name}`,
        component:
          <NotificationColumn>
            <span>Removed <strong>{incomingFile.file_name}</strong> from the cart.</span>
            <span style={center}>
              <UndoIcon />
              <a onClick={() => dispatch(toggleFilesInCart(incomingFile))}>Undo</a>
            </span>
          </NotificationColumn>,
      }));
    }

    dispatch({
      type: UPDATE_CART,
      payload: nextFiles,
    });
  };
}

function addAllFilesInCart(incomingFiles) {
  return (dispatch, getState) => {
    const incomingFilesArray = incomingFiles.constructor === Array ? incomingFiles : [incomingFiles];
    const existingFiles = getState().cart.files;
    const nextFiles = _.filter(incomingFilesArray, (file) => !_.some(existingFiles, {'file_id':file.file_id}));
    if(nextFiles && nextFiles.length < incomingFilesArray.length){
      dispatch(notify({
        action: 'add',
        component:
          <NotificationColumn>
            <span>
              Added <strong>{nextFiles.length}</strong> files to the cart. 
              <strong>{incomingFilesArray.length - nextFiles.length}</strong> files already in cart, not added.
            </span>
            <span style={center}>
              <UndoIcon />
              <a onClick={() => dispatch(toggleFilesInCart(incomingFile))}>Undo</a>
            </span>
          </NotificationColumn>,
      }));
    } else if(nextFiles){
      dispatch(notify({
        action: 'add',
        component:
          <NotificationColumn>
            <span>Added <strong>{nextFiles.length} files</strong> to the cart.</span>
            <span style={center}>
              <UndoIcon />
              <a onClick={() => dispatch(toggleFilesInCart(incomingFile))}>Undo</a>
            </span>
          </NotificationColumn>,
      }));
    }
    dispatch({
      type: ADD_CART,
      payload: nextFiles,
    });
  }
}

const initialState = {
  files: [],
};

function reducer(state = initialState, action) {
  switch (action.type) {
    case REHYDRATE: {
      const incoming = action.payload.cart;
      if (incoming) return { ...state, ...incoming };
      return state;
    }
    case ADD_CART:
      return {
        files: state.files.concat(action.payload.map(file => ({
          access: file.access,
          file_id: file.file_id,
          file_size: file.file_size,
          projects: file.projects || file.cases.map(c => c.project.project_id),
        }))),
      };
    case UPDATE_CART:
      return {
        ...state,
        files: action.payload.map(f => ({
          access: f.access,
          file_id: f.file_id,
          file_size: f.file_size,
          projects: f.projects || f.cases.map(c => c.project.project_id),
        })),
      };

    default:
      return state;
  }
}

/*----------------------------------------------------------------------------*/

export { toggleFilesInCart, addAllFilesInCart };
export default reducer;
