// @ncigdc/components/GDCAppsDropdown/GDCAppsDropdown.js
// @flow

import React from 'react';
import _ from 'lodash';

import Dropdown from '@ncigdc/uikit/Dropdown';
import { ExternalLink } from '@ncigdc/uikit/Links';

import dataPortalImg from '@ncigdc/theme/images/GDC-App-data-portal-blue.svg';
import websiteImg from '@ncigdc/theme/images/GDC-App-website-blue.svg';
import cbioImg from '@ncigdc/theme/images/icon-cBioPortal.svg';

import Row from '@ncigdc/uikit/Flex/Row';
import Column from '@ncigdc/uikit/Flex/Column';
import baseStyle from './style';

const GDCAppsDropdown = () => {
  return (
    <Dropdown
      dropdownClassName="gdc-apps-menu-container"
      style={{
        margin: 0,
        border: 'none',
        color: '#777',
      }}
      itemStyle={{
        width: 'initial',
      }}
      dropdownItemStyle={{
        backgroundColor: 'transparent !important',
      }}
      dropdownStyle={{
        marginTop: '15px',
        borderBottomLeftRadius: '5px',
        borderBottomRightRadius: '5px',
      }}
      dropdownItemClass={false}
      buttonWithIcon={
        <span className={baseStyle.dropDownButton}>
          <i className="icon-gdc-apps-menu" style={{ paddingRight: '4px' }} />
          <span className="hidden-md hidden-sm" data-translate>GDC Apps</span>
        </span>
      }
    >
      <Row
        className={baseStyle.dropDown}
      >
        <Column>
          <Row
            className={`active ${baseStyle.dropDownItem}`}
          >
            <ExternalLink
              hasExternalIcon={false}
              href="https://gdc-portal.nci.nih.gov/"
              title="Data Portal"
            >
              <img
                style={{ width: '26px' }}
                src={dataPortalImg}
                className="icon icon-gdc-cbio-portal home" alt="GDC Website"
              />
              <p>Data Portal</p>
            </ExternalLink>
          </Row>
          <Row className={baseStyle.dropDownItem}>
            <ExternalLink
              hasExternalIcon={false}
              href="https://gdc.nci.nih.gov/developers/gdc-application-programming-interface-api"
            >
              <span className="icon icon-gdc-portal-api">
                {_.range(0, 11).map(x =>
                  <span key={x} className={`path${x}`} />
                )}
              </span>
              <p>API</p>
            </ExternalLink>
          </Row>
          <Row className={baseStyle.dropDownItem}>
            <ExternalLink
              hasExternalIcon={false}
              title="GDC Docs"
              href="https://gdc-docs.nci.nih.gov/"
            >
              <span className="icon icon-gdc-docs">
                {_.range(0, 15).map(x => <span key={x} className={`path${x}`} />)}
              </span>
              <p>Documentation</p>
            </ExternalLink>
          </Row>
          <Row className={baseStyle.dropDownItem}>
            <ExternalLink
              hasExternalIcon={false}
              title="GDC Legacy Archive"
              href="https://gdc-portal.nci.nih.gov/legacy-archive"
            >
              <span className="icon icon-gdc-legacy-archive">
                {_.range(0, 11).map(x => <span key={x} className={`path${x}`} />)}
              </span>
              <p>Legacy Archive</p>
            </ExternalLink>
          </Row>
        </Column>
        <Column>
          <Row className={baseStyle.dropDownItem}>
            <ExternalLink
              hasExternalIcon={false}
              title="GDC Website"
              href="https://gdc.nci.nih.gov/"
            >
              <img
                style={{ width: '26px' }}
                src={websiteImg} className="icon icon-gdc-cbio-portal home" alt="GDC Website"
              />
              <p>Website</p>
            </ExternalLink>
          </Row>
          <Row className={baseStyle.dropDownItem}>
            <ExternalLink
              hasExternalIcon={false}
              title="GDC Data Transfer Tool"
              href="https://gdc.nci.nih.gov/access-data/gdc-data-transfer-tool"
            >
              <span className="icon icon-gdc-data-transer-tool">
                {_.range(0, 9).map(x => <span key={x} className={`path${x}`} />)}
              </span>
              <p>Data Transfer Tool</p>
            </ExternalLink>
          </Row>
          <Row className={baseStyle.dropDownItem}>
            <ExternalLink
              hasExternalIcon={false}
              title="GDC Submission Portal"
              href="https://gdc-portal.nci.nih.gov/submission/"
            >
              <span className="icon icon-gdc-submission-portal">
                {_.range(0, 11).map(x => <span key={x} className={`path${x}`} />)}
              </span>
              <p>Data Submission Portal</p>
            </ExternalLink>
          </Row>
          <Row className={baseStyle.dropDownItem}>
            <ExternalLink
              hasExternalIcon={false}
              title="GDC cBio Portal"
              href="https://gdc-cbioportal.nci.nih.gov/"
            >
              <img
                style={{ width: '26px' }}
                src={cbioImg}
                className="icon icon-gdc-cbio-portal header"
                alt="GDC cBio Portal"
              />
              <p>GDC cBio Portal</p>
            </ExternalLink>
          </Row>
        </Column>
      </Row>
    </Dropdown>
  );
};

export default GDCAppsDropdown;
