// @ncigdc/components/GDCAppsDropdown/style.js
// @flow

import { css, insertRule } from 'glamor';

insertRule(`
  .gdc-apps-menu-container:before {
    width: 0;
    height: 0;
    top: -9px;
    right: 12px;
    position: absolute;
    content: ' ';
    border-left: 10px solid transparent;
    border-right: 10px solid transparent;
    border-bottom: 10px solid #fff;
  }
`);

export default {
  dropDownButton: css({
    padding: '15px 13px',
    display: 'block',
    ':hover': {
      color: '#333',
      backgroundColor: '#dedddd',
    },
  }),
  dropDown: css({
    position: 'absolute',
    zIndex: 1,
    textAlign: 'center',
    justifyContent: 'space-between',
  }),
  dropDownItem: css({
    whiteSpace: 'nowrap',
    display: 'inline !important',
    padding: '0.5rem',
    margin: '0.25rem',
    transition: 'background 0.25s ease-in-out',
    fontWeight: 'normal',
    '&.active': {
      backgroundColor: '#ccc',
    },
    ' a': {
      color: '#333',
      fontWeight: 'normal',
      fontSize: 'small',
      display: 'block',
    },
    ' .icon': {
      fontSize: '2rem',
    },
    ':hover': {
      backgroundColor: '#ededed',
    },
  }),
};
