// @ncigdc/components/GDCAppsDropdown/style.js
// @flow

import { css } from 'glamor';

export default {
  dropDownButton: css({
    padding: '15px 13px',
    display: 'block',
    ':hover': {
      color: '#333',
      backgroundColor: '#dedddd',
    },
  }),
  dropDown: css({
    position: 'absolute',
    zIndex: 1,
    textAlign: 'center',
    justifyContent: 'space-between',
    ':before': {
      width: '0',
      height: '0',
      top: '-9px',
      right: '12px',
      position: 'absolute',
      content: ' ',
      borderLeft: '10px solid transparent',
      borderRight: '10px solid transparent',
      borderBottom: '10px solid #fff',
    },
  }),
  dropDownItem: css({
    whiteSpace: 'nowrap',
    display: 'inline !important',
    padding: '0.5rem',
    margin: '0.25rem',
    transition: 'background 0.25s ease-in-out',
    fontWeight: 'normal',
    '&.active': {
      backgroundColor: '#ccc',
    },
    ' a': {
      color: '#333',
      fontWeight: 'normal',
      fontSize: 'small',
    },
    ' .icon': {
      fontSize: '2rem',
    },
    ':hover': {
      backgroundColor: '#ededed',
    },
  }),
};
