// @ncigdc/components/Oncogrid/getQueries.js
/* @flow */
/* global API */

async function getGenes({
  api,
  projectId,
  currentFilters,
}: {
  api: string,
  projectId?: string,
  currentFilters: Object,
}): Promise<Object> {
  const filters = currentFilters.content.length ? currentFilters : {
    op: 'AND',
    content: [
      { op: '=', content: { field: 'case.project.project_id', value: projectId } },
    ],
  };

  const fields = [
    'gene_id',
    'case.case_id',
    'symbol',
    'is_cancer_gene_census',
  ];

  const r = await fetch(`${api}analysis/oncogrid/genes?fields=${fields.join()}${
    projectId ? `&filters=${JSON.stringify(filters)}` : ''
  }`);

  return r.json();
}

async function getOccurences({
  api,
  projectId,
  geneIds,
  consequenceTypes,
  impact,
  currentFilters,
}: {
  api: string,
  projectId?: string,
  geneIds: Array<string>,
  consequenceTypes: Array<string>,
  impact: string,
  currentFilters: Object,
}): Promise<Object> {
  const filters = currentFilters.content.length ? currentFilters : {
    op: 'AND',
    content: [
      { op: 'in', content: { field: 'ssm.consequence.transcript.gene.gene_id', value: geneIds } },
      { op: '=', content: { field: 'ssm.consequence.transcript.annotation.impact', value: impact } },
      {
        op: 'in',
        content: { field: 'ssm.consequence.transcript.consequence_type', value: consequenceTypes },
      },
      projectId && { op: '=', content: { field: 'case.project.project_id', value: projectId } },
    ].filter(Boolean),
  };

  const fields = [
    'ssm.consequence.transcript.consequence_type',
    'ssm.consequence.transcript.annotation.impact',
    'ssm.consequence.transcript.gene.gene_id',
    'ssm.ssm_id',
    'case.case_id',
  ];

  const r = await fetch(`${api}ssm_occurrences?size=8000&filters=${JSON.stringify(filters)}&fields=${fields.join()}`);
  return r.json();
}

async function getCases({
  api,
  geneIds,
  projectId,
  currentFilters,
}: {
  api: string,
  geneIds: Array<string>,
  projectId?: string,
  currentFilters: Object,
}): Promise<Object> {
  const filters = currentFilters.content.length ? currentFilters : {
    op: 'AND',
    content: [
      { op: 'in', content: { field: 'gene.gene_id', value: geneIds } },
      projectId && { op: '=', content: { field: 'project.project_id', value: projectId } },
    ].filter(Boolean),
  };

  const fields = [
    'diagnoses.days_to_death',
    'diagnoses.age_at_diagnosis',
    'diagnoses.vital_status',
    'diagnoses.primary_diagnosis',
    'demographic.gender',
    'demographic.race',
    'demographic.ethnicity',
    'case_id',
    'summary.data_categories.file_count',
    'summary.data_categories.data_category',
  ];

  const r = await fetch(`${api}analysis/oncogrid/cases?filters=${JSON.stringify(filters)}&fields=${fields.join()}`);
  return r.json();
}

async function getQueries({
  projectId,
  // $FlowIgnore
  api = API,
  consequenceTypes,
  impact = 'HIGH',
  currentFilters,
}: {
  projectId: string,
  api: string,
  consequenceTypes: Array<string>,
  impact?: string,
  currentFilters: Object,
}): Promise<Object> {
  const { data } = await getGenes({ api, projectId, currentFilters });
  const genes = data.hits;
  const geneIds = genes.map(gene => gene.gene_id);

  const d = await Promise.all([
    getOccurences({ api, projectId, consequenceTypes, geneIds, impact, currentFilters }),
    getCases({ api, projectId, geneIds, currentFilters }),
  ]);

  return {
    genes,
    occurences: d[0].data.hits,
    cases: d[1].data.hits,
  };
}

export default getQueries;
