// @ncigdc/components/Oncogrid/getQueries.js
/* @flow */
function getGenes({
  api,
  projectId,
}: {
  api: string,
  projectId?: string,
}): Promise<Object> {
  const filters = {
    op: 'AND',
    content: [
      { op: '=', content: { field: 'case.project.project_id', value: projectId } },
    ],
  };

  const fields = [
    'gene_id',
    'case.case_id',
    'symbol',
    'is_cancer_gene_census',
  ];

  return fetch(`${api}analysis/oncogrid/genes?fields=${fields.join()}${
    projectId ? `&filters=${JSON.stringify(filters)}` : ''
  }`)
    .then(r => r.json());
}

function getOccurences({
  api,
  projectId,
  geneIds,
  consequenceTypes,
  impact,
}: {
  api: string,
  projectId?: string,
  geneIds: Array<string>,
  consequenceTypes: Array<string>,
  impact: string,
}): Promise<Object> {
  const filters = {
    op: 'AND',
    content: [
      { op: 'in', content: { field: 'ssm.consequence.transcript.gene.gene_id', value: geneIds } },
      { op: '=', content: { field: 'ssm.consequence.transcript.annotation.impact', value: impact } },
      {
        op: 'in',
        content: { field: 'ssm.consequence.transcript.consequence_type', value: consequenceTypes },
      },
      projectId && { op: '=', content: { field: 'case.project.project_id', value: projectId } },
    ].filter(Boolean),
  };

  const fields = [
    'ssm.consequence.transcript.consequence_type',
    'ssm.consequence.transcript.annotation.impact',
    'ssm.consequence.transcript.gene.gene_id',
    'ssm.ssm_id',
    'case.case_id',
  ];

  return fetch(`${api}ssm_occurrences?size=8000&filters=${JSON.stringify(filters)}&fields=${fields.join()}`)
    .then(r => r.json());
}

function getCases({
  api,
  geneIds,
  projectId,
}: {
  api: string,
  geneIds: Array<string>,
  projectId?: string,
}): Promise<Object> {
  const filters = {
    op: 'AND',
    content: [
      { op: 'in', content: { field: 'gene.gene_id', value: geneIds } },
      projectId && { op: '=', content: { field: 'project.project_id', value: projectId } },
    ].filter(Boolean),
  };

  const fields = [
    'diagnoses.days_to_death',
    'diagnoses.age_at_diagnosis',
    'diagnoses.vital_status',
    'diagnoses.primary_diagnosis',
    'demographic.gender',
    'demographic.race',
    'demographic.ethnicity',
    'case_id',
    'summary.data_categories.file_count',
    'summary.data_categories.data_category',
  ];

  return fetch(`${api}analysis/oncogrid/cases?filters=${JSON.stringify(filters)}&fields=${fields.join()}`)
    .then(r => r.json());
}

function getQueries({
  projectId,
  api,
  consequenceTypes,
  impact = 'HIGH',
}: {
  projectId: string,
  api: string,
  consequenceTypes: Array<string>,
  impact?: string,
}): Promise<Object> {
  return getGenes({ api, projectId })
    .then(({ data }) => {
      const genes = data.hits;
      const geneIds = genes.map(gene => gene.gene_id);

      return Promise.all([
        getOccurences({ api, projectId, consequenceTypes, geneIds, impact }),
        getCases({ api, projectId, geneIds }),
      ]).then(d => ({
        genes,
        occurences: d[0].data.hits,
        cases: d[1].data.hits,
      }));
    });
}

export default getQueries;
