// @ncigdc/components/Oncogrid/oncoGridParams.js
/* @flow */
import { insertRule } from 'glamor';
import { uniq } from 'lodash';

import {
  clinicalDonorTracks,
  dataTypeTracks,
  geneTracks,
  geneSetTracks,
  gdcTracks,
  getColorValue,
} from '@ncigdc/components/Oncogrid/tracks';
import { mapDonors, mapGenes, mapOccurences } from './dataMapping';
import type { TDonor, TGene, TOccurence, TDonorInput, TGeneInput, TOccurenceInput } from './dataMapping';

const donorTracks = [
  ...clinicalDonorTracks,
  ...dataTypeTracks,
];

const trackColorMap = [
  ...donorTracks,
  ...geneTracks,
].reduce((acc, t) => ({ ...acc, [t.fieldName]: t.color }), {});

const fillFunc = (track: {fieldName: string, value?: mixed}) => getColorValue({
  color: trackColorMap[track.fieldName] || '',
  value: track.value,
});

function dataTypeLegend(): string {
  return `<b>Available Data Types:</b><br>${
    dataTypeTracks
      .map(t => t.legend ? t.legend(t) : '')
      .filter(Boolean)
      .join('<br>')
  }`;
}

function gdcLegend(max: number): string {
  return `<b># of Cases Affected:</b><br>${
    gdcTracks
      .map(t => t.legend ? t.legend({ ...t, max }) : '')
      .filter(Boolean)
      .join('<br>')
  }`;
}

function geneSetLegend(): string {
  return `<b>Gene Sets:</b><br>${
    geneSetTracks
      .map(t => t.legend ? t.legend(t) : '')
      .filter(Boolean)
      .join('<br>')
  }`;
}

export default function ({
  donorData,
  geneData,
  occurencesData,
  colorMap,
  element,
  height = 150,
  width = 680,
  addTrackFunc,
  consequenceTypes,
  impacts,
  donorHistogramClick = (d: {}) => console.log('donorHistogramClick: ', d),
  gridClick = (o: {}) => console.log('gridClick: ', o),
  geneHistogramClick = (g: {}) => console.log('geneHistogramClick: ', g),
  geneClick = (g: {}) => { console.log('geneClick', g); },
  donorClick = (d: {}) => { console.log('donorClick', d); },
}: {
  donorData: Array<TDonorInput>,
  geneData: Array<TGeneInput>,
  occurencesData: Array<TOccurenceInput>,
  colorMap: {[key: string]: string},
  element: string,
  height: number,
  width: number,
  addTrackFunc: (trackOptions: Array<{name: string}>, callback: (t: Array<{name: string}>) => void) => void,
  consequenceTypes: Array<string>,
  impacts: Array<string>,
  donorHistogramClick?: Function,
  gridClick?: Function,
  geneHistogramClick?: Function,
  geneClick?: Function,
  donorClick?: Function,
}): ?Object {
  let donors: Array<TDonor> = mapDonors(donorData); // eslint-disable-line fp/no-let
  let genes: Array<TGene> = mapGenes(geneData); // eslint-disable-line fp/no-let
  const observations: Array<TOccurence> = mapOccurences(occurencesData, donors, genes, consequenceTypes, impacts);

  if (observations.length === 0) return null;

  // Clean gene & donor data before using for oncogrid.
  const donorObs = observations.map(o => o.donorId);
  const geneObs = observations.map(o => o.geneId);
  donors = donors.filter(d => donorObs.indexOf(d.id) >= 0); // eslint-disable-line fp/no-mutation
  genes = genes.filter(g => geneObs.indexOf(g.id) >= 0); // eslint-disable-line fp/no-mutation

  const maxDaysToDeath = Math.max(...donors.map((d) => d.daysToDeath));
  const maxAgeAtDiagnosis = Math.max(...donors.map((d) => d.age));
  const maxDonorsAffected = Math.max(...genes.map(g => g.totalDonors));

  const donorOpacityFunc = ({ type, value }: { type: string, value: number }) => {
    switch (type) {
      case 'int':
        return value / 100;
      case 'vital':
      case 'gender':
      case 'ethnicity':
      case 'race':
      case 'primary_diagnosi':
        return 1;
      case 'bool':
        return value ? 1 : 0;
      case 'days_to_death':
        return value / maxDaysToDeath;
      case 'age':
        return value / maxAgeAtDiagnosis;
      default:
        return 0;
    }
  };

  const geneOpacity = ({ type, value }: {type: string, value: number}) => {
    switch (type) {
      case 'int':
        return value / maxDonorsAffected;
      case 'bool':
        return value ? 1 : 0;
      default:
        return 1;
    }
  };

  return {
    donors,
    genes,
    observations,
    height,
    width,
    element,
    colorMap,
    scaleToFit: true,
    heatMap: false,
    grid: true,
    minCellHeight: 8,
    trackHeight: 12,
    trackLegends: {
      GDC: gdcLegend(maxDonorsAffected),
      'Gene Sets': geneSetLegend(),

      Clinical: `<b>Clinical Data:</b><br>${
        clinicalDonorTracks
          .map(t => t.legend ? t.legend({
            ...t,
            maxDaysToDeath,
            values: uniq(donors.map(d => d[t.fieldName])),
          }) : '')
          .filter(Boolean)
          .join('<br>')
      }`,
      'Data Types': dataTypeLegend(),
    },
    templates: {
      mainGrid: `
        {{#observation}}
          Case: {{observation.donorId}}<br>
          Gene: {{observation.geneSymbol}}<br>
          Mutation: {{observation.id}}<br>
          Consequence: {{observation.consequence}}<br>
        {{/observation}}
      `,
      mainGridCrosshair: `
        {{#donor}}Case: {{donor.id}}<br>{{/donor}}
        {{#gene}}Gene: {{gene.symbol}}<br>{{/gene}}
        {{#obs}}Mutations: {{obs}}<br>{{/obs}}
      `,
    },
    trackLegendLabel: '<i style="font-size: 13px; margin-left: 5px" class="fa fa-question-circle"></i>',
    donorTracks,
    donorOpacityFunc,
    donorFillFunc: fillFunc,
    geneTracks,
    geneOpacityFunc: geneOpacity,
    geneFillFunc: fillFunc,
    expandableGroups: ['Clinical'],
    margin: { top: 20, right: 5, bottom: 20, left: 0 },
    leftTextWidth: 120,
    addTrackFunc,
    donorHistogramClick,
    gridClick,
    geneHistogramClick,
    geneClick,
    donorClick,
  };
}

/*----------------------------------------------------------------------------*/

insertRule(`
  .onco-track-legend {
    margin-right: 5px;
    display: inline-block;
    width: 12px;
    height: 12px;
  }
`);
