// @ncigdc/containers/AnnotationsPage.js
/* @flow */

import React from 'react';
import Relay from 'react-relay';
import { compose, withState } from 'recompose';

import SearchPage from '@ncigdc/components/SearchPage';
import { DoubleArrowLeftIcon } from '@ncigdc/theme/icons';

import AnnotationTable from './AnnotationTable';
import AnnotationAggregations from './AnnotationAggregations';
import TabbedLinks from '@ncigdc/components/TabbedLinks.js';
import UnstyledButton from '@ncigdc/uikit/UnstyledButton.js';

export type TProps = {
  location: Object,
  relay: Object,
  viewer: {
    autocomplete: {
      hits: Array<Object>,
    },
    annotations: {
      aggregations: string,
      hits: string,
    },
  },
  showFacets: boolean,
  setShowFacets: Function,
};

const enhance = compose(
  withState('showFacets', 'setShowFacets', true)
);


export const AnnotationsPageComponent = (props: TProps) => (
  <SearchPage
    toggleShowFacets={() => props.setShowFacets(!props.showFacets)}
    showFacets={props.showFacets}
    facets={
      <TabbedLinks
        queryParam="facetTab"
        defaultIndex={0}
        tabToolbar={(
          <UnstyledButton
            onClick={() => {
              props.setShowFacets(!props.showFacets);
            }}
            style={{ padding: '1.2rem 0' }}
          >
            <DoubleArrowLeftIcon /> Hide Filters
          </UnstyledButton>
        )}
        links={[
          {
            id: 'cases',
            text: 'Cases',
            component: (
              <AnnotationAggregations
                aggregations={props.viewer.annotations.aggregations}
                suggestions={(props.viewer.autocomplete || { hits: [] }).hits}
                setAutocomplete={
                  (value) => { props.relay.setVariables({ idAutocomplete: value, runAutocomplete: !!value }); }
                }
              />
            ),
          },
        ]}
      />
    }
    results={<AnnotationTable hits={props.viewer.annotations.hits} />}
  />
);

export const AnnotationsPageQuery = {
  initialVariables: {
    size: null,
    offset: null,
    filters: null,
    idAutocomplete: null,
    runAutocomplete: false,
    annotations_sort: null,
  },
  fragments: {
    viewer: () => Relay.QL`
      fragment on Root {
        autocomplete: query(query: $idAutocomplete types: ["annotation"]) @include(if: $runAutocomplete) {
          hits {
            id
            ...on Annotation {
              annotation_id
            }
          }
        },
        annotations {
          aggregations(filters: $filters) {
            ${AnnotationAggregations.getFragment('aggregations')}
          }
          hits(first: $size offset: $offset, filters: $filters, sort: $annotations_sort) {
            ${AnnotationTable.getFragment('hits')}
          }
        }
      }
    `,
  },
};

const AnnotationsPage = Relay.createContainer(
  enhance(AnnotationsPageComponent),
  AnnotationsPageQuery
);

export default AnnotationsPage;
