// @ncigdc/containers/AnnotationTr.js
/* @flow */

import React from 'react';
import Relay from 'react-relay';

import AnnotationLink from '@ncigdc/components/Links/AnnotationLink';

import { Tr, Td } from '@ncigdc/uikit/Table';

import { withTheme } from '@ncigdc/theme';

export type TProps = {
  index: number,
  node: {
    annotation_id: string,
    case_id: string,
    category: string,
    classification: string,
    created_datetime: string,
    entity_id: string,
    entity_type: string,
    project: {
      project_id: string,
    },
  },
  theme: Object,
};

export const AnnotationTrComponent = ({ node, index, theme }: TProps) => (
  <Tr
    style={{
      backgroundColor: index % 2 === 0 ? theme.tableStripe : '#fff',
    }}
  >
    <Td>
      <AnnotationLink id={node.annotation_id}>
        {node.annotation_id}
      </AnnotationLink>
    </Td>
    <Td>{node.case_id}</Td>
    <Td>{node.project.project_id}</Td>
    <Td>{node.entity_type}</Td>
    <Td>{node.entity_id}</Td>
    <Td>{node.category}</Td>
    <Td>{node.classification}</Td>
    <Td>{node.created_datetime}</Td>
  </Tr>
);

export const AnnotationTrQuery = {
  fragments: {
    node: () => Relay.QL`
      fragment on Annotation {
        annotation_id
        case_id
        project {
          project_id
        }
        entity_type
        entity_id
        category
        classification
        created_datetime
      }
    `,
  },
};

const AnnotationTr = Relay.createContainer(
  withTheme(AnnotationTrComponent),
  AnnotationTrQuery
);

export default AnnotationTr;
