// @ncigdc/containers/AnnotationAggregations.js
/* @flow */

import React from 'react';
import Relay from 'react-relay';

import TermAggregation from '@ncigdc/components/Aggregations/TermAggregation';

import type { TBucket } from '@ncigdc/components/Aggregations/types';

import theme from '@ncigdc/theme';

export type TProps = {
  aggregations: {
    category: { buckets: [TBucket] },
    classification: { buckets: [TBucket] },
    entity_type: { buckets: [TBucket] },
    project__primary_site: { buckets: [TBucket] },
    project__program__name: { buckets: [TBucket] },
    project__project_id: { buckets: [TBucket] },
    status: { buckets: [TBucket] },
  },
};

export const AnnotationAggregationsComponent = (props: TProps) => {
  const docType = 'annotations';
  const facets = [
    'classification',
    'category',
    'entity_type',
    'project__primary_site',
    'project__program__name',
    'project__project_id',
    'status',
  ];

  return (
    <div>
      {facets.map((f, i) => (
        <TermAggregation
          style={{ ...(i ? { borderTop: `1px solid ${theme().greyScale5}` } : {}) }}
          key={`${docType}.${f}`}
          field={`${docType}.${f}`}
          buckets={props.aggregations[f].buckets}
        />
      ))}
    </div>
  );
};


export const AnnotationAggregationsQuery = {
  fragments: {
    aggregations: () => Relay.QL`
      fragment on AnnotationAggregations {
        category {
          buckets {
            doc_count
            key
          }
        }
        classification {
          buckets {
            doc_count
            key
          }
        }
        entity_type {
          buckets {
            doc_count
            key
          }
        }
        project__primary_site {
          buckets {
            doc_count
            key
          }
        }
        project__program__name {
          buckets {
            doc_count
            key
          }
        }
        project__project_id {
          buckets {
            doc_count
            key
          }
        }
        status {
          buckets {
            doc_count
            key
          }
        }
      }
    `,
  },
};


const AnnotationAggregations = Relay.createContainer(
  AnnotationAggregationsComponent,
  AnnotationAggregationsQuery
);

export default AnnotationAggregations;
