// @ncigdc/containers/CancerDistributionChart.js
// @flow

import React from 'react';
import Relay from 'react-relay';
import { lifecycle, compose } from 'recompose';
import { sortBy } from 'lodash';

import withSize from '@ncigdc/utils/withSize';
import withRouter from '@ncigdc/utils/withRouter';
import { makeFilter } from '@ncigdc/utils/filters';
import { Row, Column } from '@ncigdc/uikit/Flex';
import DownloadVisualizationButton from '@ncigdc/components/DownloadVisualizationButton';
import { withTheme } from '@ncigdc/theme';
import BarChart from '@ncigdc/components/Charts/BarChart';
import Loader from '@ncigdc/uikit/Loaders/Loader';
import wrapSvg from '@ncigdc/utils/wrapSvg';

type TProps = {
  aggregations: Object;
  size: {
    width: number;
  },
  theme: Object,
};

const CancerDistributionChartComponent = compose(
  withSize(),
  withRouter,
  withTheme,
  lifecycle({
    componentDidMount(): void {
      this.props.relay.setVariables({
        fetchFilteredCaseAggs: true,
        caseAggsFilter: this.props.filters,
      });
    },
  })
)(({
  cases,
  size: {
    width,
  },
  theme,
  push,
}: TProps = {}) => {
  const casesByProjectMap = (cases.total || { project__project_id: [] })
    .project__project_id.buckets.reduce((acc, bucket) => ({ ...acc, [bucket.key]: bucket.doc_count }), {});

  const cancerDistData = (cases.filtered || { project__project_id: { buckets: [] } })
  .project__project_id.buckets.map(b => {
    const totalCasesByProject = casesByProjectMap[b.key];
    const freq = b.doc_count / totalCasesByProject;
    return {
      freq,
      project_id: b.key,
      num_affected_cases: b.doc_count,
      num_cases_total: totalCasesByProject,
    };
  });

  const sortedCancerDistData = sortBy(cancerDistData, d => -d.freq);

  const chartData = sortedCancerDistData.map(d => ({
    label: d.project_id,
    value: (d.freq * 100),
    onClick: () => push(`/projects/${d.project_id}`),
    tooltip: (
      <span>
        {d.num_affected_cases.toLocaleString()} Case{d.num_affected_cases > 1 ? 's' : ''} Affected in <b>{d.project_id}</b><br />
        {d.num_affected_cases.toLocaleString()} / {d.num_cases_total.toLocaleString()}&nbsp;({(d.freq * 100).toFixed(2)}%)
      </span>
    ),
  }));

  const chartMargin = { top: 30, right: 50, bottom: 105, left: 40 };

  const maxBars = 20;
  const bandWidth = ((width - chartMargin.right - chartMargin.left) / (maxBars + 1) / 2) * 0.7;

  return (
    <span>{
      chartData.length >= 5 &&
      <Loader
        loading={!cases.filtered}
        height="400px"
      >
        <Column>
          <h5 style={{ textTransform: 'uppercase', padding: '0 2rem' }}>
            {`${sortedCancerDistData.reduce((acc, d) => [...acc, ...d.cases], []).length} cases affected by
            ${sortedCancerDistData.reduce((acc, d) => [...acc, ...d.ssms], []).length} mutations across
            ${sortedCancerDistData.length} projects`}
          </h5>
          <Row style={{ padding: '0 2rem' }}>
            <span>
              <div style={{ textAlign: 'right' }}>
                <DownloadVisualizationButton
                  svg={() => wrapSvg({ selector: '#cancer-distribution svg', title: 'Cancer Distribution' })}
                  data={chartData.map(d => ({ label: d.label, value: d.value }))}
                  slug="bar-chart"
                  noText
                  tooltipHTML="Download image or data"
                />
              </div>
              <BarChart
                margin={chartMargin}
                bandwidth={bandWidth}
                data={chartData}
                yAxis={{ title: '% of Cases Affected' }}
                height={300}
                styles={{
                  xAxis: { stroke: theme.greyScale4, textFill: theme.greyScale3 },
                  yAxis: { stroke: theme.greyScale4, textFill: theme.greyScale3 },
                  bars: { fill: theme.secondary },
                  tooltips: {
                    fill: '#fff',
                    stroke: theme.greyScale4,
                    textFill: theme.greyScale3,
                  },
                }}
              />
            </span>
          </Row>
        </Column>
      </Loader>
    }
    </span>
  );
});

const CancerDistributionChartQuery = {
  initialVariables: {
    caseAggsFilter: null,
    fetchFilteredCaseAggs: false,
    ssmTested: makeFilter([{
      field: 'available_variation_data',
      value: 'ssm',
    }], false),
  },
  fragments: {
    cases: () => Relay.QL`
      fragment on CohortCases {
        filtered: aggregations(filters: $caseAggsFilter) @include(if: $fetchFilteredCaseAggs) {
          project__project_id {
            buckets {
              doc_count
              key
            }
          }
        }
        total: aggregations(filters: $ssmTested) {
          project__project_id {
            buckets {
              doc_count
              key
            }
          }
        }
      }
    `,
  },
};

const CancerDistributionChart = Relay.createContainer(
  CancerDistributionChartComponent,
  CancerDistributionChartQuery
);

export default CancerDistributionChart;
