// @ncigdc/containers/CancerDistributionTable.js
// @flow
import React from 'react';
import Relay from 'react-relay';

const CancerDistributionTableComponent = props => (
  <Column style={{ width: '100%', minWidth: 450 }}>
    <EntityPageHorizontalTable
      headings={[
        { key: 'project_id', title: 'Project ID' },
        { key: 'disease_type', title: 'Disease Type' },
        { key: 'site', title: 'Site' },
        { key: 'num_affected_cases',
          title: (
            <Tooltip
              Component={`Number of Cases where ${gene.symbol} contains SSM`}
            >
              # Affected Cases
            </Tooltip>
          ),
        },
        { key: 'num_mutations',
          title: <Tooltip Component={`Number of SSM observed in ${gene.symbol}`}># Mutations</Tooltip>,
          style: { textAlign: 'right' },
        },
      ]}
      data={sortedCancerDistData.map(d => ({
        ...d,
        project_id: <a href={`/projects/${d.project_id}`}>{d.project_id}</a>,
        num_affected_cases:
          `${d.cases.length}/${casesByProjectMap[d.project_id]} (${(d.freq * 100).toFixed(2)}%)`,
        num_mutations: d.ssms.length,
      }))}
    />
  </Column>
);

const CancerDistributionTableQuery = {
  fragments
}

const CancerDistributionTable = Relay.createContainer(
  CancerDistributionTableComponent,
  CancerDistributionTableQuery
);

export default CancerDistributionTable;
