// @ncigdc/containers/CaseTr.js
/* @flow */

import React from 'react';
import Relay from 'react-relay';

import CaseLink from '@ncigdc/components/Links/CaseLink';
import { findDataCategory } from '@ncigdc/utils/data';

import type { TCategory } from '@ncigdc/utils/data/types';

import { Tr, Td } from '@ncigdc/uikit/Table';

import { withTheme } from '@ncigdc/theme';

export type TProps = {|
  index: number,
  node: {|
    case_id: string,
    demographic: {|
      gender: string,
    |},
    project: {|
      primary_site: string,
      project_id: string,
    |},
    summary: {|
      data_categories: Array<{|
        case_count?: number,
        data_category: TCategory,
      |}>,
      file_count: number,
    |},
  |},
  theme: Object,
|};

export const CaseTrComponent = ({ node, index, theme }: TProps) => (
  <Tr
    style={{
      backgroundColor: index % 2 === 0 ? theme.tableStripe : '#fff',
    }}
  >
    <Td><CaseLink id={node.case_id} merge whitelist={['filters']} /></Td>
    <Td>{node.project.project_id}</Td>
    <Td>{node.project.primary_site}</Td>
    <Td>{node.demographic.gender}</Td>
    <Td>files</Td>
    <Td>{findDataCategory('Seq', node.summary.data_categories).file_count}</Td>
    <Td>{findDataCategory('Exp', node.summary.data_categories).file_count}</Td>
    <Td>{findDataCategory('SNV', node.summary.data_categories).file_count}</Td>
    <Td>{findDataCategory('CNV', node.summary.data_categories).file_count}</Td>
    <Td>{findDataCategory('Clinical', node.summary.data_categories).file_count}</Td>
    <Td>{findDataCategory('Bio', node.summary.data_categories).file_count}</Td>
    <Td>annotations</Td>
  </Tr>
);

export const CaseTrQuery = {
  fragments: {
    node: () => Relay.QL`
      fragment on Case {
        case_id
        project {
          project_id
          primary_site
        }
        demographic {
          gender
        }
        summary {
          file_count
          data_categories {
            file_count
            data_category
          }
        }
      }
    `,
  },
};

const CaseTr = Relay.createContainer(
  withTheme(CaseTrComponent),
  CaseTrQuery
);

export default CaseTr;
