// @ncigdc/containers/FileTable.js
/* @flow */

import React from 'react';
import Relay from 'react-relay';

import Pagination from '@ncigdc/components/Pagination';
import Showing from '@ncigdc/components/Pagination/Showing';
import AddToCartButtonAll from '@ncigdc/components/AddToCartButtonAll';
import { Row } from '@ncigdc/uikit/Flex';
import TableActions from '@ncigdc/components/TableActions';

import { Tr, Th } from '@ncigdc/uikit/Table';

import FileTr from './FileTr';

import type { TTableProps } from './types';

const styles = {
  table: {
    width: '100%',
    borderCollapse: 'collapse',
    borderSpacing: 0,
  },
};

export const FileTableComponent = ({downloadable, hits, relay, canAddToCart = true, tableHeader}: TTableProps) => {
  const prefix = 'files';

  return (
    <div>

      <Row style={{ backgroundColor: 'white', padding: '1rem', justifyContent: 'space-between' }}>
        { tableHeader &&
          <h3 className='panel-title'>
            {tableHeader}
          </h3>
        }
        <Showing
          docType="files"
          prefix={prefix}
          params={relay.route.params}
          total={hits.total}
        />
        <TableActions
          prefix={prefix}
          total={hits.total}
          sortKey="files_sort"
          endpoint={`${API}/files`}
          downloadable={downloadable}
          downloadFields={[
            'file_id',
            'file_name',
            'cases.project.project_id',
            'data_category',
            'data_format',
            'file_size',
          ]}
          sortOptions={[
            {
              id: 'file_id.raw',
              name: 'File UUID',
            },
            {
              id: 'submitter_id.raw',
              name: 'File Submitter ID',
            },
            {
              id: 'access',
              name: 'Access',
            },
            {
              id: 'file_name.raw',
              name: 'File Name',
            },
            {
              id: 'cases.project.project_id',
              name: 'Project',
            },
            {
              id: 'data_category',
              name: 'Data Category',
            },
            {
              id: 'data_format',
              name: 'Data Format',
            },
            {
              id: 'file_size',
              name: 'Size',
            },
          ]}
        />
      </Row>
      <div style={{ overflowX: 'auto' }}>
        <table style={styles.table}>
          <thead>
            <Tr>
              <Th>
                { canAddToCart &&
                  <AddToCartButtonAll
                    edges={hits.edges.map(e => e.node)}
                    total={hits.total}
                  />
                }
                { !canAddToCart &&
                  <span>Action</span>
                }
              </Th>
              <Th>Access</Th>
              <Th>File Name</Th>
              <Th>Cases</Th>
              <Th>Projects</Th>
              <Th>Category</Th>
              <Th>Format</Th>
              <Th>Size</Th>
            </Tr>
          </thead>
          <tbody>
            {hits.edges.map((e, i) => (
              <FileTr
                node={e.node}
                key={e.node.id}
                index={i}
                canAddToCart={canAddToCart}
              />
          ))}
          </tbody>
        </table>
      </div>
      <Pagination prefix={prefix} params={relay.route.params} total={hits.total} />
    </div>
  );
};

export const FileTableQuery = {
  fragments: {
    hits: () => Relay.QL`
      fragment on FileConnection {
        total
        edges {
          node {
            id
            file_id
            file_name
            file_size
            access
            data_category
            data_format
            cases {
              hits(first: 99) {
                total
                edges {
                  node {
                    project {
                      project_id
                    }
                  }
                }
              }
            }
          }
        }
      }
    `,
  },
};

const FileTable = Relay.createContainer(
  FileTableComponent,
  FileTableQuery
);

export default FileTable;
