// @ncigdc/containers/FilePage.js
/* @flow */

import React from 'react';
import Relay from 'react-relay';

import SideNavLayout from '@ncigdc/components/Layouts/SideNavLayout';
import File from '@ncigdc/components/File';

export type TProps = {
  node: {
    access: string,
    cases: Array<{
      project: {
        project_id: string,
      },
    }>,
    data_category: string,
    data_format: string,
    file_id: string,
    file_name: string,
    file_size: number,
    platform: string,
  },
};

export const FilePageComponent = (props: TProps) => (
  <SideNavLayout
    links={[
      { icon: 'table', id: 'summary', title: 'Summary' },
    ]}
    title={props.node.file_id}
    entityType={<i className="fa fa-file-o" />}
  >
    <File node={props.node} />
  </SideNavLayout>
);

export const FilePageQuery = {
  fragments: {
    node: () => Relay.QL`
      fragment on File {
        file_id
        submitter_id
        file_id
        file_name
        file_size
        access
        data_category
        data_format
        data_type
        experimental_strategy
        md5sum
        state
        tags
        center {
          short_name
        }
        annotations {
          annotation_id
        }
        cases {
          project {
            project_id
          }
        }
        associated_entities {
          case_id
          entity_id
          entity_type
        }
        metadata_files {
          file_id
          file_name
          data_category
          data_type
          data_format
          file_size
        }
        archive {
          archive_id
          submitter_id
          revision
        }
        platform
      }
    `,
  },
};

const FilePage = Relay.createContainer(
  FilePageComponent,
  FilePageQuery
);

export default FilePage;
