// @ncigdc/containers/FileTr.js
/* @flow */

import React from 'react';
import Relay from 'react-relay';

import FileLink from '@ncigdc/components/Links/FileLink';
import FileSize from '@ncigdc/components/FileSize';
import AddToCartButtonSingle from '@ncigdc/components/AddToCartButtonSingle';

import { Tr, Td } from '@ncigdc/uikit/Table';

import { withTheme } from '@ncigdc/theme';

export type TProps = {
  index: number,
  node: {
    access: string,
    cases: Array<{
      project: {
        project_id: string,
      },
    }>,
    data_category: string,
    data_format: string,
    file_id: string,
    file_name: string,
    file_size: number,
  },
  theme: Object,
};

export const FileTrComponent = ({ node, index, theme }: TProps) => (
  <Tr
    style={{
      backgroundColor: index % 2 === 0 ? theme.tableStripe : '#fff',
    }}
  >
    <Td><AddToCartButtonSingle file={node} /></Td>
    <Td>{node.access}</Td>
    <Td>
      <FileLink id={node.file_id}>
        {node.file_name}
      </FileLink>
    </Td>
    <Td>{node.cases.length}</Td>
    <Td>{new Set(node.cases.map(c => c.project.project_id))}</Td>
    <Td>{node.data_category}</Td>
    <Td>{node.data_format}</Td>
    <Td><FileSize bytes={node.file_size} /></Td>
  </Tr>
);

export const FileTrQuery = {
  fragments: {
    node: () => Relay.QL`
      fragment on RepositoryFile {
        file_id
        file_name
        file_size
        access
        data_category
        data_format
        cases {
          project {
            project_id
          }
        }
      }
    `,
  },
};

const FileTr = Relay.createContainer(
  withTheme(FileTrComponent),
  FileTrQuery
);

export default FileTr;
