// @ncigdc/containers/FileTr.js
/* @flow */

import React from 'react';
import Relay from 'react-relay';

import FileLink from '@ncigdc/components/Links/FileLink';
import FileSize from '@ncigdc/components/FileSize';
import AddToCartButtonSingle from '@ncigdc/components/AddToCartButtonSingle';

import { Tr, Td } from '@ncigdc/uikit/Table';

import { withTheme } from '@ncigdc/theme';

export type TProps = {
  index: number,
  node: {
    access: string,
    cases: Array<{
      project: {
        project_id: string,
      },
    }>,
    data_category: string,
    data_format: string,
    file_id: string,
    file_name: string,
    file_size: number,
  },
  theme: Object,
};

export const FileTrComponent = ({ node, index, theme }: TProps) => (
  <Tr
    style={{
      backgroundColor: index % 2 === 0 ? theme.tableStripe : '#fff',
    }}
  >
    <Td><AddToCartButtonSingle file={node} /></Td>
    <Td>
      {node.access === 'open' &&
        <i className='fa fa-unlock-alt' />
      }
      {node.access === 'controlled' &&
        <i className='fa fa-lock' />
      }
      <span
        style={{
          marginLeft: '0.3rem',
        }}
      >
        {node.access}
      </span>
    </Td>
    <Td>
      <FileLink id={node.file_id}>
        {node.file_name}
      </FileLink>
    </Td>
    <Td>{node.cases.length}</Td>
    <Td>{new Set(node.cases.hits.edges.map(e => e.node.project.project_id))}</Td>
    <Td>{node.data_category}</Td>
    <Td>{node.data_format}</Td>
    <Td><FileSize bytes={node.file_size} /></Td>
  </Tr>
);

export default withTheme(FileTrComponent);
