// @ncigdc/containers/FrequentlyMutatedGenesChart.js
// @flow

import React from 'react';
import Relay from 'react-relay';
import withSize from 'react-sizeme';
import { lifecycle, compose } from 'recompose';
import withRouter from '@ncigdc/utils/withRouter';
import { parseFilterParam } from '@ncigdc/utils/uri';
import Loader from '@ncigdc/uikit/Loaders/Loader';
import { withTheme } from '@ncigdc/theme';
import { graphTitle } from '@ncigdc/theme/mixins';
import { Row } from '@ncigdc/uikit/Flex';
import DownloadVisualizationButton from '@ncigdc/components/DownloadVisualizationButton';
import BarChart from '@ncigdc/components/Charts/BarChart';

const FrequentlyMutatedGenesChartComponent = compose(
  withRouter,
  lifecycle({
    componentDidMount() {
      this.props.relay.setVariables({
        fetchData: true,
        fmgChart_filters: parseFilterParam(
          this.props.query.fmgChart_filters,
          this.props.defaultFilters || null
        ),
      });
    },
  }),
  withTheme,
  withSize()
)(({
  numCasesAggByProject,
  projectId = '',
  theme,
  size: {
    width,
  },
  cohort: { genes, cases },
  push,
  showSurvivalPlot = true,
}) => {
  const chartMargin = { top: 30, right: 50, bottom: 105, left: 40 };

  const maxBars = 20;
  const bandWidth = (
    (width - chartMargin.right - chartMargin.left) /
    (maxBars + 1) / (showSurvivalPlot ? 1 : 2)
  ) * 0.7;

  const mutatedGenesChartData = !!genes && genes.hits.edges.map(x => x.node);

  return (
    <Loader
      loading={!genes}
      height="400px"
    >
      {!!mutatedGenesChartData &&
        <Row>
          <span>
            <div style={{ textAlign: 'right', marginRight: 50, marginLeft: 30 }}>
              <DownloadVisualizationButton
                disabled={!mutatedGenesChartData.length}
                svg="#mutated-genes-chart svg"
                data={mutatedGenesChartData}
                slug="bar-chart"
                tooltipHTML="Download image or data"
                noText
              />
            </div>
            <div style={graphTitle()}>Distribution of Most Frequently Mutated Genes</div>
            {!!mutatedGenesChartData.length &&
              <div id="mutated-genes-chart">
                <Row style={{ padding: '2rem 2rem 0' }}>
                  <BarChart
                    data={
                      mutatedGenesChartData
                        .map(g => ({
                          label: g.symbol,
                          value: projectId
                            ? ((g.filteredCases.hits.total / numCasesAggByProject[projectId]) * 100)
                            : (g.allCases.hits.total / cases.hits.total) * 100,
                          tooltip: projectId
                          ? (
                            <span>
                              <b>{g.symbol}</b><br />
                              {g.filteredCases.hits.total} Case{g.filteredCases.hits.total > 1 ? 's' : ''}
                              &nbsp;affected in {projectId}<br />
                              {g.filteredCases.hits.total} / {numCasesAggByProject[projectId]}
                              &nbsp;({((g.filteredCases.hits.total / numCasesAggByProject[projectId]) * 100)
                              .toFixed(2)}%)
                            </span>
                          )
                          : (
                            <span><b>{g.symbol}</b><br />
                              <b>{g.allCases.hits.total} Case{g.allCases.hits.total > 1 ? 's ' : ' '}
                                &nbsp;affected in all projects</b><br />
                              <b>{g.allCases.hits.total} / {cases.hits.total}
                              &nbsp;({((g.allCases.hits.total / cases.hits.total) * 100).toFixed(2)}%)</b>
                            </span>
                          ),
                          onClick: () => push(`/genes/${g.gene_id}`),
                        }))
                    }
                    yAxis={{ title: '% of Cases Affected' }}
                    bandwidth={bandWidth}
                    margin={chartMargin}
                    height={300}
                    styles={{
                      xAxis: { stroke: theme.greyScale4, textFill: theme.greyScale3 },
                      yAxis: { stroke: theme.greyScale4, textFill: theme.greyScale3 },
                      bars: { fill: theme.secondary },
                      tooltips: {
                        fill: '#fff',
                        stroke: theme.greyScale4,
                        textFill: theme.greyScale3,
                      },
                    }}
                  />
                </Row>
              </div>
            }
          </span>
        </Row>
      }
    </Loader>
  );
});

export const FrequentlyMutatedGenesChartQuery = {
  initialVariables: {
    fetchData: false,
    fmgChart_filters: null,
    score: 'case.project.project_id',
  },
  fragments: {
    cohort: () => Relay.QL`
      fragment on Cohort {
        cases {
          hits(first: 0) { total }
        }
        genes @include(if: $fetchData) {
          hits (first: 20 filters: $fmgChart_filters, score: $score) @include(if: $fetchData) {
            total
            edges {
              node {
                score
                symbol
                gene_id
                filteredCases: case {
                  hits(first: 1 filters: $fmgChart_filters) {
                    total
                  }
                }
                allCases: case {
                  hits(first: 1) {
                    total
                  }
                }
              }
            }
          }
        }
      }
    `,
  },
};

const FrequentlyMutatedGenesChart = Relay.createContainer(
  FrequentlyMutatedGenesChartComponent,
  FrequentlyMutatedGenesChartQuery
);

export default FrequentlyMutatedGenesChart;
