// @ncigdc/containers/FrequentMutationsTable.js
// @flow

import React from 'react';
import Relay from 'react-relay';
import withSize from 'react-sizeme';
import { startCase, truncate, isEqual } from 'lodash';
import { lifecycle, compose } from 'recompose';
import { scaleOrdinal, schemeCategory10 } from 'd3';
import { parseIntParam, parseFilterParam } from '@ncigdc/utils/uri';
import Showing from '@ncigdc/components/Pagination/Showing';
import { withTheme } from '@ncigdc/theme';
import { tableToolTipHint } from '@ncigdc/theme/mixins';
import GeneLink from '@ncigdc/components/Links/GeneLink';
import withRouter from '@ncigdc/utils/withRouter';
import MutationLink from '@ncigdc/components/Links/MutationLink';
import SpinnerParticle from '@ncigdc/uikit/Loaders/Particle';

import EntityPageHorizontalTable from '@ncigdc/components/EntityPageHorizontalTable';
import SurvivalIcon from '@ncigdc/theme/icons/SurvivalIcon';
import { Row } from '@ncigdc/uikit/Flex';
import TogglableUl from '@ncigdc/uikit/TogglableUl';
import { Tooltip } from '@ncigdc/uikit/Tooltip';
import Pagination from '@ncigdc/components/Pagination';
import { getSurvivalCurves } from '@ncigdc/utils/survivalplot';
import Hidden from '@ncigdc/components/Hidden';

const impactBubble = {
  color: 'white',
  padding: '2px 5px',
  borderRadius: '8px',
  fontSize: '10px',
  fontWeight: 'bold',
  display: 'inline-block',
  width: '20px',
};

const colors = scaleOrdinal(schemeCategory10);

const impactColors = {
  HIGH: 'rgb(185, 36, 36)',
  MODERATE: 'rgb(193, 158, 54)',
  LOW: 'rgb(49, 161, 60)',
};

const styles = {
  impact: {
    HIGH: {
      ...impactBubble,
      backgroundColor: impactColors.HIGH,
    },
    MODERATE: {
      ...impactBubble,
      backgroundColor: impactColors.MODERATE,
    },
    LOW: {
      ...impactBubble,
      backgroundColor: impactColors.LOW,
    },
  },
};

const mapData = (data: Array<Object>, projectId: string): Array<Object> => (
  data.map((hit) => {
    const { transcript } = hit.consequence.hits.edges.find(c => c.node.transcript.is_canonical).node;
    const { annotation = {}, consequence_type = '', gene = {}, aa_change } = transcript;
    const { symbol } = gene;
    const impact = annotation.impact;

    return {
      ...hit,
      score: hit.score,
      allOccurrences: hit.allOccurrences.hits.total,
      filteredOccurrences: hit.filteredOccurrences.hits.total,
      num_affected_cases_by_project: hit.allOccurrences.hits.edges.reduce((acc, o) => ({
        ...acc,
        [o.node.case.project.project_id]: acc[o.node.case.project.project_id] ? acc[o.node.case.project.project_id] + 1 : 1,
      }), {}),
      impact,
      consequence_type: (
        <span>
          <b>{startCase(consequence_type.replace('variant', ''))}</b>
          <span style={{ marginLeft: '5px' }}>
            <GeneLink id={symbol}>
              {symbol}
            </GeneLink>
          </span>
          <span
            style={{
              marginLeft: '5px',
              color: impactColors[impact] || 'inherit',
            }}
          >
            {truncate(aa_change)}
          </span>
        </span>
      ),
    };
  })
);

const FrequentMutationsTableComponent = compose(
  withRouter,
  lifecycle({
    componentDidMount() {
      this.props.relay.setVariables({
        fetchData: true,
        fmTable_offset: parseIntParam(this.props.query.fmTable_offset, 0),
        fmTable_size: parseIntParam(this.props.query.fmTable_size, 20),
        fmTable_filters: parseFilterParam(this.props.query.fmTable_filters, this.props.defaultFilters || null),
      });
    },
    componentWillReceiveProps(nextProps) {
      if (!isEqual(this.props.query, nextProps.query)) {
        this.props.relay.setVariables({
          fmTable_offset: parseIntParam(nextProps.query.fmTable_offset, 0),
          fmTable_size: parseIntParam(nextProps.query.fmTable_size, 20),
          fmTable_filters: parseFilterParam(nextProps.query.fmTable_filters, nextProps.defaultFilters || null),
        })
      }
    }
  }),
  withTheme,
  withSize()
)(({
  numCasesAggByProject = {},
  projectId = '',
  showSurvivalPlot = false,
  selectedSurvivalData = {},
  setSelectedSurvivalData = () => {},
  cohort: { ssms, cases },
  relay,
} = {}) => {
  if (!ssms) {
    return (
      <Row style={{ justifyContent: 'center', padding: '1rem' }}>
        <SpinnerParticle />
      </Row>
    );
  }

  if (!ssms.hits.edges.length) {
    return <Row style={{ padding: '1rem' }}>No mutation data found.</Row>;
  }

  const frequentMutations = mapData(ssms.hits.edges.map(x => x.node), projectId);

  return (
    <div>
      <Row style={{ backgroundColor: 'white', padding: '1rem', justifyContent: 'space-between' }}>
        <Showing
          docType="mutations"
          prefix={'fmTable'}
          params={relay.route.params}
          total={ssms.hits.total}
        />
      </Row>
      <EntityPageHorizontalTable
        headings={[
          {
            key: 'mutation_uuid',
            title: 'ID',
          },
          {
            key: 'genomic_dna_change',
            title: (
              <Tooltip
                Component={
                  <span>
                    Genomic DNA change, shown as <br />
                    {'{chromosome}:g{start}{ref}>{tumor}'}
                  </span>
                }
                style={tableToolTipHint()}
              >
                DNA Change
              </Tooltip>
            ),
            className: 'id-cell',
            style: { whiteSpace: 'pre-line' },
          },
          { key: 'mutation_subtype', title: 'Type' },
          { key: 'consequence_type', title: 'Consequences' },
          ...(projectId ? [{
            key: 'filteredOccurrences',
            title: (
              <Tooltip
                Component={
                  <span>
                    Breakdown of affected cases in {projectId} <br />
                    # of cases affected / # SSM tested cases
                  </span>
                }
                style={tableToolTipHint()}
              >
                # Affected Cases<br />in {projectId}
              </Tooltip>
            ),
          }] : []),
          {
            key: 'allOccurrences',
            title: (
              <Tooltip
                Component={
                  <span>
                    # of cases where mutation is observed, filtered by current criteria <br />
                    / # SSM tested cases portal wide.<br />
                    Expand to see breakdown by project.
                  </span>
                }
                style={tableToolTipHint()}
              >
                # Affected Cases<br /> (All Projects)
              </Tooltip>
            ),
          },
          {
            key: 'impact',
            title: 'Impact',
            style: { textAlign: 'center' },
          },
          ...(showSurvivalPlot ? [{
            title: 'Survival Analysis',
            key: 'survival_plot',
            style: { textAlign: 'center', width: '100px' },
          }] : []),
        ]}
        data={frequentMutations.map(x => ({
          ...x,
          mutation_uuid: (
            <MutationLink id={x.ssm_id}>
              <Tooltip
                Component={
                  <span>
                    {x.ssm_id}
                  </span>
                }
              >
                {x.ssm_id.substr(0, 8)}...
              </Tooltip>
            </MutationLink>
          ),
          genomic_dna_change: <span>{x.genomic_dna_change}</span>,
          ...(projectId ? { filteredOccurrences:
            `${x.filteredOccurrences.toLocaleString()} / ${numCasesAggByProject[projectId].toLocaleString()}`
          + ` (${((x.filteredOccurrences / numCasesAggByProject[projectId]) * 100).toFixed(2)}%)`,
          } : {}),
          allOccurrences: (
            <TogglableUl
              items={[
                /*  TODO: Commented out as part of PRTL-683
                    Should be revert back once Exploration page is ready
                <CohortLink
                  query={{
                    searchTableTab: 'cases',
                    filters: makeFilter([{ field: 'ssms.ssm_id', value: [x.ssm_id] }], false),
                  }}
                >
                  {x.num_affected_cases_all} ({((x.num_affected_cases_all / totalNumCases) * 100).toFixed(2)}%)
                </CohortLink>,*/
                <span>
                  {x.allOccurrences.toLocaleString()} / {cases.hits.total.toLocaleString()}
                  &nbsp;({((x.allOccurrences / cases.hits.total) * 100).toFixed(2)}%)
                </span>,
                ...Object.entries(x.num_affected_cases_by_project)
                  .map(([k, v]) => `${k}: ${v.toLocaleString()} (${((v / cases.hits.total) * 100).toFixed(2)}%)`),
              ]}
            />
          ),
          impact: !['LOW', 'MODERATE', 'HIGH'].includes(x.impact) ? null : (
            <Tooltip Component={x.impact}>
              <span
                style={styles.impact[x.impact]}
              >
                {x.impact.slice(0, 1)}
              </span>
            </Tooltip>
          ),
          ...(showSurvivalPlot ? {
            survival_plot: (
              <Tooltip Component={`Click icon to plot ${x.genomic_dna_change}`}>
                <button
                  onClick={() => {
                    if (x.ssm_id !== selectedSurvivalData.id) {
                      getSurvivalCurves({
                        field: 'gene.ssm.ssm_id',
                        value: x.ssm_id,
                        slug: x.ssm_id.substr(0, 8),
                        currentFilters: projectId ? { op: 'and', content: [{ op: '=', content: { field: 'cases.project.project_id', value: projectId } }] } : null,
                      })
                        .then(setSelectedSurvivalData);
                    } else {
                      setSelectedSurvivalData({});
                    }
                  }}
                >
                  <span
                    style={{
                      color: colors(selectedSurvivalData.id === x.ssm_id ? 1 : 0),
                      cursor: 'pointer',
                    }}
                  >
                    <SurvivalIcon />
                    <Hidden>add to survival plot</Hidden>
                  </span>
                </button>
              </Tooltip>
            ),
          } : {}),
        }))}
      />
      <Pagination prefix={'fmTable'} params={relay.route.params} total={ssms.hits.total} />
    </div>
  );
});


export const FrequentMutationsTableQuery = {
  initialVariables: {
    fetchData: false,
    score: 'occurrence.case.project.project_id',
    fmTable_filters: null,
    fmTable_size: 20,
    fmTable_offset: 0,
  },
  fragments: {
    cohort: () => Relay.QL`
      fragment on Cohort {
        cases {
          hits(first: 0) { total }
        }
        ssms @include (if: $fetchData) {
          hits(first: $fmTable_size offset: $fmTable_offset filters: $fmTable_filters, score: $score) {
            total
            edges {
              node {
                score
                genomic_dna_change
                mutation_subtype
                ssm_id
                consequence {
                  hits(first: 99) {
                    edges {
                      node {
                        transcript {
                          is_canonical
                          annotation {
                            impact
                          }
                          consequence_type
                          gene {
                            gene_id
                            symbol
                          }
                          aa_change
                        }
                      }
                    }
                  }
                }
                filteredOccurrences: occurrence {
                  hits(first: 1 filters: $fmTable_filters) {
                    total
                  }
                }
                allOccurrences: occurrence {
                  hits(first: 1000) {
                    total
                    edges {
                      node {
                        case {
                          project { project_id }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    `,
  },
};

const FrequentMutationsTable = Relay.createContainer(
  FrequentMutationsTableComponent,
  FrequentMutationsTableQuery
);

export default FrequentMutationsTable;
