// @ncigdc/containers/FrequentMutationsTable.js
// @flow

import React from 'react';
import Relay from 'react-relay';
import withSize from '@ncigdc/utils/withSize';
import { startCase, truncate, isEqual, get } from 'lodash';
import { lifecycle, compose, withState } from 'recompose';
import { scaleOrdinal, schemeCategory10 } from 'd3';
import { parseIntParam, parseFilterParam } from '@ncigdc/utils/uri';
import Showing from '@ncigdc/components/Pagination/Showing';
import { withTheme } from '@ncigdc/theme';
import { tableToolTipHint } from '@ncigdc/theme/mixins';
import Loader from '@ncigdc/uikit/Loaders/Loader';
import GeneLink from '@ncigdc/components/Links/GeneLink';
import withRouter from '@ncigdc/utils/withRouter';
import MutationLink from '@ncigdc/components/Links/MutationLink';
import { makeFilter } from '@ncigdc/utils/filters';
import EntityPageHorizontalTable from '@ncigdc/components/EntityPageHorizontalTable';
import SurvivalIcon from '@ncigdc/theme/icons/SurvivalIcon';
import { Row } from '@ncigdc/uikit/Flex';
import TogglableUl from '@ncigdc/uikit/TogglableUl';
import { Tooltip } from '@ncigdc/uikit/Tooltip';
import Pagination from '@ncigdc/components/Pagination';
import { getSurvivalCurves } from '@ncigdc/utils/survivalplot';
import Hidden from '@ncigdc/components/Hidden';
import BubbleIcon from '@ncigdc/theme/icons/BubbleIcon';
import { SpinnerIcon } from '@ncigdc/theme/icons';
import type { TGroupFilter } from '@ncigdc/utils/filters/types';
import type { TTheme } from '@ncigdc/theme';

type TProps = {
  projectId: string,
  showSurvivalPlot: boolean,
  selectedSurvivalData: {
    id: string,
  },
  setSelectedSurvivalData: Function,
  cohort: {
    ssms: {
      hits: {
        total: number,
        edges: Array<{
          node: {},
        }>,
      },
    },
    cases: {
      hits: {
        total: number,
      },
      aggregations: {
        project__project_id: {
          buckets: Array<{
            doc_count: number,
            key: string,
          }>,
        },
      },
    },
  },
  shouldShowGeneSymbol: boolean,
  relay: {
    route: {
      params: {},
    },
  },
  setSurvivalLoadingId: Function,
  survivalLoadingId: string,
  theme: TTheme,
  query: {
    fmTable_offset: string,
    fmTable_size: string,
    fmTable_filters: string,
  },
  defaultFilters: TGroupFilter,
};
const colors = scaleOrdinal(schemeCategory10);

const mutationSubTypeMap = {
  'single base substitution': 'Substitution',
  'small deletion': 'Deletion',
  'small insertion': 'Insertion',
};

const mapData = (data: Array<Object>, projectId: string, shouldShowGeneSymbol: boolean, theme): Array<Object> => (
  data.map((hit) => {
    const consequenceOfInterest = hit.consequence.hits.edges
      .find(consequence => get(consequence, 'node.transcript.annotation.impact'), {});
    const { transcript } = consequenceOfInterest.node || {};
    const { annotation = {}, consequence_type: consequenceType = '', gene = {}, aa_change } = transcript || {};
    const { symbol, gene_id: geneId } = gene;
    const impact = annotation.impact;

    return {
      ...hit,
      score: hit.score,
      allOccurrences: hit.allOccurrences.hits.total,
      filteredOccurrences: hit.filteredOccurrences.hits.total,
      num_affected_cases_by_project: hit.allOccurrences.hits.edges.reduce((acc, { node: { case: { project } } }) => ({
        ...acc,
        [project.project_id]: acc[project.project_id]
          ? acc[project.project_id] + 1
          : 1,
      }), {}),
      impact,
      mutation_subtype: mutationSubTypeMap[(hit.mutation_subtype || '').toLowerCase()] || hit.mutation_subtype,
      consequence_type: (
        <span>
          <b>{startCase(consequenceType.replace('variant', ''))}</b>
          {shouldShowGeneSymbol &&
            <span style={{ marginLeft: '5px' }}>
              <GeneLink id={geneId}>{symbol}</GeneLink>
            </span>
          }
          <span
            style={{
              marginLeft: '5px',
              color: theme.impacts[impact] || 'inherit',
            }}
          >
            {truncate(aa_change)}
          </span>
        </span>
      ),
    };
  })
);

const FrequentMutationsTableComponent = compose(
  withRouter,
  withState('survivalLoadingId', 'setSurvivalLoadingId', ''),
  lifecycle({
    componentDidMount(): void {
      this.props.relay.setVariables({
        fetchData: true,
        fmTable_offset: parseIntParam(this.props.query.fmTable_offset, 0),
        fmTable_size: parseIntParam(this.props.query.fmTable_size, 10),
        fmTable_filters: parseFilterParam(this.props.query.fmTable_filters, this.props.defaultFilters || null),
      });
    },
    componentWillReceiveProps(nextProps: TProps): void {
      if (!isEqual(this.props.query, nextProps.query)) {
        this.props.relay.setVariables({
          fmTable_offset: parseIntParam(nextProps.query.fmTable_offset, 0),
          fmTable_size: parseIntParam(nextProps.query.fmTable_size, 10),
          fmTable_filters: parseFilterParam(nextProps.query.fmTable_filters, nextProps.defaultFilters || null),
        });
      }
    },
  }),
  withTheme,
  withSize()
)(({
  projectId = '',
  showSurvivalPlot = false,
  selectedSurvivalData = { id: '' },
  setSelectedSurvivalData = () => {},
  cohort: { ssms, cases },
  shouldShowGeneSymbol = true,
  relay,
  setSurvivalLoadingId,
  survivalLoadingId,
  theme,
}: TProps = {}) => {
  if (ssms && !ssms.hits.edges.length) {
    return <Row style={{ padding: '1rem' }}>No mutation data found.</Row>;
  }

  const numCasesAggByProject = cases.aggregations.project__project_id.buckets.reduce((acc, b) => ({
    ...acc,
    [b.key]: b.doc_count,
  }), {});

  const frequentMutations = !ssms ? [] :
    mapData(ssms.hits.edges.map(x => x.node), projectId, shouldShowGeneSymbol, theme);

  return (
    <Loader
      loading={!ssms}
      height="387px"
    >
      <Row style={{ backgroundColor: 'white', padding: '1rem', justifyContent: 'space-between' }}>
        <Showing
          docType="mutations"
          prefix={'fmTable'}
          params={relay.variables}
          total={!ssms ? 0 : ssms.hits.total}
        />
      </Row>
      <EntityPageHorizontalTable
        headings={[
          {
            key: 'mutation_uuid',
            title: 'ID',
          },
          {
            key: 'genomic_dna_change',
            title: (
              <Tooltip
                Component={
                  <span>
                    Genomic DNA change, shown as <br />
                    {'{chromosome}:g{start}{ref}>{tumor}'}
                  </span>
                }
                style={tableToolTipHint()}
              >
                DNA Change
              </Tooltip>
            ),
            className: 'id-cell',
            style: { whiteSpace: 'pre-line' },
          },
          { key: 'mutation_subtype', title: 'Type' },
          { key: 'consequence_type', title: 'Consequences' },
          ...(projectId ? [{
            key: 'filteredOccurrences',
            title: (
              <Tooltip
                Component={
                  <span>
                    Breakdown of affected cases in {projectId} <br />
                    # of cases affected / # SSM tested cases
                  </span>
                }
                style={tableToolTipHint()}
              >
                # Affected Cases<br />in {projectId}
              </Tooltip>
            ),
          }] : []),
          {
            key: 'allOccurrences',
            title: (
              <Tooltip
                Component={
                  <span>
                    # of cases where mutation is observed, filtered by current criteria <br />
                    / # SSM tested cases portal wide.<br />
                    Expand to see breakdown by project.
                  </span>
                }
                style={tableToolTipHint()}
              >
                # Affected Cases<br /> Across all Projects
              </Tooltip>
            ),
          },
          {
            key: 'impact',
            title: <span>Impact<br />(VEP)</span>,
            style: { textAlign: 'center' },
          },
          ...(showSurvivalPlot ? [{
            title: <span>Survival<br />Analysis</span>,
            key: 'survival_plot',
            style: { textAlign: 'center', width: '100px' },
          }] : []),
        ]}
        data={frequentMutations.map(x => ({
          ...x,
          mutation_uuid: (
            <MutationLink id={x.ssm_id}>
              <Tooltip
                Component={
                  <span>
                    {x.ssm_id}
                  </span>
                }
              >
                {x.ssm_id.substr(0, 8)}...
              </Tooltip>
            </MutationLink>
          ),
          genomic_dna_change: <span>{x.genomic_dna_change}</span>,
          ...(projectId ? { filteredOccurrences:
            `${x.filteredOccurrences.toLocaleString()} / ${numCasesAggByProject[projectId].toLocaleString()}`
          + ` (${((x.filteredOccurrences / numCasesAggByProject[projectId]) * 100).toFixed(2)}%)`,
          } : {}),
          allOccurrences: (
            <TogglableUl
              items={[
                /*  TODO: Commented out as part of PRTL-683
                    Should be revert back once Exploration page is ready
                <CohortLink
                  query={{
                    searchTableTab: 'cases',
                    filters: makeFilter([{ field: 'ssms.ssm_id', value: [x.ssm_id] }], false),
                  }}
                >
                  {x.num_affected_cases_all} ({((x.num_affected_cases_all / totalNumCases) * 100).toFixed(2)}%)
                </CohortLink>,*/
                <span key="first">
                  {x.allOccurrences.toLocaleString()} / {cases.hits.total.toLocaleString()}
                  &nbsp;({((x.allOccurrences / cases.hits.total) * 100).toFixed(2)}%)
                </span>,
                ...Object.entries(x.num_affected_cases_by_project) // eslint-disable-line fp/no-mutating-methods
                  .sort(([ak, av], [bk, bv]) => (bv / numCasesAggByProject[bk]) - (av / numCasesAggByProject[ak]))
                  .map(([k, v]) => (
                    <span key={k}>
                      {k}: {v} / {numCasesAggByProject[k]}
                      &nbsp;({((v / numCasesAggByProject[k]) * 100).toFixed(2)}%)
                    </span>
                  )),
              ]}
            />
          ),
          impact: !['LOW', 'MODERATE', 'HIGH', 'MODIFIER'].includes(x.impact) ? null : (
            <BubbleIcon
              toolTipText={x.impact}
              text={x.impact.slice(0, (x.impact === 'MODIFIER' ? 2 : 1))}
              backgroundColor={theme.impacts[x.impact]}
            />
          ),
          ...(showSurvivalPlot ? {
            survival_plot: (
              <Tooltip Component={`Click icon to plot ${x.genomic_dna_change}`}>
                <button
                  style={{ padding: '2px 3px' }}
                  onClick={() => {
                    if (x.ssm_id !== selectedSurvivalData.id) {
                      setSurvivalLoadingId(x.ssm_id);
                      getSurvivalCurves({
                        field: 'gene.ssm.ssm_id',
                        value: x.ssm_id,
                        slug: x.ssm_id.substr(0, 8),
                        currentFilters: projectId
                          ? {
                            op: 'and',
                            content: [{ op: '=', content: { field: 'cases.project.project_id', value: projectId } }],
                          }
                          : null,
                      })
                        .then((data) => {
                          setSelectedSurvivalData(data);
                          setSurvivalLoadingId('');
                        });
                    } else {
                      setSelectedSurvivalData({});
                    }
                  }}
                >
                  <span
                    style={{
                      color: colors(selectedSurvivalData.id === x.ssm_id ? 1 : 0),
                      cursor: 'pointer',
                    }}
                  >
                    {survivalLoadingId === x.ssm_id ? <SpinnerIcon /> : <SurvivalIcon />}
                    <Hidden>add to survival plot</Hidden>
                  </span>
                </button>
              </Tooltip>
            ),
          } : {}),
        }))}
      />
      <Pagination
        prefix={'fmTable'}
        params={relay.variables}
        total={!ssms ? 0 : ssms.hits.total}
      />
    </Loader>
  );
});


export const FrequentMutationsTableQuery = {
  initialVariables: {
    fetchData: false,
    score: 'occurrence.case.project.project_id',
    fmTable_filters: null,
    fmTable_size: 10,
    fmTable_offset: 0,
    consequenceFilters: {
      op: 'NOT',
      content: {
        field: 'consequence.transcript.annotation.impact',
        value: 'missing',
      },
    },
    ssmTested: makeFilter([{
      field: 'available_variation_data',
      value: 'ssm',
    }], false),
  },
  fragments: {
    cohort: () => Relay.QL`
      fragment on Cohort {
        cases {
          aggregations(filters: $ssmTested) {
            project__project_id {
              buckets {
                doc_count
                key
              }
            }
          }
          hits(first: 0, filters: $ssmTested) { total }
        }
        ssms @include (if: $fetchData) {
          hits(first: $fmTable_size offset: $fmTable_offset filters: $fmTable_filters, score: $score) {
            total
            edges {
              node {
                score
                genomic_dna_change
                mutation_subtype
                ssm_id
                consequence {
                  hits(first: 1 filters: $consequenceFilters) {
                    edges {
                      node {
                        transcript {
                          is_canonical
                          annotation {
                            impact
                          }
                          consequence_type
                          gene {
                            gene_id
                            symbol
                          }
                          aa_change
                        }
                      }
                    }
                  }
                }
                filteredOccurrences: occurrence {
                  hits(first: 1 filters: $fmTable_filters) {
                    total
                  }
                }
                allOccurrences: occurrence {
                  hits(first: 1000) {
                    total
                    edges {
                      node {
                        case {
                          case_id
                          project { project_id }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    `,
  },
};

const FrequentMutationsTable = Relay.createContainer(
  FrequentMutationsTableComponent,
  FrequentMutationsTableQuery
);

export default FrequentMutationsTable;
