// @ncigdc/containers/GeneProjectBreakdown.js
// @flow

import React from 'react';
import Relay from 'react-relay';
import { compose, withState } from 'recompose';
import { makeFilter } from '@ncigdc/utils/filters';
import TogglableUl from '@ncigdc/uikit/TogglableUl';

const GeneProjectBreakdownComponent = compose(
  withState('loading', 'setLoading', false)
)(({
  loading,
  setLoading,
  cohort: { cases },
  numCasesAggByProject,
  allCases,
  relay,
  id,
}) => {
  const filteredAggs = !cases.aggregations ? {} : cases.aggregations.project__project_id.buckets.reduce((acc, b) => ({
    ...acc,
    [b.key]: b.doc_count,
  }), {});

  return (
    <TogglableUl
      onToggle={() => {
        relay.setVariables({
          fetchData: true,
          filters: makeFilter([{ field: 'gene.gene_id', value: id }], false),
        });
        setLoading(l => !l);
      }}
      items={[
        <span key="total">
          {allCases.hits.total.toLocaleString()} / {cases.hits.total.toLocaleString()}
          &nbsp;({((allCases.hits.total / cases.hits.total) * 100).toFixed(2)}%)
        </span>,
        ...(loading && !cases.aggregations && [<i className="fa fa-spinner fa-spin" />]),
        ...Object.entries(filteredAggs)
          .sort(([ak, av], [bk, bv]) => (bv / numCasesAggByProject[bk]) - (av / numCasesAggByProject[ak]))
          .map(([k, v]) => (
            <span key={k}>
              {k}: {v} / {numCasesAggByProject[k]}
              &nbsp;({((v / numCasesAggByProject[k]) * 100).toFixed(2)}%)
            </span>
          )),
      ]}
    />
  );
});

export const GeneProjectBreakdownQuery = {
  initialVariables: {
    fetchData: false,
    filters: null,
    ssmTested: makeFilter([{
      field: 'available_variation_data',
      value: 'ssm',
    }], false),
  },
  fragments: {
    cohort: () => Relay.QL`
      fragment on Cohort {
        cases {
          hits(first: 1 filters: $ssmTested) { total }
          aggregations(filters: $filters) @include(if: $fetchData) {
            project__project_id {
              buckets {
                doc_count
                key
              }
            }
          }
        }
      }
    `,
  },
};

const GeneProjectBreakdown = Relay.createContainer(
  GeneProjectBreakdownComponent,
  GeneProjectBreakdownQuery
);

export default GeneProjectBreakdown;
