// @ncigdc/containers/HomePage.js
/* @flow */

import React from 'react';
import Relay from 'react-relay';

import Home from '@ncigdc/components/Home';

import ProjectsCount from './ProjectsCount';

export type TProps = {
  viewer: {
    projects: {
      hits: {
        edges: Array<{|
          node: {|
           id: string,
         |},
       |}>,
      },
    },
  },
};

export const HomePageComponent = (props: TProps) => (
  <Home
    humanBodyData={props.viewer.projects.hits.edges.map(x => x.node)}
    projectsCountsData={props.viewer.projects.hits}
  />
);

export const HomePageQuery = {
  initialVariables: {
    size: 1000,
    offset: 0,
    filters: null,
    sort: 'summary.case_count:desc',
  },
  fragments: {
    viewer: () => Relay.QL`
      fragment on Root {
        projects {
          hits(first: $size offset: $offset, sort: $sort, filters: $filters) {
            edges {
              node {
                id
                project_id
                primary_site
                summary {
                  case_count
                  file_count
                }
              }
            }
            ${ProjectsCount.getFragment('hits')}
          }
        }
      }
    `,
  },
};

const HomePage = Relay.createContainer(
  HomePageComponent,
  HomePageQuery
);

export default HomePage;
