// @ncigdc/containers/MostAffectedCasesChart.js
// @flow

import React from 'react';
import Relay from 'react-relay';
import withSize from 'react-sizeme';
import { compose, lifecycle } from 'recompose';
import withRouter from '@ncigdc/utils/withRouter';
import { parseFilterParam } from '@ncigdc/utils/uri';
import SpinnerParticle from '@ncigdc/uikit/Loaders/Particle';
import { Row } from '@ncigdc/uikit/Flex';
import BarChart from '@ncigdc/components/Charts/BarChart';
import { withTheme } from '@ncigdc/theme';

const MostAffectedCasesChartComponent = compose(
  withTheme,
  withSize(),
  withRouter,
  lifecycle({
    componentDidMount() {
      this.props.relay.setVariables({
        fetchData: true,
        macChart_filters: parseFilterParam(
          this.props.query.macChart_filters,
          this.props.defaultFilters || null
        ),
      });
    },
  })
)(({
  cohort: { cases },
  theme,
  size: {
    width,
  },
  push,
}) => {
  if (!cases) {
    return (
      <Row style={{ justifyContent: 'center', padding: '1rem' }}>
        <SpinnerParticle />
      </Row>
    );
  }

  const chartMargin = { top: 30, right: 50, bottom: 105, left: 40 };

  const bandWidth = (
    (width - chartMargin.right - chartMargin.left) /
    (cases.hits.edges.length + 1) / 2
  ) * 0.7;

  return (
    <span>
      {!!cases.hits.edges.length &&
        <Row style={{ padding: '0 2rem' }}>
          <BarChart
            data={cases.hits.edges.map(x => x.node).map(c => ({
              label: `${c.case_id.substring(0, 8)}\u2026`,
              value: c.score,
              tooltip: <span>{c.case_id}<br />{c.score} Genes Affected</span>,
              onClick: () => push(`/cases/${c.case_id}`),
            }))}
            margin={chartMargin}
            height={300}
            bandwidth={bandWidth}
            yAxis={{ title: '# Affected Genes' }}
            styles={{
              xAxis: { stroke: theme.greyScale4, textFill: theme.greyScale3 },
              yAxis: { stroke: theme.greyScale4, textFill: theme.greyScale3 },
              bars: { fill: theme.secondary },
              tooltips: {
                fill: '#fff',
                stroke: theme.greyScale4,
                textFill: theme.greyScale3,
              },
            }}
          />
        </Row>
      }
    </span>
  );
});

export const MostAffectedCasesChartQuery = {
  initialVariables: {
    fetchData: false,
    macChart_filters: null,
    score: 'gene.gene_id',
  },
  fragments: {
    cohort: () => Relay.QL`
      fragment on Cohort {
        allCases: cases {
          hits(first: 0) { total }
        }
        cases @include(if: $fetchData) {
          hits (first: 20 filters: $macChart_filters, score: $score) @include(if: $fetchData) {
            total
            edges {
              node {
                score
                case_id
              }
            }
          }
        }
      }
    `,
  },
};

const MostAffectedCasesChart = Relay.createContainer(
  MostAffectedCasesChartComponent,
  MostAffectedCasesChartQuery
);

export default MostAffectedCasesChart;
