// @ncigdc/containers/MutationsCount.js
// @flow
import React from 'react';
import Relay from 'react-relay';
import { lifecycle, compose } from 'recompose';
import { makeFilter } from '@ncigdc/utils/filters';
import GreyBox from '@ncigdc/uikit/GreyBox';
import { withTheme } from '@ncigdc/theme';

const MutationsCountComponent = compose(
  withTheme,
  lifecycle({
    componentDidMount(): void {
      this.props.relay.setVariables({
        fetchData: true,
        ssmsFilters: this.props.filters,
      });
    },
  })
)(({
  ssms,
  theme,
} = {}) => ssms.hits ?
  <span>{ssms.hits.total}</span> : <GreyBox />
);

export const MutationsCountQuery = {
  initialVariables: {
    ssmsFilters: null,
    fetchData: false,
  },
  fragments: {
    ssms: () => Relay.QL`
      fragment on Ssms {
        blah: hits(first: 0) { total }
        hits(filters: $ssmsFilters) @include(if: $fetchData) {
          total
        }
      }
    `,
  },
};

const MutationsCount = Relay.createContainer(
  MutationsCountComponent,
  MutationsCountQuery
);

export default MutationsCount;
