// @ncigdc/containers/ProjectPage.js
/* @flow */

import React from 'react';
import Relay from 'react-relay';

import SideNavLayout from '@ncigdc/components/Layouts/SideNavLayout';

export type TProps = {
  node: {
    disease_type: string,
    name: string,
    primary_site: string,
    program: {
      name: string,
    },
    project_id: string,
  },
};

export const ProjectPageComponent = (props: TProps) => (
  <SideNavLayout
    links={[
      { icon: 'table', title: 'Summary', id: 'summary' },
      { icon: 'bar-chart-o', title: 'Mutated Genes', id: 'mutated-genes' },
      { icon: 'th', title: 'OncoGrid', id: 'oncogrid' },
      { icon: 'bar-chart-o', title: 'Frequent Mutations', id: 'frequent-mutations' },
      { icon: 'bar-chart-o', title: 'Most Affected Cases', id: 'most-affected-cases' },
    ]}
    title={props.node.project_id}
    entityType="PR"
  >
    <div>{props.node.project_id}</div>
    <div>{props.node.name}</div>
    <div>{props.node.disease_type}</div>
    <div>{props.node.program.name}</div>
    <div>{props.node.primary_site}</div>
  </SideNavLayout>
);

export const ProjectPageQuery = {
  fragments: {
    node: () => Relay.QL`
      fragment on Project {
        project_id
        name
        disease_type
        program {
          name
        }
        primary_site
      }
    `,
  },
};

const ProjectPage = Relay.createContainer(
  ProjectPageComponent,
  ProjectPageQuery
);

export default ProjectPage;
