// @ncigdc/containers/ProjectPage.js
/* @flow */

import React from 'react';
import Relay from 'react-relay';

import SideNavLayout from '@ncigdc/components/Layouts/SideNavLayout';
import Project from '@ncigdc/components/Project';

export type TProps = {
  node: {
    disease_type: string,
    name: string,
    primary_site: string,
    program: {
      name: string,
    },
    project_id: string,
    summary: {
      case_count: number,
      data_categories: {
        case_count: number,
        data_category: string,
        file_count: number,
      },
      experimental_strategies: {
        case_count: number,
        experimental_strategy: string,
        file_count: number,
      },
      file_count: number,
    },
  },
};

export const ProjectPageComponent = (props: TProps) => (
  <SideNavLayout
    links={[
      { icon: 'table', title: 'Summary', id: 'summary' },
      { icon: 'bar-chart-o', title: 'Mutated Genes', id: 'mutated-genes' },
      { icon: 'th', title: 'OncoGrid', id: 'oncogrid' },
      { icon: 'bar-chart-o', title: 'Frequent Mutations', id: 'frequent-mutations' },
      { icon: 'bar-chart-o', title: 'Most Affected Cases', id: 'most-affected-cases' },
    ]}
    title={props.node.project_id}
    entityType="PR"
  >
    <Project
      caseCount={props.node.summary.case_count}
      fileCount={props.node.summary.file_count}
      projectId={props.node.project_id}
      projectName={props.node.name}
      programName={props.node.program.name}
      diseaseType={props.node.disease_type}
      primarySite={props.node.primary_site}
      dataCategories={props.node.summary.data_categories}
      experimentalStrategies={props.node.summary.experimental_strategies}
    />
  </SideNavLayout>
);

export const ProjectPageQuery = {
  fragments: {
    node: () => Relay.QL`
      fragment on Project {
        project_id
        name
        disease_type
        program {
          name
        }
        primary_site
        summary {
          case_count
          file_count
          experimental_strategies {
            experimental_strategy
            file_count
            case_count
          }
          data_categories {
            data_category
            file_count
            case_count
          }
        }
      }
    `,
  },
};

const ProjectPage = Relay.createContainer(
  ProjectPageComponent,
  ProjectPageQuery
);

export default ProjectPage;
