// @ncigdc/containers/ProjectTr.js
/* @flow */

import React from 'react';
import Relay from 'react-relay';

import ProjectLink from '@ncigdc/components/Links/ProjectLink';
import { findDataCategory } from '@ncigdc/utils/data';

import type { TCategory } from '@ncigdc/utils/data/types';

import { Tr, Td } from '@ncigdc/uikit/Table';

import { withTheme } from '@ncigdc/theme';

export type TProps = {|
  index: number,
  node: {|
    disease_type: string,
    name: string,
    primary_site: string,
    program: {|
      name: string,
    |},
    project_id: string,
    summary: {|
      case_count: number,
      data_categories: Array<{|
        case_count?: number,
        data_category: TCategory,
      |}>,
      file_count: number,
    |},
  |},
  theme: Object,
|};

export const ProjectTrComponent = ({ node, index, theme }: TProps) => (
  <Tr
    style={{
      backgroundColor: index % 2 === 0 ? theme.tableStripe : '#fff',
    }}
  >
    <Td>
      <ProjectLink id={node.project_id} />
    </Td>
    <Td>{node.disease_type}</Td>
    <Td>{node.primary_site}</Td>
    <Td>{node.program.name}</Td>
    <Td>{node.summary.case_count}</Td>
    <Td>{findDataCategory('Seq', node.summary.data_categories).case_count}</Td>
    <Td>{findDataCategory('Exp', node.summary.data_categories).case_count}</Td>
    <Td>{findDataCategory('SNV', node.summary.data_categories).case_count}</Td>
    <Td>{findDataCategory('CNV', node.summary.data_categories).case_count}</Td>
    <Td>{findDataCategory('Clinical', node.summary.data_categories).case_count}</Td>
    <Td>{findDataCategory('Bio', node.summary.data_categories).case_count}</Td>
    <Td>{node.summary.file_count}</Td>
  </Tr>
);

export const ProjectTrQuery = {
  fragments: {
    node: () => Relay.QL`
      fragment on Project {
        project_id
        disease_type
        program {
          name
        }
        primary_site
        summary {
          case_count
          data_categories {
            case_count
            data_category
          }
          file_count
        }
      }
    `,
  },
};

const ProjectTr = Relay.createContainer(
  withTheme(ProjectTrComponent),
  ProjectTrQuery
);

export default ProjectTr;
