// @ncigdc/containers/ProjectBreakdown.js
// @flow

import React from 'react';
import Relay from 'react-relay';
import { compose, withState } from 'recompose';
import { makeFilter } from '@ncigdc/utils/filters';
import TogglableUl from '@ncigdc/uikit/TogglableUl';

const ProjectBreakdownComponent = compose(
  withState('loading', 'setLoading', false)
)(({
  loading,
  setLoading,
  cohort: { cases },
  caseTotal,
  relay,
  filters,
}) => {
  const allAggs = !cases.allAggs ? {} : cases.allAggs.project__project_id.buckets.reduce((acc, b) => ({
    ...acc,
    [b.key]: b.doc_count,
  }), {});

  const filteredAggs = !cases.aggregations ? {} : cases.aggregations.project__project_id.buckets.reduce((acc, b) => ({
    ...acc,
    [b.key]: b.doc_count,
  }), {});

  return (
    <TogglableUl
      onToggle={() => {
        relay.setVariables({
          fetchData: true,
          aggFilters: filters,
        });
        setLoading(l => !l);
      }}
      items={[
        <span key="total">
          {caseTotal.toLocaleString()} / {cases.hits.total.toLocaleString()}
          &nbsp;({((caseTotal / cases.hits.total) * 100).toFixed(2)}%)
        </span>,
        ...(loading && !cases.aggregations && [<i className="fa fa-spinner fa-spin" />]),
        ...Object.entries(filteredAggs)
          .sort(([ak, av], [bk, bv]) => (bv / allAggs[bk]) - (av / allAggs[ak]))
          .map(([k, v]) => (
            <span key={k}>
              {k}: {v} / {allAggs[k]}
              &nbsp;({((v / allAggs[k]) * 100).toFixed(2)}%)
            </span>
          )),
      ]}
    />
  );
});

export const ProjectBreakdownQuery = {
  initialVariables: {
    fetchData: false,
    aggFilters: null,
    ssmTested: makeFilter([{
      field: 'cases.available_variation_data',
      value: 'ssm',
    }], false),
  },
  fragments: {
    cohort: () => Relay.QL`
      fragment on Cohort {
        cases {
          hits(first: 0 filters: $ssmTested) { total }
          allAggs: aggregations(filters: $ssmTested) {
            project__project_id {
              buckets {
                doc_count
                key
              }
            }
          }
          aggregations(filters: $aggFilters) @include(if: $fetchData) {
            project__project_id {
              buckets {
                doc_count
                key
              }
            }
          }
        }
      }
    `,
  },
};

const ProjectBreakdown = Relay.createContainer(
  ProjectBreakdownComponent,
  ProjectBreakdownQuery
);

export default ProjectBreakdown;
