// @ncigdc/containers/SSMPage.js
/* @flow */

import React from 'react';
import Relay from 'react-relay';

import SideNavLayout from '@ncigdc/components/Layouts/SideNavLayout';
import SSM from '@ncigdc/components/SSM';

export type TProps = {
  node: {
    genomic_dna_change: string,
  },
};

export const SSMPageComponent = (props: TProps) => (
  <SideNavLayout
    links={[
      { icon: 'table', id: 'summary', title: 'Summary' },
    ]}
    title={props.node.genomic_dna_change}
    entityType={<i className="fa fa-file-o" />}
  >
    <SSM node={props.node} />
  </SideNavLayout>
);

export const SSMPageQuery = {
  fragments: {
    node: () => Relay.QL`
      fragment on CohortSSM {
        ssm_id
        chromosome
        start_position
        reference_allele
        genomic_dna_change
        tumor_allele
        variant_type
        ncbi_build
        consequence {
          consequence__transcript__aa_change
          consequence__transcript__consequence_type
          consequence__transcript__transcript_id
          consequence__transcript__is_canonical
          consequence__transcript__transcript_id
          consequence__transcript__gene_symbol
          consequence__transcript__annotation__impact
          consequence__transcript__annotation__hgvsc
          consequence__transcript__gene__gene_strand
          consequence__transcript__gene__gene_id

        }
        occurrence__case__case_id
        occurrence__case__project__project_id
        occurrence__case__project__disease_type
        occurrence__case__project__cancer_type
        occurrence__case__project__primary_site
      }
    `,
  },
};

const SSMPage = Relay.createContainer(
  SSMPageComponent,
  SSMPageQuery
);

export default SSMPage;
